<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Patrice Godard" />

<meta name="date" content="2022-01-14" />

<title>TKCat User guide</title>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>
<link href="libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="libs/highlightjs-9.12.0/highlight.js"></script>
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="libs/vis-9.1.0/vis-network.min.css" rel="stylesheet" />
<script src="libs/vis-9.1.0/vis-network.min.js"></script>
<script src="libs/visNetwork-binding-2.1.0/visNetwork.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">TKCat User guide</h1>
<h4 class="author">Patrice Godard</h4>
<h4 class="date">January 14, 2022</h4>

</div>


<style type="text/css">
code{
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<div style="width:200px;">
<p><img src="libs/figs/TKCat.png" style="width:100.0%" /></p>
</div>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>The aim of TKCat (Tailored Knowledge Catalog) is to facilitate the management of data from knowledge resources that are frequently used alone or together in research environments. In TKCat, knowledge resources are manipulated as modeled database (MDB) objects. These objects provide access to the data tables along with a general description of the resource and a detail data model generated with <a href="https://github.com/patzaw/ReDaMoR">ReDaMoR</a> documenting the tables, their fields and their relationships. These MDB are then gathered in catalogs that can be easily explored an shared. TKCat provides tools to easily subset, filter and combine MDBs and create new catalogs suited for specific needs.</p>
<p>The TKCat R package is licensed under <a href="https://www.gnu.org/licenses/gpl-3.0.en.html">GPL-3</a>.</p>
<p>This vignette focuses on a local usage of TKCat in R console. Two other vignettes describe more specifically how TKCat can be used with a <a href="https://clickhouse.com/">ClickHouse</a> database from a <a href="chTKCat-User-guide.html">user</a> or an <a href="chTKCat-Operations-manual.html">operational</a> perspectives. A final vignette is dedicated to an <a href="TKCat-Collections.html">extended documentation of collections</a>.</p>
</div>
<div id="modeled-databases-and-embedded-information" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Modeled databases and embedded information</h1>
<p>A modeled database (MDB) in TKCat gathers the following information:</p>
<ul>
<li>General database information including a mandatory <em>name</em> and optionally the following fields: <em>title</em>, <em>description</em>, <em>url</em>, <em>version</em> and <em>maintainer</em>.</li>
<li>A data model created using the <a href="https://github.com/patzaw/ReDaMoR">ReDaMoR</a> package.</li>
<li>A list of tables corresponding to reference concepts shared by different MDBs. The way these concepts are identified is defined in specific documents called collections.</li>
<li>The data themselves organized according to the data model.</li>
</ul>
<div id="reading-examples" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Reading examples</h2>
<div id="hpo" class="section level3" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> HPO</h3>
<p>A subset of the <a href="#hpo">Human Phenotype Ontology (HPO)</a> is provided within the <a href="https://github.com/patzaw/ReDaMoR">ReDaMoR</a> package. The HPO aims to provide a standardized vocabulary of phenotypic abnormalities encountered in human diseases <span class="citation">(<a href="#ref-kohler_expansion_2019" role="doc-biblioref">Köhler et al. 2019</a>)</span>. An MDB object based on files (see <a href="#mdb-implementations">MDB implementations</a>) can be read as shown below. As explained above, the data provided by the <code>path</code> parameter are documented with a model (<code>dataModel</code> parameter) and general information (<code>dbInfo</code> parameter).</p>
<pre class="r"><code>file_hpo &lt;- read_fileMDB(
   path=system.file(&quot;examples/HPO-subset&quot;, package=&quot;ReDaMoR&quot;),
   dataModel=system.file(&quot;examples/HPO-model.json&quot;, package=&quot;ReDaMoR&quot;),
   dbInfo=list(
      &quot;name&quot;=&quot;HPO&quot;,
      &quot;title&quot;=&quot;Data extracted from the HPO database&quot;,
      &quot;description&quot;=paste(
         &quot;This is a very small subset of the HPO!&quot;,
         &quot;Visit the reference URL for more information.&quot;
      ),
      &quot;url&quot;=&quot;http://human-phenotype-ontology.github.io/&quot;
   )
)</code></pre>
<pre><code>## HPO
## SUCCESS
## 
## Check configuration
##    - Optional checks: 
##    - Maximum number of records: 10</code></pre>
<p>The message displayed in the console indicates if the data fit the data model. It relies on the <code>ReDaMoR::confront_data()</code> functions and check by default the first 10 rows of each file.</p>
<p>The data model can then be drawn.</p>
<pre class="r"><code>plot(data_model(file_hpo))</code></pre>
<div id="htmlwidget-e439af4d688354b28c25" style="width:672px;height:480px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-e439af4d688354b28c25">{"x":{"nodes":{"tableName":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms"],"label":["<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_altId<\/b>\n    - id {character - idx.1}\n    - alt {character}","<b>HPO_sourceFiles<\/b>\n    - <b>*url<\/b> {character - uidx.1}\n    - current {Date}","<b>HPO_diseases<\/b>\n    - <b>db<\/b> {character - uidx.1}\n    - <b>id<\/b> {character - uidx.1}\n    - label {character}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_diseaseSynonyms<\/b>\n    - db {character - idx.1}\n    - id {character - idx.1}\n    - synonym {character}\n    - preferred {logical}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}","<b>HPO_descendants<\/b>\n    - id {character - idx.1}\n    - descendant {character - idx.2}","<b>HPO_synonyms<\/b>\n    - id {character}\n    - synonym {character}\n    - type {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_altId<\/strong> (Alternative HP identifiers)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>alt<\/strong>: Alternative identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_sourceFiles<\/strong> (Files from which data were extracted)<\/p> <ul> <li><strong>url<\/strong>: Source file URL<\/li> <li><strong>current<\/strong>: The date of the current source file<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong> (Diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>label<\/strong>: Disease lable (preferred synonym)<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseSynonyms<\/strong> (Disease synonyms)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>synonym<\/strong>: Disease synonym<\/li> <li><strong>preferred<\/strong>: Is synonym preferred<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_descendants<\/strong> (Indirect descendants of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>descendant<\/strong>: Descendant (child or child of child...) HP ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_synonyms<\/strong> (Synonyms of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>synonym<\/strong>: HP synonym<\/li> <li><strong>type<\/strong>: Synonym type<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left"],"x":[-167,-154,262,320,75,321,-328,-101,-360],"y":[-177,-38,-336,-170,-172,-24,-331,-360,-36],"color.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE"],"color.border":["black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE"],"id":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms"]},"edges":{"id":["HPO_altId->HPO_hp: id->id","HPO_diseaseHP->HPO_hp: hp->id","HPO_diseaseHP->HPO_diseases: db->db && id->id","HPO_diseaseSynonyms->HPO_diseases: db->db && id->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id","HPO_descendants->HPO_hp: id->id","HPO_descendants->HPO_hp: descendant->id","HPO_synonyms->HPO_hp: id->id"],"to":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseases","HPO_hp","HPO_hp","HPO_hp","HPO_hp","HPO_hp"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_altId<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseSynonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">descendant<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_synonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["id","hp",["db","id"],["db","id"],"id","parent","id","descendant","id"],"tf":["id","id",["db","id"],["db","id"],"id","id","id","id","id"],"from":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_descendants","HPO_descendants","HPO_synonyms"],"arrows":["to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0,0.2,-0.2,-0.2,0.2,0],"selfReferenceSize":[30,30,30,30,30,50,30,50,30],"uef":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseases","HPO_hp","HPO_hp","HPO_descendants","HPO_descendants","HPO_hp"],"uet":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_hp","HPO_hp","HPO_synonyms"],"ue":["HPO_altId<->HPO_hp","HPO_diseaseHP<->HPO_hp","HPO_diseaseHP<->HPO_diseases","HPO_diseases<->HPO_diseaseSynonyms","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents","HPO_descendants<->HPO_hp","HPO_descendants<->HPO_hp","HPO_hp<->HPO_synonyms"],"color.color":["black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
<p>In this model, the <em>HPO_hp</em> table refers to the concept of phenotype and the <em>HPO_disease</em> to the concept of disease. These concepts are used to define the condition of individuals. The <strong>Condition</strong> collection is built in the TKCat package. Identifying the collection members in an MDB is done by providing a table of the shape as displayed below and using the <code>collection_members()</code> function. As described in the <a href="#Merging-with-collections">Merging with collections</a> section, collections identify concepts shared by different MDB and can be used to merge resources according to these concepts.</p>
<pre class="r"><code>cn &lt;- c(
   &quot;collection&quot;, &quot;cid&quot;,                &quot;resource&quot;, &quot;mid&quot;, &quot;table&quot;,        &quot;field&quot;,     &quot;static&quot;, &quot;value&quot;,    &quot;type&quot;
)
cm &lt;- matrix(data=c(
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      1,     &quot;HPO_hp&quot;,       &quot;condition&quot;,  TRUE,    &quot;Phenotype&quot;, NA,
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      1,     &quot;HPO_hp&quot;,       &quot;source&quot;,     TRUE,    &quot;HP&quot;,        NA,
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      1,     &quot;HPO_hp&quot;,       &quot;identifier&quot;, FALSE,   &quot;id&quot;,        NA,
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      2,     &quot;HPO_diseases&quot;, &quot;condition&quot;,  TRUE,    &quot;Disease&quot;,   NA,
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      2,     &quot;HPO_diseases&quot;, &quot;source&quot;,     FALSE,   &quot;db&quot;,        NA,
   &quot;Condition&quot;,  &quot;HPO_conditions_1.0&quot;, &quot;HPO&quot;,      2,     &quot;HPO_diseases&quot;, &quot;identifier&quot;, FALSE,   &quot;id&quot;,        NA
   ),
   ncol=9, byrow=TRUE
) %&gt;%
   set_colnames(cn) %&gt;% 
   as_tibble() %&gt;% 
   mutate(mid=as.integer(mid), static=as.logical(static))
collection_members(file_hpo) &lt;- cm
file_hpo</code></pre>
<pre><code>## fileMDB HPO: Data extracted from the HPO database
##    - 9 tables with 25 fields
## 
## Collection members: 
##    - 2 Condition members
## 
## This is a very small subset of the HPO! Visit the reference URL for more information.
## (http://human-phenotype-ontology.github.io/)
## 
## </code></pre>
</div>
<div id="clinvar" class="section level3" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> ClinVar</h3>
<p>A subset of the <a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a> database is provided within this package. ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence <span class="citation">(<a href="#ref-landrum_clinvar_2018" role="doc-biblioref">Landrum et al. 2018</a>)</span>. This resource can be read as a <code>fileMDB</code> as shown above, excepted that all the documenting information is included in the resource directory in this case and it is organized as following:</p>
<ul>
<li><p><strong>DESCRIPTION.json</strong> contains db_information</p></li>
<li><p><strong>data</strong> is a directory with all the data tables</p></li>
<li><p><strong>model</strong> is a directory with model information:</p>
<ul>
<li>A <strong>json</strong> file with the ClinVar data model created with the <a href="https://github.com/patzaw/ReDaMoR">ReDaMoR</a> package</li>
<li>A <strong>Collections</strong> subfolder with one json file per collection with members in the ClinVar resource</li>
</ul></li>
</ul>
<pre class="r"><code>file_clinvar &lt;- read_fileMDB(
   path=system.file(&quot;examples/ClinVar&quot;, package=&quot;TKCat&quot;)
)</code></pre>
<pre><code>## ClinVar
## SUCCESS
## 
## Check configuration
##    - Optional checks: 
##    - Maximum number of records: 10</code></pre>
</div>
<div id="chembl" class="section level3" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> CHEMBL</h3>
<p>A self-documented subset of the <a href="https://www.ebi.ac.uk/chembl/">CHEMBL</a> database is also provided in this package. It can be read the same way as the ClinVar resource.</p>
<pre class="r"><code>file_chembl &lt;- read_fileMDB(
   path=system.file(&quot;examples/CHEMBL&quot;, package=&quot;TKCat&quot;)
)</code></pre>
<pre><code>## CHEMBL
## SUCCESS
## 
## Check configuration
##    - Optional checks: 
##    - Maximum number of records: 10</code></pre>
<p>CHEMBL is a manually curated chemical database of bioactive molecules with drug-like properties <span class="citation">(<a href="#ref-mendez_chembl_2019" role="doc-biblioref">Mendez et al. 2019</a>)</span>.</p>
</div>
</div>
<div id="mdb-implementations" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> MDB implementations</h2>
<p>There are 3 main implementations of MDBs:</p>
<ul>
<li><p><strong>fileMDB</strong> objects keep the data in files and load them only when requested by the user. These implementation is the first one which is used when reading MDB as demonstrated in the examples above.</p></li>
<li><p><strong>memoMDB</strong> objects have all the data loaded in memory. These objects are very easy to use but can take time to load and can use a lot of memory.</p></li>
<li><p><strong>chMDB</strong> objects get the data from a ClickHouse database providing a catalog of MDBs as described in the <a href="#chTKCat">chTKCat</a> section below. More information about chTKCat and chMDB objects can also be found in the <a href="chTKCat-User-guide.html">chTKCat user guide</a> and the <a href="chTKCat-Operations-manual.html">chTKCat operational manual</a>.</p></li>
</ul>
<p>The different implementations can be converted to each others using <code>as_fileMDB()</code>, <code>as_memoMDB()</code> and <code>as_chMDB()</code> functions.</p>
<pre class="r"><code>memo_clinvar &lt;- as_memoMDB(file_clinvar)
object.size(file_clinvar) %&gt;% print(units=&quot;Kb&quot;)</code></pre>
<pre><code>## 155.2 Kb</code></pre>
<pre class="r"><code>object.size(memo_clinvar) %&gt;% print(units=&quot;Kb&quot;)</code></pre>
<pre><code>## 760.5 Kb</code></pre>
<p>A fourth implementation is <strong>metaMDB</strong> which combines several MDBs glued together with relational tables (see the <a href="#Merging-with-collections">Merging with collections</a> section).</p>
<p>Most of the functions described below work with any MDB implementation. A few functions are specific to each implementation.</p>
</div>
<div id="getting-information" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Getting information</h2>
<div id="general-information" class="section level3" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> General information</h3>
<pre class="r"><code>db_info(file_clinvar)</code></pre>
<pre><code>## $name
## [1] &quot;ClinVar&quot;
## 
## $title
## [1] &quot;Data extracted from the ClinVar database&quot;
## 
## $description
## [1] &quot;ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. This is a very small subset of ClinVar! Visit the reference URL for more information.&quot;
## 
## $url
## [1] &quot;https://www.ncbi.nlm.nih.gov/clinvar/&quot;
## 
## $version
## [1] &quot;0.9&quot;
## 
## $maintainer
## [1] &quot;Patrice Godard &lt;patrice.godard@ucb.com&gt;&quot;
## 
## $timestamp
## [1] NA</code></pre>
<p>The function <code>db_info()&lt;-</code> can be used to update this information.</p>
</div>
<div id="data-model" class="section level3" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Data model</h3>
<p>As shown above the data model of an MDB can be retrieved and plot the following way.</p>
<pre class="r"><code>plot(data_model(file_clinvar))</code></pre>
<div id="htmlwidget-4ab3c932166b3d4c2a5a" style="width:672px;height:480px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-4ab3c932166b3d4c2a5a">{"x":{"nodes":{"tableName":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames"],"label":["<b>ClinVar_ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>ClinVar_rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVar_ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>ClinVar_rcvaInhMode<\/b>\n    - rcvaId {integer}\n    - inhMode {character}","<b>ClinVar_rcvaObservedIn<\/b>\n    - rcvaId {integer}\n    - origin {character}\n    - taxonomyId {character}\n    - species {character}\n    - affectedStatus {character}\n    - (numberTested {integer})","<b>ClinVar_rcvaTraits<\/b>\n    - rcvaId {integer}\n    - t.id {integer - idx.1}\n    - traitType {character}","<b>ClinVar_clinSigOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_revStatOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_variants<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - type {character}\n    - (name {character})","<b>ClinVar_cvaObservedIn<\/b>\n    - cvaId {integer}\n    - origin {character}\n    - species {character}\n    - affectedStatus {character}","<b>ClinVar_cvaSubmitters<\/b>\n    - cvaId {integer}\n    - submitter {character}\n    - primary {logical}","<b>ClinVar_traits<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - name {character}","<b>ClinVar_varEntrez<\/b>\n    - varId {integer - idx.1}\n    - entrez {integer}\n    - type {character}","<b>ClinVar_varAttributes<\/b>\n    - varId {integer - idx.1}\n    - Type {character}\n    - (integerValue {integer})\n    - (Change {character})\n    - (value {character})","<b>ClinVar_varCytoLoc<\/b>\n    - varId {integer - idx.1}\n    - location {character}","<b>ClinVar_varNames<\/b>\n    - varId {integer - idx.1}\n    - name {character}\n    - type {character}","<b>ClinVar_varSeqLoc<\/b>\n    - varId {integer - idx.1}\n    - Accession {character}\n    - (alternateAllele {character})\n    - Assembly {character}\n    - AssemblyAccessionVersion {character}\n    - AssemblyStatus {character}\n    - Chr {character}\n    - display_start {integer}\n    - (display_stop {integer})\n    - (innerStart {integer})\n    - (innerStop {integer})\n    - (outerStart {integer})\n    - (outerStop {integer})\n    - (referenceAllele {character})\n    - (start {integer})\n    - (stop {integer})\n    - (Strand {character})\n    - (variantLength {integer})","<b>ClinVar_varXRef<\/b>\n    - varId {integer - idx.1}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitCref<\/b>\n    - t.id {integer}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitNames<\/b>\n    - t.id {integer}\n    - name {character}\n    - type {character}","<b>ClinVar_entrezNames<\/b>\n    - <b>*entrez<\/b> {integer - uidx.1}\n    - name {character}\n    - symbol {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaInhMode<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>inhMode<\/strong>: Mode of inheritance<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaObservedIn<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>taxonomyId<\/strong>: Taxonomy ID<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <li><strong>numberTested<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaTraits<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>traitType<\/strong>: Trait type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_clinSigOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of clinical significance: the lowest the best clinical outcome.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_revStatOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of review status: the lowest the least significant.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_variants<\/strong><\/p> <ul> <li><strong>id<\/strong>: Variant ID<\/li> <li><strong>type<\/strong>: Variant type<\/li> <li><strong>name<\/strong>: Variant preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaObservedIn<\/strong><\/p> <ul> <li><strong>cvaId<\/strong><\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaSubmitters<\/strong><\/p> <ul> <li><strong>cvaId<\/strong>: ClinVar Assertion ID<\/li> <li><strong>submitter<\/strong>: Submitter name<\/li> <li><strong>primary<\/strong>: The submitter is of pirmary type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traits<\/strong><\/p> <ul> <li><strong>id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varEntrez<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>type<\/strong>: Type of the association between the variant and the gene<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varAttributes<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Type<\/strong><\/li> <li><strong>integerValue<\/strong><\/li> <li><strong>Change<\/strong><\/li> <li><strong>value<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varCytoLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>location<\/strong>: Cytogenic location<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varNames<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>name<\/strong>: Variant name<\/li> <li><strong>type<\/strong>: Variant name type: Preferred or Alternate<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varSeqLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Accession<\/strong>: Sequence accession ID<\/li> <li><strong>alternateAllele<\/strong>: Alternate allele<\/li> <li><strong>Assembly<\/strong>: Assembly<\/li> <li><strong>AssemblyAccessionVersion<\/strong>: Version of assembly accession<\/li> <li><strong>AssemblyStatus<\/strong>: Status of the assembly: current or previous<\/li> <li><strong>Chr<\/strong>: Chromosome name<\/li> <li><strong>display_start<\/strong><\/li> <li><strong>display_stop<\/strong><\/li> <li><strong>innerStart<\/strong><\/li> <li><strong>innerStop<\/strong><\/li> <li><strong>outerStart<\/strong><\/li> <li><strong>outerStop<\/strong><\/li> <li><strong>referenceAllele<\/strong>: Reference allele<\/li> <li><strong>start<\/strong><\/li> <li><strong>stop<\/strong><\/li> <li><strong>Strand<\/strong><\/li> <li><strong>variantLength<\/strong>: Length of the variant<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varXRef<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitCref<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitNames<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait name<\/li> <li><strong>type<\/strong>: \"Alternate\" or \"Preferred\"<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_entrezNames<\/strong><\/p> <ul> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>name<\/strong>: Gene name<\/li> <li><strong>symbol<\/strong>: Gene symbol<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left"],"x":[-370,55,-767,-311,-114,-368,-540,-443,430,-766,-771,-371,266,243,486,494,751,721,-507,-264,265],"y":[344,345,338,521,498,168,524,603,346,186,488,31,178,503,489,194,214,461,-87,-91,37],"color.background":["#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0"],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0"],"id":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames"]},"edges":{"id":["ClinVar_ReferenceClinVarAssertion->ClinVar_clinSigOrder: clinicalSignificance->label","ClinVar_ReferenceClinVarAssertion->ClinVar_revStatOrder: reviewStatus->label","ClinVar_rcvaVariant->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaVariant->ClinVar_variants: varId->id","ClinVar_ClinVarAssertions->ClinVar_ReferenceClinVarAssertion: cvs->cvs","ClinVar_rcvaInhMode->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaObservedIn->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaTraits->ClinVar_traits: t.id->id","ClinVar_rcvaTraits->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_cvaObservedIn->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_cvaSubmitters->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_varEntrez->ClinVar_variants: varId->id","ClinVar_varEntrez->ClinVar_entrezNames: entrez->entrez","ClinVar_varAttributes->ClinVar_variants: varId->id","ClinVar_varCytoLoc->ClinVar_variants: varId->id","ClinVar_varNames->ClinVar_variants: varId->id","ClinVar_varSeqLoc->ClinVar_variants: varId->id","ClinVar_varXRef->ClinVar_variants: varId->id","ClinVar_traitCref->ClinVar_traits: t.id->id","ClinVar_traitNames->ClinVar_traits: t.id->id"],"to":["ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_variants","ClinVar_entrezNames","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traits","ClinVar_traits"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_clinSigOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">clinicalSignificance<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_revStatOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">reviewStatus<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaInhMode<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaSubmitters<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_entrezNames<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varAttributes<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varCytoLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varSeqLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varXRef<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["clinicalSignificance","reviewStatus","rcvaId","varId","cvs","rcvaId","rcvaId","t.id","rcvaId","cvaId","cvaId","varId","entrez","varId","varId","varId","varId","varId","t.id","t.id"],"tf":["label","label","id","id","cvs","id","id","id","id","id","id","id","entrez","id","id","id","id","id","id","id"],"from":["ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_varEntrez","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames"],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[0,-0,-0,-0,-0,-0,-0,-0,-0,0,0,-0,0,-0,-0,0,0,0,-0,-0],"selfReferenceSize":[30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30],"uef":["ClinVar_clinSigOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_varEntrez","ClinVar_entrezNames","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traitCref","ClinVar_traitNames"],"uet":["ClinVar_ReferenceClinVarAssertion","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_variants","ClinVar_varEntrez","ClinVar_variants","ClinVar_variants","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traits","ClinVar_traits"],"ue":["ClinVar_clinSigOrder<->ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion<->ClinVar_revStatOrder","ClinVar_rcvaVariant<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant<->ClinVar_variants","ClinVar_ClinVarAssertions<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaInhMode<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaObservedIn<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaTraits<->ClinVar_traits","ClinVar_rcvaTraits<->ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions<->ClinVar_cvaObservedIn","ClinVar_ClinVarAssertions<->ClinVar_cvaSubmitters","ClinVar_varEntrez<->ClinVar_variants","ClinVar_entrezNames<->ClinVar_varEntrez","ClinVar_varAttributes<->ClinVar_variants","ClinVar_varCytoLoc<->ClinVar_variants","ClinVar_variants<->ClinVar_varNames","ClinVar_variants<->ClinVar_varSeqLoc","ClinVar_variants<->ClinVar_varXRef","ClinVar_traitCref<->ClinVar_traits","ClinVar_traitNames<->ClinVar_traits"],"color.color":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
<p>Tables names can be listed with the <code>names()</code> function and changed with <code>names()&lt;-</code> or <code>rename()</code>.</p>
<pre class="r"><code>names(file_clinvar)</code></pre>
<pre><code>##  [1] &quot;ClinVar_ReferenceClinVarAssertion&quot; &quot;ClinVar_rcvaVariant&quot;              
##  [3] &quot;ClinVar_ClinVarAssertions&quot;         &quot;ClinVar_rcvaInhMode&quot;              
##  [5] &quot;ClinVar_rcvaObservedIn&quot;            &quot;ClinVar_rcvaTraits&quot;               
##  [7] &quot;ClinVar_clinSigOrder&quot;              &quot;ClinVar_revStatOrder&quot;             
##  [9] &quot;ClinVar_variants&quot;                  &quot;ClinVar_cvaObservedIn&quot;            
## [11] &quot;ClinVar_cvaSubmitters&quot;             &quot;ClinVar_traits&quot;                   
## [13] &quot;ClinVar_varEntrez&quot;                 &quot;ClinVar_varAttributes&quot;            
## [15] &quot;ClinVar_varCytoLoc&quot;                &quot;ClinVar_varNames&quot;                 
## [17] &quot;ClinVar_varSeqLoc&quot;                 &quot;ClinVar_varXRef&quot;                  
## [19] &quot;ClinVar_traitCref&quot;                 &quot;ClinVar_traitNames&quot;               
## [21] &quot;ClinVar_entrezNames&quot;</code></pre>
<p>The different collection members of an MDBs are listed with the <code>collection_members()</code> function and updated with <code>collection_members()&lt;-</code>.</p>
<pre class="r"><code>collection_members(file_clinvar)</code></pre>
<pre><code>## # A tibble: 10 × 9
##    collection cid        resource   mid table     field   static value   type   
##    &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;   &lt;lgl&gt;  &lt;chr&gt;   &lt;chr&gt;  
##  1 BE         ClinVar_B… ClinVar      1 ClinVar_… be      TRUE   Gene    &lt;NA&gt;   
##  2 BE         ClinVar_B… ClinVar      1 ClinVar_… identi… FALSE  entrez  &lt;NA&gt;   
##  3 BE         ClinVar_B… ClinVar      1 ClinVar_… organi… TRUE   Homo s… Scient…
##  4 BE         ClinVar_B… ClinVar      1 ClinVar_… source  TRUE   Entrez… &lt;NA&gt;   
##  5 Condition  ClinVar_c… ClinVar      1 ClinVar_… condit… TRUE   Disease &lt;NA&gt;   
##  6 Condition  ClinVar_c… ClinVar      1 ClinVar_… identi… FALSE  id      &lt;NA&gt;   
##  7 Condition  ClinVar_c… ClinVar      1 ClinVar_… source  FALSE  db      &lt;NA&gt;   
##  8 Condition  ClinVar_c… ClinVar      2 ClinVar_… condit… TRUE   Disease &lt;NA&gt;   
##  9 Condition  ClinVar_c… ClinVar      2 ClinVar_… identi… FALSE  id      &lt;NA&gt;   
## 10 Condition  ClinVar_c… ClinVar      2 ClinVar_… source  TRUE   ClinVar &lt;NA&gt;</code></pre>
</div>
<div id="size" class="section level3" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> Size</h3>
<p>The following functions are use to get the number of tables, the number of fields per table and the number of records.</p>
<pre class="r"><code>length(file_clinvar)        # Number of tables</code></pre>
<pre><code>## [1] 21</code></pre>
<pre class="r"><code>lengths(file_clinvar)       # Number of fields per table</code></pre>
<pre><code>## ClinVar_ReferenceClinVarAssertion               ClinVar_rcvaVariant 
##                                 8                                 2 
##         ClinVar_ClinVarAssertions               ClinVar_rcvaInhMode 
##                                 4                                 2 
##            ClinVar_rcvaObservedIn                ClinVar_rcvaTraits 
##                                 6                                 3 
##              ClinVar_clinSigOrder              ClinVar_revStatOrder 
##                                 2                                 2 
##                  ClinVar_variants             ClinVar_cvaObservedIn 
##                                 3                                 4 
##             ClinVar_cvaSubmitters                    ClinVar_traits 
##                                 3                                 2 
##                 ClinVar_varEntrez             ClinVar_varAttributes 
##                                 3                                 5 
##                ClinVar_varCytoLoc                  ClinVar_varNames 
##                                 2                                 3 
##                 ClinVar_varSeqLoc                   ClinVar_varXRef 
##                                18                                 4 
##                 ClinVar_traitCref                ClinVar_traitNames 
##                                 4                                 3 
##               ClinVar_entrezNames 
##                                 3</code></pre>
<pre class="r"><code>count_records(file_clinvar) # Number of records per table</code></pre>
<pre><code>## ClinVar_ReferenceClinVarAssertion               ClinVar_rcvaVariant 
##                               166                               166 
##         ClinVar_ClinVarAssertions               ClinVar_rcvaInhMode 
##                               409                                16 
##            ClinVar_rcvaObservedIn                ClinVar_rcvaTraits 
##                               337                               166 
##              ClinVar_clinSigOrder              ClinVar_revStatOrder 
##                                11                                 2 
##                  ClinVar_variants             ClinVar_cvaObservedIn 
##                               138                               412 
##             ClinVar_cvaSubmitters                    ClinVar_traits 
##                               416                                18 
##                 ClinVar_varEntrez             ClinVar_varAttributes 
##                               145                              2262 
##                ClinVar_varCytoLoc                  ClinVar_varNames 
##                               138                               188 
##                 ClinVar_varSeqLoc                   ClinVar_varXRef 
##                               280                               244 
##                 ClinVar_traitCref                ClinVar_traitNames 
##                                50                                44 
##               ClinVar_entrezNames 
##                                20</code></pre>
<p>The <code>count_records()</code> function can take a lot of time when dealing with <em>fileMDB</em> objects if the data files are very large. In such case it could be more clever to list data file size.</p>
<pre class="r"><code>data_file_size(file_clinvar, hr=TRUE)</code></pre>
<pre><code>## # A tibble: 21 × 3
##    table                             size   compressed
##    &lt;chr&gt;                             &lt;chr&gt;  &lt;lgl&gt;     
##  1 ClinVar_ReferenceClinVarAssertion 4.6 KB TRUE      
##  2 ClinVar_rcvaVariant               947 B  TRUE      
##  3 ClinVar_ClinVarAssertions         4.2 KB TRUE      
##  4 ClinVar_rcvaInhMode               152 B  TRUE      
##  5 ClinVar_rcvaObservedIn            1.4 KB TRUE      
##  6 ClinVar_rcvaTraits                788 B  TRUE      
##  7 ClinVar_clinSigOrder              145 B  TRUE      
##  8 ClinVar_revStatOrder              101 B  TRUE      
##  9 ClinVar_variants                  2.1 KB TRUE      
## 10 ClinVar_cvaObservedIn             1.8 KB TRUE      
## # … with 11 more rows</code></pre>
</div>
</div>
<div id="pulling-subsetting-and-combining" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Pulling, subsetting and combining</h2>
<p>There are several possible ways to pull data tables from MDBs. The following lines return the same results.</p>
<pre class="r"><code>data_tables(file_clinvar, &quot;ClinVar_traitNames&quot;)[[1]]
file_clinvar[[&quot;ClinVar_traitNames&quot;]]
file_clinvar$&quot;ClinVar_traitNames&quot;</code></pre>
<pre class="r"><code>file_clinvar %&gt;% pull(ClinVar_traitNames)</code></pre>
<pre><code>## # A tibble: 44 × 3
##     t.id name                                                           type    
##    &lt;int&gt; &lt;chr&gt;                                                          &lt;chr&gt;   
##  1   912 Chudley-McCullough syndrome                                    Preferr…
##  2   912 Deafness, autosomal recessive 82                               Alterna…
##  3   912 Deafness, bilateral sensorineural, and hydrocephalus due to f… Alterna…
##  4   912 Deafness, sensorineural, with partial agenesis of the corpus … Alterna…
##  5  1352 CTSD-Related Neuronal Ceroid-Lipofuscinosis                    Alterna…
##  6  1352 Ceroid lipofuscinosis neuronal Cathepsin D-deficient           Alterna…
##  7  1352 Neuronal ceroid lipofuscinosis 10                              Preferr…
##  8  1352 Neuronal ceroid lipofuscinosis due to Cathepsin D deficiency   Alterna…
##  9  1481 Diabetes mellitus, neonatal, with congenital hypothyroidism    Preferr…
## 10  1481 NDH SYNDROME                                                   Alterna…
## # … with 34 more rows</code></pre>
<p>MDBs can also be subset and combined. The corresponding functions ensure that the data model is fulfilled by the data tables.</p>
<pre class="r"><code>file_clinvar[1:3]</code></pre>
<pre><code>## fileMDB ClinVar (version 0.9, Patrice Godard &lt;patrice.godard@ucb.com&gt;): Data extracted from the ClinVar database
##    - 3 tables with 14 fields
## 
## No collection member
## 
## ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. This is a very small subset of ClinVar! Visit the reference URL for more information.
## (https://www.ncbi.nlm.nih.gov/clinvar/)
## 
## </code></pre>
<pre class="r"><code>c(file_clinvar[1:3], file_hpo[c(1,5,7)]) %&gt;% 
   data_model() %&gt;% auto_layout(force=TRUE) %&gt;% plot()</code></pre>
<div id="htmlwidget-ed542d28949ecba29970" style="width:672px;height:480px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-ed542d28949ecba29970">{"x":{"nodes":{"tableName":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","HPO_hp","HPO_diseaseHP","HPO_parents"],"label":["<b>ClinVar_ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>ClinVar_rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVar_ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true],"font.align":["left","left","left","left","left","left"],"x":[-133.313834862548,3.07690002774892,-240,127.40272483682,240,20.1312321301397],"y":[141.958860941012,240,11.271923922906,-156.376546231074,-26.3201500676791,-240],"color.background":["#DFFB86","#D9D9D9","#DFFB86","#9BC8FE","#D9D9D9","#F67FC4"],"color.border":["black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#DFFB86","#D9D9D9","#DFFB86","#9BC8FE","#D9D9D9","#F67FC4"],"id":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","HPO_hp","HPO_diseaseHP","HPO_parents"]},"edges":{"id":["ClinVar_rcvaVariant->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_ClinVarAssertions->ClinVar_ReferenceClinVarAssertion: cvs->cvs","HPO_diseaseHP->HPO_hp: hp->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id"],"to":["ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","HPO_hp","HPO_hp","HPO_hp"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["rcvaId","cvs","hp","id","parent"],"tf":["id","cvs","id","id","id"],"from":["ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","HPO_diseaseHP","HPO_parents","HPO_parents"],"arrows":["to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0.2,-0.2],"selfReferenceSize":[30,30,30,30,50],"uef":["ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","HPO_diseaseHP","HPO_hp","HPO_hp"],"uet":["ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","HPO_hp","HPO_parents","HPO_parents"],"ue":["ClinVar_rcvaVariant<->ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions<->ClinVar_ReferenceClinVarAssertion","HPO_diseaseHP<->HPO_hp","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents"],"color.color":["black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
<p>The function <code>c()</code> concatenates the provided MDB after checking that tables names are not duplicated. It does not integrate the data with any relational table. This can achieved by merging the MDBs as described in the <a href="#Merging-with-collections">Merging with collections</a> section.</p>
</div>
<div id="filtering-and-joining" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Filtering and joining</h2>
<p>An MDB can be filtered by filtering one or several tables based on field values. The filtering is propagated to other tables using the embedded data model.</p>
<p>In the example below, the <code>file_clinvar</code> object is filtered in order to focus on a few genes with pathogenic variants (the tables have been renamed using the <code>set_names()</code> function to improve the readability of the example). The object returned by <code>filter()</code> or <code>slice</code> is a <em>memoMDB</em>: all the data are in memory.</p>
<pre class="r"><code>filtered_clinvar &lt;- file_clinvar %&gt;% 
   set_names(sub(&quot;ClinVar_&quot;, &quot;&quot;, names(.))) %&gt;%
   filter(
      entrezNames = symbol %in% c(&quot;PIK3R2&quot;, &quot;UGT1A8&quot;)
   ) %&gt;% 
   slice(ReferenceClinVarAssertion=grep(
      &quot;pathogen&quot;,
      .$ReferenceClinVarAssertion$clinicalSignificance,
      ignore.case=TRUE
   ))</code></pre>
<p>Tables can be easily joined to get diseases associated to the genes of interest.</p>
<pre class="r"><code>gene_traits &lt;- filtered_clinvar %&gt;% 
   join_mdb_tables(
      &quot;entrezNames&quot;, &quot;varEntrez&quot;, &quot;variants&quot;, &quot;rcvaVariant&quot;,
      &quot;ReferenceClinVarAssertion&quot;, &quot;rcvaTraits&quot;, &quot;traits&quot;
   )
gene_traits$entrezNames %&gt;%
   select(symbol, name, variants.type, variants.name, traitType, traits.name)</code></pre>
<pre><code>## # A tibble: 4 × 6
##   symbol name       variants.type   variants.name    traitType traits.name      
##   &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;           &lt;chr&gt;            &lt;chr&gt;     &lt;chr&gt;            
## 1 PIK3R2 phosphoin… single nucleot… NM_005027.4(PIK… Disease   Megalencephaly-p…
## 2 PIK3R2 phosphoin… single nucleot… NM_005027.4(PIK… Disease   not provided     
## 3 PIK3R2 phosphoin… single nucleot… NM_005027.4(PIK… Disease   not provided     
## 4 UGT1A8 UDP glucu… Microsatellite  UGT1A1*28        Disease   Gilbert&#39;s syndro…</code></pre>
</div>
<div id="merging-with-collections" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Merging with collections</h2>
<div id="collections-and-collection-members" class="section level3" number="2.6.1">
<h3><span class="header-section-number">2.6.1</span> Collections and collection members</h3>
<p>Some databases refer to the same concepts and could be merged accordingly. However they often use different vocabularies.</p>
<p>For example, CHEMBL refers to biological entities (BE) in the <em>CHEMBL_component_sequence</em> table using mainly Uniprot peptide identifiers from different species.</p>
<pre class="r"><code>file_chembl$CHEMBL_component_sequence %&gt;% head()</code></pre>
<pre><code>## # A tibble: 6 × 5
##   component_id accession organism              db_source db_version
##          &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;                 &lt;chr&gt;     &lt;chr&gt;     
## 1          259 P15260    Homo sapiens          Uniprot   2019_09   
## 2          327 Q99062    Homo sapiens          Uniprot   2019_09   
## 3          752 P35563    Rattus norvegicus     Uniprot   2019_09   
## 4          917 P07339    Homo sapiens          Uniprot   2019_09   
## 5         1807 Q54A96    Plasmodium falciparum Uniprot   2019_09   
## 6         2180 P67774    Bos taurus            Uniprot   2019_09</code></pre>
<p>Whereas ClinVar refers to BE in the <em>ClinVar_entrezNames</em> table using human Entrez gene identifiers.</p>
<pre class="r"><code>file_clinvar$ClinVar_entrezNames %&gt;% head()</code></pre>
<pre><code>## # A tibble: 6 × 3
##   entrez name                                             symbol
##    &lt;int&gt; &lt;chr&gt;                                            &lt;chr&gt; 
## 1   1509 cathepsin D                                      CTSD  
## 2   1903 sphingosine-1-phosphate receptor 3               S1PR3 
## 3   3300 DnaJ heat shock protein family (Hsp40) member B2 DNAJB2
## 4   3423 iduronate 2-sulfatase                            IDS   
## 5   3910 laminin subunit alpha 4                          LAMA4 
## 6   5296 phosphoinositide-3-kinase regulatory subunit 2   PIK3R2</code></pre>
<p>Some tools exist to convert such BE identifiers from one scope to the other (<a href="https://github.com/patzaw/BED">BED</a>, <a href="https://bioconductor.org/packages/release/bioc/html/mygene.html">mygene</a>, <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>). TKCat provides mechanism to document these scopes in order to allow automatic conversions from and to any of them. Those concepts are called <strong>Collections</strong> in TKCat and they should be formally defined before being able to document any of their members. Two collection definitions are provided within the TKCat package and other can be imported with the <code>import_local_collection()</code> function.</p>
<pre class="r"><code>list_local_collections()</code></pre>
<pre><code>## # A tibble: 2 × 2
##   title     description                                  
##   &lt;chr&gt;     &lt;chr&gt;                                        
## 1 BE        Collection of biological entity (BE) concepts
## 2 Condition Collection of condition concepts</code></pre>
<p>The way to describe the scope of a collection member is formally defined by a JSON schema (use <code>get_local_collection()</code> to get the JSON of a collection). Here are the definition of the BE collection members provided by the <em>CHEMBL_component_sequence</em> and the <em>ClinVar_entrezNames</em> tables.</p>
<pre class="r"><code>collection_members(file_chembl, &quot;BE&quot;)</code></pre>
<pre><code>## # A tibble: 4 × 9
##   collection cid           resource   mid table      field  static value type   
##   &lt;chr&gt;      &lt;chr&gt;         &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;      &lt;chr&gt;  &lt;lgl&gt;  &lt;chr&gt; &lt;chr&gt;  
## 1 BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_co… be     TRUE   Pept… &lt;NA&gt;   
## 2 BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_co… ident… FALSE  acce… &lt;NA&gt;   
## 3 BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_co… source FALSE  db_s… &lt;NA&gt;   
## 4 BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_co… organ… FALSE  orga… Scient…</code></pre>
<pre class="r"><code>collection_members(file_clinvar, &quot;BE&quot;)</code></pre>
<pre><code>## # A tibble: 4 × 9
##   collection cid            resource   mid table    field  static value   type  
##   &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;  &lt;lgl&gt;  &lt;chr&gt;   &lt;chr&gt; 
## 1 BE         ClinVar_BE_1.0 ClinVar      1 ClinVar… be     TRUE   Gene    &lt;NA&gt;  
## 2 BE         ClinVar_BE_1.0 ClinVar      1 ClinVar… ident… FALSE  entrez  &lt;NA&gt;  
## 3 BE         ClinVar_BE_1.0 ClinVar      1 ClinVar… organ… TRUE   Homo s… Scien…
## 4 BE         ClinVar_BE_1.0 ClinVar      1 ClinVar… source TRUE   Entrez… &lt;NA&gt;</code></pre>
<p>The <em>Collection</em> column indicates the collection to which the table refers. The <em>cid</em> column indicates the version of the collection definition which should correspond to the <code>$id</code> of JSON schema. The <em>resource</em> column indicated the name of the resource and the <em>mid</em> column an identifier which is unique for each member of a collection in each resource. The <em>field</em> column indicated each part of the scope of collection. In the case of BE, 4 fields should be documented:</p>
<ul>
<li>be: the type of BE (e.g. Gene or Peptide)</li>
<li>source: the source of the identifier (e.g. EntrezGene or Peptide)</li>
<li>organism: the organism to which the identifier refers (e.g Homo sapiens)</li>
<li>identifier: the identifier itself.</li>
</ul>
<p>Each of these fields can be <em>static</em> or not. <code>TRUE</code> means that the value of this field is the same for all the records and is provided in the <em>value</em> column. Whereas <code>FALSE</code> means that the value can be different for each record and is provided in the column the name of which is given in the <em>value</em> column. The <em>type</em> column is only used for the organism field in the case of the BE collection and can take 2 values: “Scientific name” or “NCBI taxon identifier.” The definition of the pre-build BE collection members follows the terminology used in the <a href="https://github.com/patzaw/BED">BED</a> package <span class="citation">(<a href="#ref-godard_bed:_2018" role="doc-biblioref">Godard and Eyll 2018</a>)</span>. But it can be adapted according to the solution chosen for converting BE identifiers from one scope to another.</p>
<p>Setting up the definition of such scope is done using the <code>collection_members&lt;-()</code> function as shown in the <a href="#hpo">Reading HPO example</a> above.</p>
</div>
<div id="shared-collections-and-merging" class="section level3" number="2.6.2">
<h3><span class="header-section-number">2.6.2</span> Shared collections and merging</h3>
<p>The aim of collections is to identify potential bridges between MDBs. The <code>get_shared_collection()</code> function is used to list all the collections shared by two MDBs.</p>
<pre class="r"><code>get_shared_collections(filtered_clinvar, file_chembl)</code></pre>
<pre><code>## # A tibble: 3 × 5
##   collection mid.x table.x     mid.y table.y                  
##   &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;                    
## 1 Condition      2 traits          1 CHEMBL_drug_indication   
## 2 Condition      1 traitCref       1 CHEMBL_drug_indication   
## 3 BE             1 entrezNames     1 CHEMBL_component_sequence</code></pre>
<p>In this example, there are 3 different ways to merge the two MDBs <em>filtered_clinvar</em> and <em>file_chembl</em>:</p>
<ul>
<li>Based on conditions provided respectively in the <em>traits</em> and in the <em>CHEMBL_drug_indication</em> tables</li>
<li>Based on conditions provided respectively in the <em>traitsCref</em> and in the <em>CHEMBL_drug_indication</em> tables</li>
<li>Based on BE provided respectively in the <em>entrezNames</em> and in the <em>CHEMBL_component_sequence</em> tables</li>
</ul>
<p>The code below shows how to merge these two resources based on BE information. To achieve this task it relies on a function provided with TKCat along with BE collection definition (to get the function: <code>get_collection_mapper("BE")</code>). This function uses the BED package and you need this package to be installed with a connection to BED database in order to run the code below.</p>
<pre class="r"><code>sel_coll &lt;- get_shared_collections(file_clinvar, file_chembl) %&gt;% 
   filter(collection==&quot;BE&quot;)
filtered_cv_chembl &lt;- merge(
   x=file_clinvar,
   y=file_chembl,
   by=sel_coll
)</code></pre>
<p>The returned object is <strong>metaMDB</strong> gathering the original MDBs and a relational table between members of the same collection as defined by the <code>by</code> parameter.</p>
<p>Additional information about collection can be found <a href="TKCat-Collections.html">here</a>.</p>
</div>
<div id="merging-without-collection" class="section level3" number="2.6.3">
<h3><span class="header-section-number">2.6.3</span> Merging without collection</h3>
<p>If the <em>Collection</em> column of the <code>by</code> parameter is <code>NA</code>, then the relational table is built by merging identical columns in table.x and table.y (No conversion occurs). For example, <em>file_hpo</em> and <em>file_clinvar</em> MDBs could be merged according to conditions provided in the <em>HPO_diseases</em> and the <em>ClinVar_traitCref</em> tables respectively.</p>
<pre class="r"><code>get_shared_collections(file_hpo, file_clinvar)</code></pre>
<pre><code>## # A tibble: 4 × 5
##   collection mid.x table.x      mid.y table.y          
##   &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;            
## 1 Condition      1 HPO_hp           1 ClinVar_traitCref
## 2 Condition      1 HPO_hp           2 ClinVar_traits   
## 3 Condition      2 HPO_diseases     1 ClinVar_traitCref
## 4 Condition      2 HPO_diseases     2 ClinVar_traits</code></pre>
<p>These conditions could be converted using a function provided with TKCat (<code>get_collection_mapper("Condition")</code>) and which rely on the DODO package. The two tables can also be simply concatenated without applying any conversion (loosing the advantage of such conversion obviously).</p>
<pre class="r"><code>sel_coll &lt;- get_shared_collections(file_hpo, file_clinvar) %&gt;% 
   filter(table.x==&quot;HPO_diseases&quot;, table.y==&quot;ClinVar_traitCref&quot;) %&gt;% 
   mutate(collection=NA)
sel_coll</code></pre>
<pre><code>## # A tibble: 1 × 5
##   collection mid.x table.x      mid.y table.y          
##   &lt;lgl&gt;      &lt;int&gt; &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;            
## 1 NA             2 HPO_diseases     1 ClinVar_traitCref</code></pre>
<pre class="r"><code>hpo_clinvar &lt;- merge(file_hpo, file_clinvar, by=sel_coll)
plot(data_model(hpo_clinvar))</code></pre>
<div id="htmlwidget-8b6dac1a3ca4bd6569e0" style="width:672px;height:480px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-8b6dac1a3ca4bd6569e0">{"x":{"nodes":{"tableName":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames","HPO_diseases_ClinVar_traitCref"],"label":["<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_altId<\/b>\n    - id {character - idx.1}\n    - alt {character}","<b>HPO_sourceFiles<\/b>\n    - <b>*url<\/b> {character - uidx.1}\n    - current {Date}","<b>HPO_diseases<\/b>\n    - <b>db<\/b> {character - uidx.1}\n    - <b>id<\/b> {character - uidx.1}\n    - label {character}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_diseaseSynonyms<\/b>\n    - db {character - idx.1}\n    - id {character - idx.1}\n    - synonym {character}\n    - preferred {logical}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}","<b>HPO_descendants<\/b>\n    - id {character - idx.1}\n    - descendant {character - idx.2}","<b>HPO_synonyms<\/b>\n    - id {character}\n    - synonym {character}\n    - type {character}","<b>ClinVar_ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>ClinVar_rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVar_ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>ClinVar_rcvaInhMode<\/b>\n    - rcvaId {integer}\n    - inhMode {character}","<b>ClinVar_rcvaObservedIn<\/b>\n    - rcvaId {integer}\n    - origin {character}\n    - taxonomyId {character}\n    - species {character}\n    - affectedStatus {character}\n    - (numberTested {integer})","<b>ClinVar_rcvaTraits<\/b>\n    - rcvaId {integer}\n    - t.id {integer - idx.1}\n    - traitType {character}","<b>ClinVar_clinSigOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_revStatOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_variants<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - type {character}\n    - (name {character})","<b>ClinVar_cvaObservedIn<\/b>\n    - cvaId {integer}\n    - origin {character}\n    - species {character}\n    - affectedStatus {character}","<b>ClinVar_cvaSubmitters<\/b>\n    - cvaId {integer}\n    - submitter {character}\n    - primary {logical}","<b>ClinVar_traits<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - name {character}","<b>ClinVar_varEntrez<\/b>\n    - varId {integer - idx.1}\n    - entrez {integer}\n    - type {character}","<b>ClinVar_varAttributes<\/b>\n    - varId {integer - idx.1}\n    - Type {character}\n    - (integerValue {integer})\n    - (Change {character})\n    - (value {character})","<b>ClinVar_varCytoLoc<\/b>\n    - varId {integer - idx.1}\n    - location {character}","<b>ClinVar_varNames<\/b>\n    - varId {integer - idx.1}\n    - name {character}\n    - type {character}","<b>ClinVar_varSeqLoc<\/b>\n    - varId {integer - idx.1}\n    - Accession {character}\n    - (alternateAllele {character})\n    - Assembly {character}\n    - AssemblyAccessionVersion {character}\n    - AssemblyStatus {character}\n    - Chr {character}\n    - display_start {integer}\n    - (display_stop {integer})\n    - (innerStart {integer})\n    - (innerStop {integer})\n    - (outerStart {integer})\n    - (outerStop {integer})\n    - (referenceAllele {character})\n    - (start {integer})\n    - (stop {integer})\n    - (Strand {character})\n    - (variantLength {integer})","<b>ClinVar_varXRef<\/b>\n    - varId {integer - idx.1}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitCref<\/b>\n    - t.id {integer}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitNames<\/b>\n    - t.id {integer}\n    - name {character}\n    - type {character}","<b>ClinVar_entrezNames<\/b>\n    - <b>*entrez<\/b> {integer - uidx.1}\n    - name {character}\n    - symbol {character}","<b>HPO_diseases_ClinVar_traitCref<\/b>\n    - db {character}\n    - id {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_altId<\/strong> (Alternative HP identifiers)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>alt<\/strong>: Alternative identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_sourceFiles<\/strong> (Files from which data were extracted)<\/p> <ul> <li><strong>url<\/strong>: Source file URL<\/li> <li><strong>current<\/strong>: The date of the current source file<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong> (Diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>label<\/strong>: Disease lable (preferred synonym)<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseSynonyms<\/strong> (Disease synonyms)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>synonym<\/strong>: Disease synonym<\/li> <li><strong>preferred<\/strong>: Is synonym preferred<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_descendants<\/strong> (Indirect descendants of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>descendant<\/strong>: Descendant (child or child of child...) HP ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_synonyms<\/strong> (Synonyms of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>synonym<\/strong>: HP synonym<\/li> <li><strong>type<\/strong>: Synonym type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaInhMode<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>inhMode<\/strong>: Mode of inheritance<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaObservedIn<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>taxonomyId<\/strong>: Taxonomy ID<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <li><strong>numberTested<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaTraits<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>traitType<\/strong>: Trait type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_clinSigOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of clinical significance: the lowest the best clinical outcome.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_revStatOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of review status: the lowest the least significant.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_variants<\/strong><\/p> <ul> <li><strong>id<\/strong>: Variant ID<\/li> <li><strong>type<\/strong>: Variant type<\/li> <li><strong>name<\/strong>: Variant preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaObservedIn<\/strong><\/p> <ul> <li><strong>cvaId<\/strong><\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaSubmitters<\/strong><\/p> <ul> <li><strong>cvaId<\/strong>: ClinVar Assertion ID<\/li> <li><strong>submitter<\/strong>: Submitter name<\/li> <li><strong>primary<\/strong>: The submitter is of pirmary type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traits<\/strong><\/p> <ul> <li><strong>id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varEntrez<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>type<\/strong>: Type of the association between the variant and the gene<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varAttributes<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Type<\/strong><\/li> <li><strong>integerValue<\/strong><\/li> <li><strong>Change<\/strong><\/li> <li><strong>value<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varCytoLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>location<\/strong>: Cytogenic location<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varNames<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>name<\/strong>: Variant name<\/li> <li><strong>type<\/strong>: Variant name type: Preferred or Alternate<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varSeqLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Accession<\/strong>: Sequence accession ID<\/li> <li><strong>alternateAllele<\/strong>: Alternate allele<\/li> <li><strong>Assembly<\/strong>: Assembly<\/li> <li><strong>AssemblyAccessionVersion<\/strong>: Version of assembly accession<\/li> <li><strong>AssemblyStatus<\/strong>: Status of the assembly: current or previous<\/li> <li><strong>Chr<\/strong>: Chromosome name<\/li> <li><strong>display_start<\/strong><\/li> <li><strong>display_stop<\/strong><\/li> <li><strong>innerStart<\/strong><\/li> <li><strong>innerStop<\/strong><\/li> <li><strong>outerStart<\/strong><\/li> <li><strong>outerStop<\/strong><\/li> <li><strong>referenceAllele<\/strong>: Reference allele<\/li> <li><strong>start<\/strong><\/li> <li><strong>stop<\/strong><\/li> <li><strong>Strand<\/strong><\/li> <li><strong>variantLength<\/strong>: Length of the variant<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varXRef<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitCref<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitNames<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait name<\/li> <li><strong>type<\/strong>: \"Alternate\" or \"Preferred\"<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_entrezNames<\/strong><\/p> <ul> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>name<\/strong>: Gene name<\/li> <li><strong>symbol<\/strong>: Gene symbol<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases_ClinVar_traitCref<\/strong><\/p> <ul> <li><strong>db<\/strong><\/li> <li><strong>id<\/strong><\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left"],"x":[177.996663506481,-117.94594000298,1131.98659629531,823.931339993625,498.181508589049,1145.21224499016,-98.6962705837853,179.044116586047,509.161391632982,98.1757516471211,-333.522369303817,-135.126197269141,392.262735279841,127.401361302693,619.47440741341,-255.190939961579,345.418092316093,-735.667320764574,-484.703738500545,7.82514150136861,988.663473715024,-1088.30168799,-916.89825718393,-407.462085584701,-632.851648986486,-1088.55467150013,-785.877515327878,1122.75576982681,1240,-1240,1013.08595491324],"y":[988.32201490421,829.60690397157,881.898006543361,449.037239091526,685.440154616876,468.194056810558,1111.68967903108,1240,1077.84969557759,-599.222187484239,-239.157708384773,-975.180926056477,-423.990990073063,-263.449325688654,-632.176735140329,-660.41981699549,-905.621069935087,138.163630652958,-1056.30189954394,-1240,-476.140900209595,-64.8491208908148,524.848641470959,256.943206369842,497.068256869352,289.811751021027,-196.218409498351,-156.976806987799,-613.920875848696,-323.388116479613,144.02804514682],"color.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0","yellow"],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0","yellow"],"id":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames","HPO_diseases_ClinVar_traitCref"]},"edges":{"id":["HPO_altId->HPO_hp: id->id","HPO_diseaseHP->HPO_hp: hp->id","HPO_diseaseHP->HPO_diseases: db->db && id->id","HPO_diseaseSynonyms->HPO_diseases: db->db && id->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id","HPO_descendants->HPO_hp: id->id","HPO_descendants->HPO_hp: descendant->id","HPO_synonyms->HPO_hp: id->id","ClinVar_ReferenceClinVarAssertion->ClinVar_clinSigOrder: clinicalSignificance->label","ClinVar_ReferenceClinVarAssertion->ClinVar_revStatOrder: reviewStatus->label","ClinVar_rcvaVariant->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaVariant->ClinVar_variants: varId->id","ClinVar_ClinVarAssertions->ClinVar_ReferenceClinVarAssertion: cvs->cvs","ClinVar_rcvaInhMode->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaObservedIn->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaTraits->ClinVar_traits: t.id->id","ClinVar_rcvaTraits->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_cvaObservedIn->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_cvaSubmitters->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_varEntrez->ClinVar_variants: varId->id","ClinVar_varEntrez->ClinVar_entrezNames: entrez->entrez","ClinVar_varAttributes->ClinVar_variants: varId->id","ClinVar_varCytoLoc->ClinVar_variants: varId->id","ClinVar_varNames->ClinVar_variants: varId->id","ClinVar_varSeqLoc->ClinVar_variants: varId->id","ClinVar_varXRef->ClinVar_variants: varId->id","ClinVar_traitCref->ClinVar_traits: t.id->id","ClinVar_traitNames->ClinVar_traits: t.id->id","HPO_diseases_ClinVar_traitCref->HPO_diseases: db->db && id->id","HPO_diseases_ClinVar_traitCref->ClinVar_traitCref: db->db && id->id"],"to":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseases","HPO_hp","HPO_hp","HPO_hp","HPO_hp","HPO_hp","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_variants","ClinVar_entrezNames","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traits","ClinVar_traits","HPO_diseases","ClinVar_traitCref"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_altId<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseSynonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">descendant<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_synonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_clinSigOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">clinicalSignificance<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_revStatOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">reviewStatus<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaInhMode<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaSubmitters<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_entrezNames<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varAttributes<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varCytoLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varSeqLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varXRef<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases_ClinVar_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(0..n)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases_ClinVar_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitCref<br>(0..n)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["id","hp",["db","id"],["db","id"],"id","parent","id","descendant","id","clinicalSignificance","reviewStatus","rcvaId","varId","cvs","rcvaId","rcvaId","t.id","rcvaId","cvaId","cvaId","varId","entrez","varId","varId","varId","varId","varId","t.id","t.id",["db","id"],["db","id"]],"tf":["id","id",["db","id"],["db","id"],"id","id","id","id","id","label","label","id","id","cvs","id","id","id","id","id","id","id","entrez","id","id","id","id","id","id","id",["db","id"],["db","id"]],"from":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_descendants","HPO_descendants","HPO_synonyms","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_varEntrez","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","HPO_diseases_ClinVar_traitCref","HPO_diseases_ClinVar_traitCref"],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0,0.2,-0.2,-0.2,0.2,0,0,-0,-0,-0,-0,-0,-0,-0,-0,0,0,-0,0,-0,-0,0,0,0,-0,-0,0,0],"selfReferenceSize":[30,30,30,30,30,50,30,50,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30],"uef":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseases","HPO_hp","HPO_hp","HPO_descendants","HPO_descendants","HPO_hp","ClinVar_clinSigOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_varEntrez","ClinVar_entrezNames","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traitCref","ClinVar_traitNames","HPO_diseases","ClinVar_traitCref"],"uet":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_hp","HPO_hp","HPO_synonyms","ClinVar_ReferenceClinVarAssertion","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_variants","ClinVar_varEntrez","ClinVar_variants","ClinVar_variants","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traits","ClinVar_traits","HPO_diseases_ClinVar_traitCref","HPO_diseases_ClinVar_traitCref"],"ue":["HPO_altId<->HPO_hp","HPO_diseaseHP<->HPO_hp","HPO_diseaseHP<->HPO_diseases","HPO_diseases<->HPO_diseaseSynonyms","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents","HPO_descendants<->HPO_hp","HPO_descendants<->HPO_hp","HPO_hp<->HPO_synonyms","ClinVar_clinSigOrder<->ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion<->ClinVar_revStatOrder","ClinVar_rcvaVariant<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant<->ClinVar_variants","ClinVar_ClinVarAssertions<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaInhMode<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaObservedIn<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaTraits<->ClinVar_traits","ClinVar_rcvaTraits<->ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions<->ClinVar_cvaObservedIn","ClinVar_ClinVarAssertions<->ClinVar_cvaSubmitters","ClinVar_varEntrez<->ClinVar_variants","ClinVar_entrezNames<->ClinVar_varEntrez","ClinVar_varAttributes<->ClinVar_variants","ClinVar_varCytoLoc<->ClinVar_variants","ClinVar_variants<->ClinVar_varNames","ClinVar_variants<->ClinVar_varSeqLoc","ClinVar_variants<->ClinVar_varXRef","ClinVar_traitCref<->ClinVar_traits","ClinVar_traitNames<->ClinVar_traits","HPO_diseases<->HPO_diseases_ClinVar_traitCref","ClinVar_traitCref<->HPO_diseases_ClinVar_traitCref"],"color.color":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>hpo_clinvar$HPO_diseases_ClinVar_traitCref %&gt;% head()</code></pre>
<pre><code>## # A tibble: 6 × 2
##   db       id    
##   &lt;chr&gt;    &lt;chr&gt; 
## 1 DECIPHER 15    
## 2 DECIPHER 45    
## 3 DECIPHER 65    
## 4 OMIM     100050
## 5 OMIM     100650
## 6 OMIM     101800</code></pre>
</div>
</div>
</div>
<div id="mdb-catalogs-as-tkcat-objects" class="section level1" number="3">
<h1><span class="header-section-number">3</span> MDB catalogs as TKCat objects</h1>
<div id="local-tkcat" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Local TKCat</h2>
<p>MDB can be gathered in a <em>TKCat</em> (Tailored Knowledge Catalog) object.</p>
<pre class="r"><code>k &lt;- TKCat(file_hpo, file_clinvar)</code></pre>
<p>Gathering MDBs in such a catalog facilitate their exploration and their preparation for potential integration. Several functions are available to achieve this goal.</p>
<pre class="r"><code>list_MDBs(k)                     # list all the MDBs in a TKCat object</code></pre>
<pre><code>## # A tibble: 2 × 7
##   name    title             description     url    version maintainer  timestamp
##   &lt;chr&gt;   &lt;chr&gt;             &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;       &lt;dttm&gt;   
## 1 HPO     Data extracted f… This is a very… http:… &lt;NA&gt;    &lt;NA&gt;        NA       
## 2 ClinVar Data extracted f… ClinVar is a f… https… 0.9     Patrice Go… NA</code></pre>
<pre class="r"><code>get_MDB(k, &quot;HPO&quot;)                # get a specific MDBs from the catalog</code></pre>
<pre><code>## fileMDB HPO: Data extracted from the HPO database
##    - 9 tables with 25 fields
## 
## Collection members: 
##    - 2 Condition members
## 
## This is a very small subset of the HPO! Visit the reference URL for more information.
## (http://human-phenotype-ontology.github.io/)
## 
## </code></pre>
<pre class="r"><code>search_MDB_tables(k, &quot;disease&quot;)  # Search table about &quot;disease&quot;</code></pre>
<pre><code>## # A tibble: 3 × 3
##   resource name                comment                 
##   &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                   
## 1 HPO      HPO_diseases        Diseases                
## 2 HPO      HPO_diseaseHP       HP presented by diseases
## 3 HPO      HPO_diseaseSynonyms Disease synonyms</code></pre>
<pre class="r"><code>search_MDB_fields(k, &quot;disease&quot;)  # Search a field about &quot;disease&quot;</code></pre>
<pre><code>## # A tibble: 8 × 7
##   resource table               name    type      nullable unique comment        
##   &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;   &lt;chr&gt;     &lt;lgl&gt;    &lt;lgl&gt;  &lt;chr&gt;          
## 1 HPO      HPO_diseases        db      character FALSE    FALSE  Disease databa…
## 2 HPO      HPO_diseases        id      character FALSE    FALSE  Disease ID     
## 3 HPO      HPO_diseases        label   character FALSE    FALSE  Disease lable …
## 4 HPO      HPO_diseaseHP       db      character FALSE    FALSE  Disease databa…
## 5 HPO      HPO_diseaseHP       id      character FALSE    FALSE  Disease ID     
## 6 HPO      HPO_diseaseSynonyms db      character FALSE    FALSE  Disease databa…
## 7 HPO      HPO_diseaseSynonyms id      character FALSE    FALSE  Disease ID     
## 8 HPO      HPO_diseaseSynonyms synonym character FALSE    FALSE  Disease synonym</code></pre>
<pre class="r"><code>collection_members(k)            # Get collection members of the different MDBs</code></pre>
<pre><code>## # A tibble: 5 × 3
##   resource collection table              
##   &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;              
## 1 HPO      Condition  HPO_hp             
## 2 HPO      Condition  HPO_diseases       
## 3 ClinVar  BE         ClinVar_entrezNames
## 4 ClinVar  Condition  ClinVar_traitCref  
## 5 ClinVar  Condition  ClinVar_traits</code></pre>
<pre class="r"><code>c(k, TKCat(file_chembl))         # Merge 2 TKCat objects</code></pre>
<pre><code>## TKCat gathering 3 MDB objects</code></pre>
</div>
<div id="chtkcat" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> chTKCat</h2>
<p>A <em>chTKCat</em> object is a catalog of MDB as a <em>TKCat</em> object described above but relying on a <a href="https://clickhouse.com/">ClickHouse</a> database. Therefore it requires the installation and the initialization of such a database. Two additional vignettes describes:</p>
<ul>
<li><a href="chTKCat-Operations-manual.html">How to install and administrate a chTKCat database</a></li>
<li><a href="chTKCat-User-guide.html">How to connect to an existing chTKCat database and use it</a></li>
</ul>
</div>
<div id="a-shiny-app-for-exploring-mdbs" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> A shiny app for exploring MDBs</h2>
<p>The function <code>explore_MDBs(k)</code> launches a shiny interface to explore MDBs in a <em>TKCat</em> or a <em>chTKCat</em> object. This exploration interface can be easily deployed using an <em>app.R</em> file with content similar to the one below.</p>
<pre class="r"><code>library(TKCat)
explore_MDBs(k, download=TRUE)</code></pre>
<p><img src="libs/figs/explore_MDBs-local.png" /></p>
<p>In this interface the users can explore the resources available in the catalog. They can browse the data model of each of them with some sample data. They can also search for information provided in resources, tables or fields. Finally, if the parameter <code>download</code> is set to <code>TRUE</code>, the users will also be able to download the data: either each table individually or an archive of the whole MDB.</p>
</div>
</div>
<div id="acknowledgments" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Acknowledgments</h1>
<p>This work was entirely supported by <a href="https://www.ucb.com/">UCB Pharma</a> (Early Solutions department).</p>
</div>
<div id="references" class="section level1" number="5">
<h1><span class="header-section-number">5</span> References</h1>
<!-- Bibliography -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-godard_bed:_2018" class="csl-entry">
Godard, Patrice, and Jonathan van Eyll. 2018. <span>“<span>BED</span>: A <span>Biological</span> <span>Entity</span> <span>Dictionary</span> Based on a Graph Data Model.”</span> <em>F1000Research</em> 7: 195. <a href="https://doi.org/10.12688/f1000research.13925.3">https://doi.org/10.12688/f1000research.13925.3</a>.
</div>
<div id="ref-kohler_expansion_2019" class="csl-entry">
Köhler, Sebastian, Leigh Carmody, Nicole Vasilevsky, Julius O B Jacobsen, Daniel Danis, Jean-Philippe Gourdine, Michael Gargano, et al. 2019. <span>“Expansion of the <span>Human</span> <span>Phenotype</span> <span>Ontology</span> (<span>HPO</span>) Knowledge Base and Resources.”</span> <em>Nucleic Acids Research</em> 47 (D1): D1018–27. <a href="https://doi.org/10.1093/nar/gky1105">https://doi.org/10.1093/nar/gky1105</a>.
</div>
<div id="ref-landrum_clinvar_2018" class="csl-entry">
Landrum, Melissa J., Jennifer M. Lee, Mark Benson, Garth R. Brown, Chen Chao, Shanmuga Chitipiralla, Baoshan Gu, et al. 2018. <span>“<span>ClinVar</span>: Improving Access to Variant Interpretations and Supporting Evidence.”</span> <em>Nucleic Acids Research</em> 46 (D1): D1062–67. <a href="https://doi.org/10.1093/nar/gkx1153">https://doi.org/10.1093/nar/gkx1153</a>.
</div>
<div id="ref-mendez_chembl_2019" class="csl-entry">
Mendez, David, Anna Gaulton, A Patrícia Bento, Jon Chambers, Marleen De Veij, Eloy Félix, María Paula Magariños, et al. 2019. <span>“<span>ChEMBL</span>: Towards Direct Deposition of Bioassay Data.”</span> <em>Nucleic Acids Research</em> 47 (D1): D930–40. <a href="https://doi.org/10.1093/nar/gky1075">https://doi.org/10.1093/nar/gky1075</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
