<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="TKCat">
<title>Introduction to TKCat • TKCat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to TKCat">
<meta property="og:description" content="TKCat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-7RG3ZVT8WZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-7RG3ZVT8WZ');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TKCat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/TKCat.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/TKCat-KMR-POK.html">Defining and using Requirements for Knowledge Management (KMR) in TKCat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/patzaw/TKCat/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="TKCat_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="TKCat_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="TKCat_files/vis-9.1.0/vis-network.min.css" rel="stylesheet">
<script src="TKCat_files/vis-9.1.0/vis-network.min.js"></script><script src="TKCat_files/visNetwork-binding-2.1.2/visNetwork.js"></script><link href="TKCat_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="TKCat_files/datatables-binding-0.33/datatables.js"></script><link href="TKCat_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="TKCat_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="TKCat_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="TKCat_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="TKCat_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to TKCat</h1>
                        <h4 data-toc-skip class="author">Patrice
Godard</h4>
            
            <h4 data-toc-skip class="date">May 22, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/patzaw/TKCat/blob/HEAD/vignettes/TKCat.Rmd" class="external-link"><code>vignettes/TKCat.Rmd</code></a></small>
      <div class="d-none name"><code>TKCat.Rmd</code></div>
    </div>

    
    
<div style="width:200px;">
<p><img src="TKCat.png" style="width:100.0%"></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Research organizations generate, manage, and use more and more
knowledge resources which can be highly heterogenous in their origin,
their scope, and their structure. Making this knowledge compliant to
F.A.I.R. (Findable, Accessible, Interoperable, Reusable) principles is
critical for facilitating the generation of new insights leveraging it.
The aim of the TKCat (Tailored Knowledge Catalog) R package is to
facilitate the management of such resources that are frequently used
alone or in combination in research environments.</p>
<p>In TKCat, knowledge resources are manipulated as modeled database
(MDB) objects. These objects provide access to the data tables along
with a general description of the resource and a detail data model
generated with <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a>
documenting the tables, their fields and their relationships. These MDB
are then gathered in catalogs that can be easily explored an shared.
TKCat provides tools to easily subset, filter and combine MDBs and
create new catalogs suited for specific needs.</p>
<p>Currently, there are 3 different implementations of MDBs which are
supported by TKCat: in R memory (memoMDB), in files (fileMDB) and in <a href="https://clickhouse.com/" class="external-link">ClickHouse</a> (chMDB).</p>
<p>This is document is divided in four main sections:</p>
<ul>
<li><p>The first one describes how to build an MDB object, starting with
a minimal example</p></li>
<li><p>The second section shows how to interact with MDB objects to
extract and combine information of interest</p></li>
<li><p>The third section focuses on the use of the <a href="https://clickhouse.com/" class="external-link">ClickHouse</a> implementation of MDB
(chMDB)</p></li>
<li><p>The fourth section corresponds to appendices providing technical
information regarding <a href="https://clickhouse.com/" class="external-link">ClickHouse</a>
related admin tasks and the implementation of <em>collections</em> which
are used to identify and leverage potential relationships between
different MDBs.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="create-an-mdb-a-minimal-example">Create an MDB: a minimal example<a class="anchor" aria-label="anchor" href="#create-an-mdb-a-minimal-example"></a>
</h2>
<p>This section shows how to create an MDB object starting from a set of
tables in three steps:</p>
<ul>
<li>Create a data model</li>
<li>Create and validate a modeled database (MDB) by binding the data
model to the dataset</li>
<li>Document concept collections that can be used to make bridges across
different MDBs</li>
</ul>
<p>This example focuses on the <a href="#hpo">Human Phenotype Ontology
(HPO)</a>. The HPO aims to provide a standardized vocabulary of
phenotypic abnormalities encountered in human diseases <span class="citation">(<a href="#ref-kohler_expansion_2019">Köhler et al.
2019</a>)</span>.</p>
<div class="section level3">
<h3 id="loading-example-data">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data"></a>
</h3>
<p>A subset of the HPO is provided within the <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a> package. We can
read some of the tables as follow:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="va">hpo_data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/HPO-subset"</span>, package<span class="op">=</span><span class="st">"ReDaMoR"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>HPO_hp</code> table gathers human phenotype identifiers,
names and descriptions:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HPO_hp</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">hpo_data_dir</span>, <span class="st">"HPO_hp.txt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">HPO_hp</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 500 × 4</span></span></span>
<span><span class="co">##    id      name                                              description   level</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 0000002 Abnormality of body height                        Deviation fr…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 0000009 Functional abnormality of the bladder             Dysfunction …     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 0000014 Abnormality of the bladder                        An abnormali…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 0000017 Nocturia                                          Abnormally i…     7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 0000019 Urinary hesitancy                                 Difficulty i…     7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 0000021 Megacystis                                        Dilatation o…     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 0000022 Abnormality of male internal genitalia            An abnormali…     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 0000024 Prostatitis                                       The presence…     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 0000025 Functional abnormality of male internal genitalia <span style="color: #BB0000;">NA</span>                6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 0000030 Testicular gonadoblastoma                         The presence…     9</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 490 more rows</span></span></span></code></pre>
<p>The <code>HPO_diseases</code> table gathers disease identifiers and
labels from different disease database.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HPO_diseases</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">hpo_data_dir</span>, <span class="st">"HPO_diseases.txt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">HPO_diseases</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,903 × 3</span></span></span>
<span><span class="co">##    db           id label                                                        </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> DECIPHER     15 NF1-microdeletion syndrome                                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> DECIPHER     45 Xq28 (MECP2) duplication                                     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> DECIPHER     65 ATR-16 syndrome                                              </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> OMIM     <span style="text-decoration: underline;">100</span>050 AARSKOG SYNDROME, AUTOSOMAL DOMINANT                         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> OMIM     <span style="text-decoration: underline;">100</span>650 ALDEHYDE DEHYDROGENASE 2 FAMILY                              </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> OMIM     <span style="text-decoration: underline;">101</span>800 ACRODYSOSTOSIS 1, WITH OR WITHOUT HORMONE RESISTANCE; ACRDYS1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> OMIM     <span style="text-decoration: underline;">102</span>500 HAJDU-CHENEY SYNDROME; HJCYS                                 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> OMIM     <span style="text-decoration: underline;">102</span>510 ACROPECTOROVERTEBRAL DYSPLASIA, F-FORM OF                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> OMIM     <span style="text-decoration: underline;">102</span>700 SEVERE COMBINED IMMUNODEFICIENCY, AUTOSOMAL RECESSIVE, T CEL…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> OMIM     <span style="text-decoration: underline;">102</span>800 ADENOSINE TRIPHOSPHATASE DEFICIENCY, ANEMIA DUE TO           </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,893 more rows</span></span></span></code></pre>
<p>The <code>HPO_diseaseHP</code> table indicates which phenotype is
triggered by each disease.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HPO_diseaseHP</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">hpo_data_dir</span>, <span class="st">"HPO_diseaseHP.txt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">HPO_diseaseHP</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2,594 × 3</span></span></span>
<span><span class="co">##    db           id hp     </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ORPHA    <span style="text-decoration: underline;">140</span>976 0000002</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ORPHA       432 0000002</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> DECIPHER     45 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> OMIM     <span style="text-decoration: underline;">300</span>076 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ORPHA    <span style="text-decoration: underline;">100</span>996 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ORPHA    <span style="text-decoration: underline;">100</span>997 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ORPHA      <span style="text-decoration: underline;">2</span>571 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ORPHA    <span style="text-decoration: underline;">391</span>487 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ORPHA    <span style="text-decoration: underline;">488</span>594 0000009</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ORPHA     <span style="text-decoration: underline;">71</span>211 0000009</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2,584 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="creating-a-data-model-with-redamor">Creating a data model with ReDaMoR<a class="anchor" aria-label="anchor" href="#creating-a-data-model-with-redamor"></a>
</h3>
<p>The <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a> package
can be used for drafting a data model from a set of table:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_dm</span> <span class="op">&lt;-</span> <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/df_to_model.html" class="external-link">df_to_model</a></span><span class="op">(</span><span class="va">HPO_hp</span>, <span class="va">HPO_diseases</span>, <span class="va">HPO_diseaseHP</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">igraph_available</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">mhpo_dm</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/auto_layout.html" class="external-link">auto_layout</a></span><span class="op">(</span>lengthMultiplier<span class="op">=</span><span class="fl">80</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>   <span class="va">mhpo_dm</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:700px;height:200px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"nodes":{"tableName":["HPO_hp","HPO_diseases","HPO_diseaseHP"],"label":["<b>HPO_hp<\/b>\n    - (id {character})\n    - (name {character})\n    - (description {character})\n    - (level {numeric})","<b>HPO_diseases<\/b>\n    - (db {character})\n    - (id {numeric})\n    - (label {character})","<b>HPO_diseaseHP<\/b>\n    - (db {character})\n    - (id {numeric})\n    - (hp {character})"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong><\/p> <ul> <li><strong>id<\/strong><\/li> <li><strong>name<\/strong><\/li> <li><strong>description<\/strong><\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong><\/p> <ul> <li><strong>db<\/strong><\/li> <li><strong>id<\/strong><\/li> <li><strong>label<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong><\/p> <ul> <li><strong>db<\/strong><\/li> <li><strong>id<\/strong><\/li> <li><strong>hp<\/strong><\/li> <\/ul><\/div>"],"shape":["box","box","box"],"font.multi":[true,true,true],"font.align":["left","left","left"],"x":[-22.19239468230881,-80,80],"y":[-80,80,52.914658658765],"color.background":["lightgrey","lightgrey","lightgrey"],"color.border":["black","black","black"],"color.highlight.border":["orange","orange","orange"],"color.highlight.background":["lightgrey","lightgrey","lightgrey"],"id":["HPO_hp","HPO_diseases","HPO_diseaseHP"]},"edges":{"id":[],"from":[],"to":[]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><p>This data model is minimal: only the name of the tables, their fields
and their types are documented. There is no additional constrain
regarding the uniqueness or the completeness of the fields. Also there
is no information regarding the relationships between the different
tables. The <code><a href="https://patzaw.github.io/ReDaMoR/reference/model_relational_data.html" class="external-link">model_relational_data()</a></code> can be used to improve
the documentation of the dataset according to what we know about it.
This function raises a graphical interface for manipulating and
modifying the data model (see <a href="https://patzaw.github.io/ReDaMoR/articles/ReDaMoR.html" class="external-link">ReDaMoR
documentation</a>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_dm</span> <span class="op">&lt;-</span> <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/model_relational_data.html" class="external-link">model_relational_data</a></span><span class="op">(</span><span class="va">mhpo_dm</span><span class="op">)</span></span></code></pre></div>
<p>Below is the model we get after completing it using the function
above.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mhpo_dm</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:700px;height:200px;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"nodes":{"tableName":["HPO_hp","HPO_diseases","HPO_diseaseHP"],"label":["<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_diseases<\/b>\n    - <b>db<\/b> {character - uidx.1}\n    - <b>id<\/b> {character - uidx.1}\n    - label {character}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong> (Diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>label<\/strong>: Disease lable (preferred synonym)<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>"],"shape":["box","box","box"],"font.multi":[true,true,true],"font.align":["left","left","left"],"x":[-167,320,75],"y":[-177,-170,-172],"color.background":["#9BC8FE","#DFFB86","#D9D9D9"],"color.border":["black","black","black"],"color.highlight.border":["orange","orange","orange"],"color.highlight.background":["#9BC8FE","#DFFB86","#D9D9D9"],"id":["HPO_hp","HPO_diseases","HPO_diseaseHP"]},"edges":{"id":["HPO_diseaseHP->HPO_hp: hp->id","HPO_diseaseHP->HPO_diseases: db->db && id->id"],"to":["HPO_hp","HPO_diseases"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["hp",["db","id"]],"tf":["id",["db","id"]],"from":["HPO_diseaseHP","HPO_diseaseHP"],"arrows":["to","to"],"font.align":["bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0],"selfReferenceSize":[30,30],"uef":["HPO_diseaseHP","HPO_diseaseHP"],"uet":["HPO_hp","HPO_diseases"],"ue":["HPO_diseaseHP<->HPO_hp","HPO_diseaseHP<->HPO_diseases"],"color.color":["black","black"],"color.highlight":["orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><p>In this model, we can see that:</p>
<ul>
<li>
<em>id</em> is the <strong>primary key</strong> of the
<em>HPO_hp</em> table, and therefore this field must be
<strong>unique</strong>;</li>
<li>
<em>db</em>/<em>id</em> form the <strong>primary key</strong> of the
<em>HPO_diseases</em> table and must also be <strong>unique</strong>
when taken together;</li>
<li>all the fields excepted <em>description</em> (in the <em>HPO_hp</em>
table) are complete (they cannot be NA);</li>
<li>the <em>HPO_diseaseHP</em> table refers to the <em>HPO_hp</em> table
using its <em>HPO_hp</em> fields and to the <em>HPO_diseases</em> table
using its <em>db</em> and <em>id</em> fields (such details are shown
when putting cursor over the edges).</li>
</ul>
<p>Moreover, some comments are added at the table and at the field level
to give a better understanding of the data (shown when putting the
cursor over the tables).</p>
</div>
<div class="section level3">
<h3 id="binding-the-model-to-the-data-in-an-mdb-object">Binding the model to the data in an MDB object<a class="anchor" aria-label="anchor" href="#binding-the-model-to-the-data-in-an-mdb-object"></a>
</h3>
<p>The data model can be explicitly bound to the data in an MDB (Modeled
DataBase) object as shown below. However, when trying to build the
object with the tables we’ve read and the data model we have edited, we
get the following error message.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/memoMDB.html">memoMDB</a></span><span class="op">(</span></span>
<span>   dataTables<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      HPO_hp<span class="op">=</span><span class="va">HPO_hp</span>, HPO_diseases<span class="op">=</span><span class="va">HPO_diseases</span>, HPO_diseaseHP<span class="op">=</span><span class="va">HPO_diseaseHP</span></span>
<span>   <span class="op">)</span>,</span>
<span>   dataModel<span class="op">=</span><span class="va">mhpo_dm</span>,</span>
<span>   dbInfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"miniHPO"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4" style="background-color:#FFFACD; border-top:5px solid black; border-bottom:5px solid black; margin-top:50px; margin-bottom:50px">
<h4 class="unlisted unnumbered" id="minihpo">miniHPO<a class="anchor" aria-label="anchor" href="#minihpo"></a>
</h4>
<p><span style="background-color:red; color:white; padding:2px;">FAILURE</span></p>
<div class="section level5">
<h5 class="unlisted unnumbered" id="check-configuration">Check configuration<a class="anchor" aria-label="anchor" href="#check-configuration"></a>
</h5>
<ul>
<li>
<strong>Optional checks</strong>: unique, not nullable, foreign
keys</li>
<li>
<strong>Maximum number of records</strong>: Inf</li>
</ul>
</div>
<div class="section level5">
<h5 class="unlisted unnumbered" id="hpo_hp">HPO_hp<a class="anchor" aria-label="anchor" href="#hpo_hp"></a>
</h5>
<p><span style="background-color:red; color:white; padding:2px;">FAILURE</span></p>
<div class="section level6">
<h6 class="unlisted unnumbered" id="field-issues-or-warnings">Field issues or warnings<a class="anchor" aria-label="anchor" href="#field-issues-or-warnings"></a>
</h6>
<ul>
<li>description: <span style="background-color:green; color:black; padding:2px;">SUCCESS</span>
<span style="background-color:#FFBB33; color:white; padding:2px;">Missing
values 117/500 = 23%</span>
</li>
<li>level: <span style="background-color:red; color:white; padding:2px;">FAILURE</span>
<span style="background-color:#FFBB33; color:white; padding:2px;">Unexpected
“numeric”</span>
</li>
</ul>
</div>
</div>
<div class="section level5">
<h5 class="unlisted unnumbered" id="hpo_diseases">HPO_diseases<a class="anchor" aria-label="anchor" href="#hpo_diseases"></a>
</h5>
<p><span style="background-color:red; color:white; padding:2px;">FAILURE</span></p>
<div class="section level6">
<h6 class="unlisted unnumbered" id="field-issues-or-warnings-1">Field issues or warnings<a class="anchor" aria-label="anchor" href="#field-issues-or-warnings-1"></a>
</h6>
<ul>
<li>id: <span style="background-color:red; color:white; padding:2px;">FAILURE</span>
<span style="background-color:#FFBB33; color:white; padding:2px;">Unexpected
“numeric”</span>
</li>
</ul>
</div>
</div>
<div class="section level5">
<h5 class="unlisted unnumbered" id="hpo_diseasehp">HPO_diseaseHP<a class="anchor" aria-label="anchor" href="#hpo_diseasehp"></a>
</h5>
<p><span style="background-color:red; color:white; padding:2px;">FAILURE</span></p>
<div class="section level6">
<h6 class="unlisted unnumbered" id="field-issues-or-warnings-2">Field issues or warnings<a class="anchor" aria-label="anchor" href="#field-issues-or-warnings-2"></a>
</h6>
<ul>
<li>id: <span style="background-color:red; color:white; padding:2px;">FAILURE</span>
<span style="background-color:#FFBB33; color:white; padding:2px;">Unexpected
“numeric”</span>
</li>
</ul>
</div>
</div>
</div>
<p>Indeed, according to the edited model (not the very first one
automatically created by ReDaMoR), the <code>HPO_hp$level</code> field
should contain <em>integer</em> values and the
<code>HPO_diseases$id</code> and <code>HPO_diseaseHP$id</code> fields
should contain <em>character</em> values. The type of the data is among
the data model features that are automatically checked when building an
MDB object (along with uniqueness or NA values for example).</p>
<p>To avoid this error, we can either change the type of the columns of
the data tables:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HPO_hp</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">HPO_hp</span>, level<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">level</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">HPO_diseases</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">HPO_diseases</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">HPO_diseaseHP</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">HPO_diseaseHP</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mhpo_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/memoMDB.html">memoMDB</a></span><span class="op">(</span></span>
<span>   dataTables<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      HPO_hp<span class="op">=</span><span class="va">HPO_hp</span>, HPO_diseases<span class="op">=</span><span class="va">HPO_diseases</span>, HPO_diseaseHP<span class="op">=</span><span class="va">HPO_diseaseHP</span></span>
<span>   <span class="op">)</span>,</span>
<span>   dataModel<span class="op">=</span><span class="va">mhpo_dm</span>,</span>
<span>   dbInfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"miniHPO"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Or we can use the data model to read the data in a fileMDB
object:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f_mhpo_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fileMDB.html">read_fileMDB</a></span><span class="op">(</span></span>
<span>   path<span class="op">=</span><span class="va">hpo_data_dir</span>,</span>
<span>   dbInfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"miniHPO"</span><span class="op">)</span>,</span>
<span>   dataModel<span class="op">=</span><span class="va">mhpo_dm</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold; text-decoration: underline;">miniHPO</span></span></span>
<span><span class="co">## <span style="color: #BBBBBB; background-color: #00BB00;">SUCCESS</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="font-weight: bold;">Check configuration</span></span></span>
<span><span class="co">##    - Optional checks: </span></span>
<span><span class="co">##    - Maximum number of records: 10</span></span></code></pre>
<p>The <code><a href="../reference/read_fileMDB.html">read_fileMDB()</a></code> function identifies the text files to
read in <code>path</code> according to the <code>dataModel</code>. It
uses the types documented in the data model to read the files. By
default, the field delimiter is <code>\t</code>, but another can be
defined by writing a <code>delim</code> slot in the <code>dbInfo</code>
parameter
(e.g. <code>dbInfo=list(name="miniHPO", delim="\t")</code>).</p>
<p>As shown in the message above, by default,
<code><a href="../reference/read_fileMDB.html">read_fileMDB()</a></code> does not perform optional checks
(<em>unique</em> fields, <em>not nullable</em> fields, <em>foreign
keys</em>) and it only checks data on the 10 first records. Also, the
fileMDB data are not loaded in memory until requested by the user. The
object is then smaller than the memoMDB object even if they gather the
same information.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span>, units<span class="op">=</span><span class="st">"Kb"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 691.9 Kb</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">f_mhpo_db</span><span class="op">)</span>, units<span class="op">=</span><span class="st">"Kb"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 23.5 Kb</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compare_MDB.html">compare_MDB</a></span><span class="op">(</span>former<span class="op">=</span><span class="va">mhpo_db</span>, new<span class="op">=</span><span class="va">f_mhpo_db</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span></span>
<span>      rownames<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>      width<span class="op">=</span><span class="st">"75%"</span>,</span>
<span>      options<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dom<span class="op">=</span><span class="st">"t"</span>, pageLength<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-36aa3d2a04d42bbc2145" style="width:75%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"data":[["name","title","description","url","version","maintainer","timestamp","Model","Model display","Table HPO_diseaseHP","Table HPO_diseases","Table HPO_hp","Total","Collections"],["miniHPO",null,null,null,null,null,null,"3 tables","","2,594","1,903","  500","4,997","0"],["miniHPO",null,null,null,null,null,null,"3 tables","","2,594","1,903","  500","4,997","0"],[true,null,null,null,null,null,null,true,true,true,true,true,true,true]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>Information<\/th>\n      <th>Former<\/th>\n      <th>New<\/th>\n      <th>Identical<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","pageLength":14,"columnDefs":[{"name":"Information","targets":0},{"name":"Former","targets":1},{"name":"New","targets":2},{"name":"Identical","targets":3}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[10,14,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="adding-information-about-an-mdb">Adding information about an MDB<a class="anchor" aria-label="anchor" href="#adding-information-about-an-mdb"></a>
</h3>
<p>In the table above we can see that several pieces of information are
expected in an MDB object even if not mandatory (<em>title</em>,
<em>description</em>, <em>url</em>, <em>version</em>,
<em>maintainer</em>, <em>timestamp</em>). They can be provided in the
<code>dbInfo</code> parameter of the MDB creator function
(e.g. <code><a href="../reference/memoMDB.html">memoMDB()</a></code>) or added afterward:</p>
<ul>
<li>
<em>title</em>, <em>description</em> and <em>url</em> are used to
give more details about the scope of the data and their origin.</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">title</span> <span class="op">&lt;-</span> <span class="st">"Very small extract of the human phenotype ontology"</span></span>
<span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">description</span> <span class="op">&lt;-</span> <span class="st">"For demonstrating ReDaMoR and TKCat capabilities, a very few information from the HPO (human phenotype ontology) has been extracted"</span></span>
<span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://hpo.jax.org/app/"</span></span></code></pre></div>
<ul>
<li>
<em>version</em> and <em>maintainer</em> are related to db
information and the data model whereas <em>timestamp</em> should be used
to document the data themselves.</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">version</span> <span class="op">&lt;-</span> <span class="st">"0.1"</span></span>
<span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">maintainer</span> <span class="op">&lt;-</span> <span class="st">"Patrice Godard"</span></span>
<span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span><span class="op">$</span><span class="va">timestamp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>All this information is displayed when printing the object:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span></span></code></pre></div>
<pre><code><span><span class="co">## memoMDB miniHPO (version 0.1, Patrice Godard): Very small extract of the human phenotype ontology</span></span>
<span><span class="co">##    - 3 tables with 10 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## No collection member</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## For demonstrating ReDaMoR and TKCat capabilities, a very few information from the HPO (human phenotype ontology) has been extracted</span></span>
<span><span class="co">## (https://hpo.jax.org/app/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Timestamp: 2024-05-22 15:42:44.394634</span></span>
<span><span class="co">## </span></span></code></pre>
</div>
<div class="section level3">
<h3 id="min-coll-memb">Documenting collection members<a class="anchor" aria-label="anchor" href="#min-coll-memb"></a>
</h3>
<p>In the HPO example, one table regards human phenotypes
(<em>HPO_hp</em>) and another human diseases (<em>HPO_diseases</em>).
These concepts are general and referenced in many other knowledge or
data resources (e.g. database providing information about disease
genetics). Therefore, documenting formally such concepts will help to
identify how to connect the HPO example to other resources referencing
the same or related concepts.</p>
<p>In TKCat, these central concepts are referred as members of
<em>collections</em>. <em>Collections</em> are pre-defined and members
must be documented according to this definition. There are currently two
collections provided within the TKCat package:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/list_local_collections.html">list_local_collections</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   title     description                                  </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> BE        Collection of biological entity (BE) concepts</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition Collection of condition concepts</span></span></code></pre>
<p>Additional collections can be defined by users according to their
needs. Further information about collections implementation is provided
in the <a href="#collections">appendix</a>.</p>
<p>So far, there is no collection member documented in the HPO example
described above, as indicated by the <em>“No collection member”</em>
statement displayed when printing the object:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span></span></code></pre></div>
<pre><code><span><span class="co">## memoMDB miniHPO (version 0.1, Patrice Godard): Very small extract of the human phenotype ontology</span></span>
<span><span class="co">##    - 3 tables with 10 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## No collection member</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## For demonstrating ReDaMoR and TKCat capabilities, a very few information from the HPO (human phenotype ontology) has been extracted</span></span>
<span><span class="co">## (https://hpo.jax.org/app/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Timestamp: 2024-05-22 15:42:44.394634</span></span>
<span><span class="co">## </span></span></code></pre>
<p>However, as just discussed, the <em>HPO_hp</em> table refers to human
phenotypes and the <em>HPO_diseases</em> table to human diseases. These
concept corresponds to conditions and those tables can be documented as
member of the <em>Condition</em> collection.</p>
<p><em>Condition</em> members are documented calling the
<code><a href="../reference/add_collection_member.html">add_collection_member()</a></code> function on the MDB object. The two
other main arguments are the name of the <code>collection</code> and the
name of the <code>table</code> in the MDB object. The other arguments to
be provided depend on the collection. For <em>Condition</em> members,
three additional arguments must be provided:</p>
<ul>
<li>
<code>condition</code> indicate the type of the condition
(“Phenotype” or “Disease”)</li>
<li>
<code>source</code> a reference source of the condition
identifier</li>
<li>
<code>identifier</code> a condition identifier</li>
</ul>
<p>The functions <code><a href="../reference/get_local_collection.html">get_local_collection()</a></code> and
<code><a href="../reference/show_collection_def.html">show_collection_def()</a></code> can be used together to identify
valid arguments:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_local_collection.html">get_local_collection</a></span><span class="op">(</span><span class="st">"Condition"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu"><a href="../reference/show_collection_def.html">show_collection_def</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Condition collection: Collection of condition concepts</span></span>
<span><span class="co">## Arguments (non-mandatory arguments are between parentheses):</span></span>
<span><span class="co">##    - condition:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span>
<span><span class="co">##    - source:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span>
<span><span class="co">##    - identifier:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span></code></pre>
<p>When calling <code><a href="../reference/add_collection_member.html">add_collection_member()</a></code>, these arguments
must be provided as a list with 2 elements named “value” (a character)
and “static” (a logical). If “static” is TRUE, “value” corresponds to
the information shared by all the rows of the table. If “static” is
FALSE, “value” indicates the name of the column which provides this
information for each row.</p>
<p>The example below shows how the <em>HPO_hp</em> table is documented
as a member of the <em>Condition</em> collection.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span><span class="op">$</span><span class="va">HPO_hp</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 500 × 4</span></span></span>
<span><span class="co">##    id      name                                              description   level</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 0000002 Abnormality of body height                        Deviation fr…     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 0000009 Functional abnormality of the bladder             Dysfunction …     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 0000014 Abnormality of the bladder                        An abnormali…     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 0000017 Nocturia                                          Abnormally i…     7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 0000019 Urinary hesitancy                                 Difficulty i…     7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 0000021 Megacystis                                        Dilatation o…     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 0000022 Abnormality of male internal genitalia            An abnormali…     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 0000024 Prostatitis                                       The presence…     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 0000025 Functional abnormality of male internal genitalia <span style="color: #BB0000;">NA</span>                6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 0000030 Testicular gonadoblastoma                         The presence…     9</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 490 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_collection_member.html">add_collection_member</a></span><span class="op">(</span></span>
<span>   <span class="va">mhpo_db</span>, collection<span class="op">=</span><span class="st">"Condition"</span>, table<span class="op">=</span><span class="st">"HPO_hp"</span>,</span>
<span>   condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"Phenotype"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>   source<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"HP"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>   identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"id"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>All rows in this table correspond to a condition of type “Phenotype”
(<code>condition=list(value="Phenotype", static=TRUE)</code>). The
phenotype identifiers are all taken from the same source, “HP”
(<code>source=list(value="HP", static=TRUE)</code>). The phenotype
identifiers are provided in the “id” column of the table
(<code>identifier=list(value="id", static=FALSE)</code>).</p>
<p>The example below shows how the <em>HPO_disease</em> table is
documented also as a member of the <em>Condition</em> collection. In
this case, the source of disease identifier can be different from one
row to the other and is provided in the “db” column
(<code>source=list(value="db", static=FALSE)</code>).</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_collection_member.html">add_collection_member</a></span><span class="op">(</span></span>
<span>   <span class="va">mhpo_db</span>, collection<span class="op">=</span><span class="st">"Condition"</span>, table<span class="op">=</span><span class="st">"HPO_diseases"</span>,</span>
<span>   condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"Disease"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>   source<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"db"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>   identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"id"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now, the existence of collection members is shown when printing the
MDB object:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mhpo_db</span></span></code></pre></div>
<pre><code><span><span class="co">## memoMDB miniHPO (version 0.1, Patrice Godard): Very small extract of the human phenotype ontology</span></span>
<span><span class="co">##    - 3 tables with 10 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Collection members: </span></span>
<span><span class="co">##    - 2 Condition members</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## For demonstrating ReDaMoR and TKCat capabilities, a very few information from the HPO (human phenotype ontology) has been extracted</span></span>
<span><span class="co">## (https://hpo.jax.org/app/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Timestamp: 2024-05-22 15:42:44.394634</span></span>
<span><span class="co">## </span></span></code></pre>
<p>And the documented collection members of an MDB can be displayed as
following:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">mhpo_db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 9</span></span></span>
<span><span class="co">##   collection cid                   resource   mid table field static value type </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Condition  miniHPO_Condition_1.0 miniHPO      1 HPO_… cond… TRUE   Phen… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition  miniHPO_Condition_1.0 miniHPO      1 HPO_… sour… TRUE   HP    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Condition  miniHPO_Condition_1.0 miniHPO      1 HPO_… iden… FALSE  id    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Condition  miniHPO_Condition_1.0 miniHPO      2 HPO_… cond… TRUE   Dise… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Condition  miniHPO_Condition_1.0 miniHPO      2 HPO_… sour… FALSE  db    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Condition  miniHPO_Condition_1.0 miniHPO      2 HPO_… iden… FALSE  id    <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>The use of collection members to link or integrate different MDBs
will be described <a href="#merging-with-collections">later</a> in this
document</p>
</div>
<div class="section level3">
<h3 id="min-writing">Writing an MDB in files<a class="anchor" aria-label="anchor" href="#min-writing"></a>
</h3>
<p>Once an MDB has been created and documented in can be written in a
directory:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmpDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/as_fileMDB.html">as_fileMDB</a></span><span class="op">(</span><span class="va">mhpo_db</span>, path<span class="op">=</span><span class="va">tmpDir</span>, htmlModel<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The structure of the created directory is the following:</p>
<pre><code><span><span class="co">## miniHPO                                         </span></span>
<span><span class="co">##  ¦--data                                        </span></span>
<span><span class="co">##  ¦   ¦--HPO_diseaseHP.txt.gz                    </span></span>
<span><span class="co">##  ¦   ¦--HPO_diseases.txt.gz                     </span></span>
<span><span class="co">##  ¦   °--HPO_hp.txt.gz                           </span></span>
<span><span class="co">##  ¦--DESCRIPTION.json                            </span></span>
<span><span class="co">##  °--model                                       </span></span>
<span><span class="co">##      ¦--Collections                             </span></span>
<span><span class="co">##      ¦   °--Condition-miniHPO_Condition_1.0.json</span></span>
<span><span class="co">##      °--miniHPO.json</span></span></code></pre>
<p>All the data are in the <em>data</em> folder whereas the data model
and collection members are written in json files in the <em>model</em>
folder. The <em>DESCRIPTION.json</em> file gather db information and
information about how to read the data files (i.e. <code>delim</code>,
<code>na</code>).</p>
<p>This folder can be shared and it’s then easy to get all the data and
the corresponding documentation from it back in R:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/read_fileMDB.html">read_fileMDB</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmpDir</span>, <span class="st">"miniHPO"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold; text-decoration: underline;">miniHPO</span></span></span>
<span><span class="co">## <span style="color: #BBBBBB; background-color: #00BB00;">SUCCESS</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="font-weight: bold;">Check configuration</span></span></span>
<span><span class="co">##    - Optional checks: </span></span>
<span><span class="co">##    - Maximum number of records: 10</span></span></code></pre>
<pre><code><span><span class="co">## fileMDB miniHPO (version 0.1, Patrice Godard): Very small extract of the human phenotype ontology</span></span>
<span><span class="co">##    - 3 tables with 10 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Collection members: </span></span>
<span><span class="co">##    - 2 Condition members</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## For demonstrating ReDaMoR and TKCat capabilities, a very few information from the HPO (human phenotype ontology) has been extracted</span></span>
<span><span class="co">## (https://hpo.jax.org/app/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Timestamp: 2024-05-22 15:42:44</span></span>
<span><span class="co">## </span></span></code></pre>
<p>Also writing these data and related information in text files make
them convenient to share with people using them in other analytical
environments than R.</p>
</div>
</div>
<div class="section level2">
<h2 id="leveraging-mdb">Leveraging MDB<a class="anchor" aria-label="anchor" href="#leveraging-mdb"></a>
</h2>
<p>The former section showed how to create and save an MDB object. This
section describes how MDBs can be used, filtered and combined to
efficiently leverage their content.</p>
<p>As a reminder, a modeled database (MDB) in TKCat gathers the
following information:</p>
<ul>
<li>General database information including a mandatory <em>name</em> and
optionally the following fields: <em>title</em>, <em>description</em>,
<em>url</em>, <em>version</em> and <em>maintainer</em>.</li>
<li>A <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a> data
model.</li>
<li>A list of tables corresponding to reference concepts shared by
different MDBs. The way these concepts are identified is defined in
specific documents called collections.</li>
<li>The data themselves organized according to the data model.</li>
</ul>
<div class="section level3">
<h3 id="loading-example-data-1">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data-1"></a>
</h3>
<p>To illustrate how MDBs can be used, some example data are provided
within the <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a> and
the TKCat package. The following paragraphs show how to load them in the
R session.</p>
<div class="section level4">
<h4 id="hpo">HPO<a class="anchor" aria-label="anchor" href="#hpo"></a>
</h4>
<p>A subset of the <a href="https://hpo.jax.org/app/" class="external-link">Human Phenotype
Ontology (HPO)</a> is provided within the <a href="https://patzaw.github.io/ReDaMoR/" class="external-link">ReDaMoR</a> package. The HPO
aims to provide a standardized vocabulary of phenotypic abnormalities
encountered in human diseases <span class="citation">(<a href="#ref-kohler_expansion_2019">Köhler et al. 2019</a>)</span>. An MDB
object based on files (see <a href="#mdb-implementations">MDB
implementations</a>) can be read as shown below. As explained above, the
data provided by the <code>path</code> parameter are documented with a
model (<code>dataModel</code> parameter) and general information
(<code>dbInfo</code> parameter).</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_hpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fileMDB.html">read_fileMDB</a></span><span class="op">(</span></span>
<span>   path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/HPO-subset"</span>, package<span class="op">=</span><span class="st">"ReDaMoR"</span><span class="op">)</span>,</span>
<span>   dataModel<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/HPO-model.json"</span>, package<span class="op">=</span><span class="st">"ReDaMoR"</span><span class="op">)</span>,</span>
<span>   dbInfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      <span class="st">"name"</span><span class="op">=</span><span class="st">"HPO"</span>,</span>
<span>      <span class="st">"title"</span><span class="op">=</span><span class="st">"Data extracted from the HPO database"</span>,</span>
<span>      <span class="st">"description"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>         <span class="st">"This is a very small subset of the HPO!"</span>,</span>
<span>         <span class="st">"Visit the reference URL for more information."</span></span>
<span>      <span class="op">)</span>,</span>
<span>      <span class="st">"url"</span><span class="op">=</span><span class="st">"http://human-phenotype-ontology.github.io/"</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold; text-decoration: underline;">HPO</span></span></span>
<span><span class="co">## <span style="color: #BBBBBB; background-color: #00BB00;">SUCCESS</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="font-weight: bold;">Check configuration</span></span></span>
<span><span class="co">##    - Optional checks: </span></span>
<span><span class="co">##    - Maximum number of records: 10</span></span></code></pre>
<p>The message displayed in the console indicates if the data fit the
data model. It relies on the <code><a href="https://patzaw.github.io/ReDaMoR/reference/confront_data.html" class="external-link">ReDaMoR::confront_data()</a></code>
functions and check by default the first 10 rows of each file.</p>
<p>The data model can then be drawn.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="va">file_hpo</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"nodes":{"tableName":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms"],"label":["<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_altId<\/b>\n    - id {character - idx.1}\n    - alt {character}","<b>HPO_sourceFiles<\/b>\n    - <b>*url<\/b> {character - uidx.1}\n    - current {Date}","<b>HPO_diseases<\/b>\n    - <b>db<\/b> {character - uidx.1}\n    - <b>id<\/b> {character - uidx.1}\n    - label {character}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_diseaseSynonyms<\/b>\n    - db {character - idx.1}\n    - id {character - idx.1}\n    - synonym {character}\n    - preferred {logical}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}","<b>HPO_descendants<\/b>\n    - id {character - idx.1}\n    - descendant {character - idx.2}","<b>HPO_synonyms<\/b>\n    - id {character}\n    - synonym {character}\n    - type {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_altId<\/strong> (Alternative HP identifiers)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>alt<\/strong>: Alternative identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_sourceFiles<\/strong> (Files from which data were extracted)<\/p> <ul> <li><strong>url<\/strong>: Source file URL<\/li> <li><strong>current<\/strong>: The date of the current source file<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong> (Diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>label<\/strong>: Disease lable (preferred synonym)<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseSynonyms<\/strong> (Disease synonyms)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>synonym<\/strong>: Disease synonym<\/li> <li><strong>preferred<\/strong>: Is synonym preferred<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_descendants<\/strong> (Indirect descendants of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>descendant<\/strong>: Descendant (child or child of child...) HP ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_synonyms<\/strong> (Synonyms of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>synonym<\/strong>: HP synonym<\/li> <li><strong>type<\/strong>: Synonym type<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left"],"x":[-167,-154,262,320,75,321,-328,-101,-360],"y":[-177,-38,-336,-170,-172,-24,-331,-360,-36],"color.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE"],"color.border":["black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE"],"id":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms"]},"edges":{"id":["HPO_altId->HPO_hp: id->id","HPO_diseaseHP->HPO_hp: hp->id","HPO_diseaseHP->HPO_diseases: db->db && id->id","HPO_diseaseSynonyms->HPO_diseases: db->db && id->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id","HPO_descendants->HPO_hp: id->id","HPO_descendants->HPO_hp: descendant->id","HPO_synonyms->HPO_hp: id->id"],"to":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseases","HPO_hp","HPO_hp","HPO_hp","HPO_hp","HPO_hp"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_altId<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseSynonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">descendant<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_synonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["id","hp",["db","id"],["db","id"],"id","parent","id","descendant","id"],"tf":["id","id",["db","id"],["db","id"],"id","id","id","id","id"],"from":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_descendants","HPO_descendants","HPO_synonyms"],"arrows":["to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0,0.2,-0.2,-0.2,0.2,0],"selfReferenceSize":[30,30,30,30,30,50,30,50,30],"uef":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseases","HPO_hp","HPO_hp","HPO_descendants","HPO_descendants","HPO_hp"],"uet":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_hp","HPO_hp","HPO_synonyms"],"ue":["HPO_altId<->HPO_hp","HPO_diseaseHP<->HPO_hp","HPO_diseaseHP<->HPO_diseases","HPO_diseases<->HPO_diseaseSynonyms","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents","HPO_descendants<->HPO_hp","HPO_descendants<->HPO_hp","HPO_hp<->HPO_synonyms"],"color.color":["black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><p>The data model shows that this MDB contains the 3 tables taken into
account in the minimal example. The additional tables provides mainly
supplementary details regarding phenotype and diseases. Still, the
<em>HPO_hp</em> and the <em>HPO_disease</em> table are members of the
<em>Condition</em> collection and can be documented as such, as <a href="#min-coll-memb">explained above</a>.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_hpo</span> <span class="op">&lt;-</span> <span class="va">file_hpo</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu"><a href="../reference/add_collection_member.html">add_collection_member</a></span><span class="op">(</span></span>
<span>      collection<span class="op">=</span><span class="st">"Condition"</span>, table<span class="op">=</span><span class="st">"HPO_hp"</span>,</span>
<span>      condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"Phenotype"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      source<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"HP"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"id"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>   <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu"><a href="../reference/add_collection_member.html">add_collection_member</a></span><span class="op">(</span></span>
<span>      collection<span class="op">=</span><span class="st">"Condition"</span>, table<span class="op">=</span><span class="st">"HPO_diseases"</span>,</span>
<span>      condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"Disease"</span>, static<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      source<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"db"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>      identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>value<span class="op">=</span><span class="st">"id"</span>, static<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="clinvar">ClinVar<a class="anchor" aria-label="anchor" href="#clinvar"></a>
</h4>
<p>A subset of the <a href="https://www.ncbi.nlm.nih.gov/clinvar/" class="external-link">ClinVar</a> database is
provided within this package. ClinVar is a freely accessible, public
archive of reports of the relationships among human variations and
phenotypes, with supporting evidence <span class="citation">(<a href="#ref-landrum_clinvar_2018">Landrum et al. 2018</a>)</span>. This
resource can be read as a <code>fileMDB</code> as shown above. However,
in this case all the documenting information is included in the resource
directory, making it easier to read as <a href="#min-writing">explained
above</a>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_clinvar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fileMDB.html">read_fileMDB</a></span><span class="op">(</span></span>
<span>   path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/ClinVar"</span>, package<span class="op">=</span><span class="st">"TKCat"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold; text-decoration: underline;">ClinVar</span></span></span>
<span><span class="co">## <span style="color: #BBBBBB; background-color: #00BB00;">SUCCESS</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="font-weight: bold;">Check configuration</span></span></span>
<span><span class="co">##    - Optional checks: </span></span>
<span><span class="co">##    - Maximum number of records: 10</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_clinvar</span></span></code></pre></div>
<pre><code><span><span class="co">## fileMDB ClinVar (version 0.9, Patrice Godard &lt;patrice.godard@ucb.com&gt;): Data extracted from the ClinVar database</span></span>
<span><span class="co">##    - 21 tables with 86 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Collection members: </span></span>
<span><span class="co">##    - 1 BE member</span></span>
<span><span class="co">##    - 2 Condition members</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. This is a very small subset of ClinVar! Visit the reference URL for more information.</span></span>
<span><span class="co">## (https://www.ncbi.nlm.nih.gov/clinvar/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span></code></pre>
</div>
<div class="section level4">
<h4 id="chembl">CHEMBL<a class="anchor" aria-label="anchor" href="#chembl"></a>
</h4>
<p>Similarly, a self-documented subset of the <a href="https://www.ebi.ac.uk/chembl/" class="external-link">CHEMBL</a> database is also
provided in the TKCat package. It can be read the same way.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_chembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fileMDB.html">read_fileMDB</a></span><span class="op">(</span></span>
<span>   path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/CHEMBL"</span>, package<span class="op">=</span><span class="st">"TKCat"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold; text-decoration: underline;">CHEMBL</span></span></span>
<span><span class="co">## <span style="color: #BBBBBB; background-color: #00BB00;">SUCCESS</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="font-weight: bold;">Check configuration</span></span></span>
<span><span class="co">##    - Optional checks: </span></span>
<span><span class="co">##    - Maximum number of records: 10</span></span></code></pre>
<p>CHEMBL is a manually curated chemical database of bioactive molecules
with drug-like properties <span class="citation">(<a href="#ref-mendez_chembl_2019">Mendez et al. 2019</a>)</span>.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_chembl</span></span></code></pre></div>
<pre><code><span><span class="co">## fileMDB CHEMBL (version 0.2, Liesbeth François &lt;liesbeth.francois@ucb.com&gt;): Data extracted from the CHEMBL database</span></span>
<span><span class="co">##    - 10 tables with 61 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Collection members: </span></span>
<span><span class="co">##    - 1 BE member</span></span>
<span><span class="co">##    - 1 Condition member</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## CHEMBL is a manually curated chemical database of bioactive molecules with drug-like properties. This is a very small subset of CHEMBL! Visit the reference URL for more information.</span></span>
<span><span class="co">## (https://www.ebi.ac.uk/chembl/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="mdb-implementations">MDB implementations<a class="anchor" aria-label="anchor" href="#mdb-implementations"></a>
</h3>
<p>There are 3 main implementations of MDBs:</p>
<ul>
<li><p><strong>fileMDB</strong> objects keep the data in files and load
them only when requested by the user. These implementation is the first
one which is used when reading MDB as demonstrated in the examples
above.</p></li>
<li><p><strong>memoMDB</strong> objects have all the data loaded in
memory. These objects are very easy to use but can take time to load and
can use a lot of memory.</p></li>
<li><p><strong>chMDB</strong> objects get the data from a <a href="https://clickhouse.com/" class="external-link">ClickHouse</a> database providing a
catalog of MDBs as described in the <a href="#chTKCat">dedicated
section</a>.</p></li>
</ul>
<p>The different implementations can be converted to each others using
<code><a href="../reference/as_fileMDB.html">as_fileMDB()</a></code>, <code><a href="../reference/as_memoMDB.html">as_memoMDB()</a></code> and
<code><a href="../reference/as_chMDB.html">as_chMDB()</a></code> functions.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">memo_clinvar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_memoMDB.html">as_memoMDB</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>units<span class="op">=</span><span class="st">"Kb"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 155.2 Kb</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">memo_clinvar</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>units<span class="op">=</span><span class="st">"Kb"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 760.5 Kb</span></span></code></pre>
<p>A fourth implementation is <strong>metaMDB</strong> which combines
several MDBs glued together with relational tables (see the <a href="#merging-with-collections">Merging with collections</a> part).</p>
<p>Most of the functions described below work with any MDB
implementation, and a few functions are specific to each
implementation.</p>
</div>
<div class="section level3">
<h3 id="exploring-information">Exploring information<a class="anchor" aria-label="anchor" href="#exploring-information"></a>
</h3>
<p>General information can be retrieved (and potentialy updated) using
the <code><a href="../reference/db_info.html">db_info()</a></code> function.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/db_info.html">db_info</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $name</span></span>
<span><span class="co">## [1] "ClinVar"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $title</span></span>
<span><span class="co">## [1] "Data extracted from the ClinVar database"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $description</span></span>
<span><span class="co">## [1] "ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. This is a very small subset of ClinVar! Visit the reference URL for more information."</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $url</span></span>
<span><span class="co">## [1] "https://www.ncbi.nlm.nih.gov/clinvar/"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $version</span></span>
<span><span class="co">## [1] "0.9"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $maintainer</span></span>
<span><span class="co">## [1] "Patrice Godard &lt;patrice.godard@ucb.com&gt;"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $timestamp</span></span>
<span><span class="co">## [1] NA</span></span></code></pre>
<p>As shown above the data model of an MDB can be retrieved and plot the
following way.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-1fb4450895fe099f74a1" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-1fb4450895fe099f74a1">{"x":{"nodes":{"tableName":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames"],"label":["<b>ClinVar_ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>ClinVar_rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVar_ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>ClinVar_rcvaInhMode<\/b>\n    - rcvaId {integer}\n    - inhMode {character}","<b>ClinVar_rcvaObservedIn<\/b>\n    - rcvaId {integer}\n    - origin {character}\n    - taxonomyId {character}\n    - species {character}\n    - affectedStatus {character}\n    - (numberTested {integer})","<b>ClinVar_rcvaTraits<\/b>\n    - rcvaId {integer}\n    - t.id {integer - idx.1}\n    - traitType {character}","<b>ClinVar_clinSigOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_revStatOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>ClinVar_variants<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - type {character}\n    - (name {character})","<b>ClinVar_cvaObservedIn<\/b>\n    - cvaId {integer}\n    - origin {character}\n    - species {character}\n    - affectedStatus {character}","<b>ClinVar_cvaSubmitters<\/b>\n    - cvaId {integer}\n    - submitter {character}\n    - primary {logical}","<b>ClinVar_traits<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - name {character}","<b>ClinVar_varEntrez<\/b>\n    - varId {integer - idx.1}\n    - entrez {integer}\n    - type {character}","<b>ClinVar_varAttributes<\/b>\n    - varId {integer - idx.1}\n    - Type {character}\n    - (integerValue {integer})\n    - (Change {character})\n    - (value {character})","<b>ClinVar_varCytoLoc<\/b>\n    - varId {integer - idx.1}\n    - location {character}","<b>ClinVar_varNames<\/b>\n    - varId {integer - idx.1}\n    - name {character}\n    - type {character}","<b>ClinVar_varSeqLoc<\/b>\n    - varId {integer - idx.1}\n    - Accession {character}\n    - (alternateAllele {character})\n    - Assembly {character}\n    - AssemblyAccessionVersion {character}\n    - AssemblyStatus {character}\n    - Chr {character}\n    - display_start {integer}\n    - (display_stop {integer})\n    - (innerStart {integer})\n    - (innerStop {integer})\n    - (outerStart {integer})\n    - (outerStop {integer})\n    - (referenceAllele {character})\n    - (start {integer})\n    - (stop {integer})\n    - (Strand {character})\n    - (variantLength {integer})","<b>ClinVar_varXRef<\/b>\n    - varId {integer - idx.1}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitCref<\/b>\n    - t.id {integer}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>ClinVar_traitNames<\/b>\n    - t.id {integer}\n    - name {character}\n    - type {character}","<b>ClinVar_entrezNames<\/b>\n    - <b>*entrez<\/b> {integer - uidx.1}\n    - name {character}\n    - symbol {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaInhMode<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>inhMode<\/strong>: Mode of inheritance<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaObservedIn<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>taxonomyId<\/strong>: Taxonomy ID<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <li><strong>numberTested<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_rcvaTraits<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>traitType<\/strong>: Trait type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_clinSigOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of clinical significance: the lowest the best clinical outcome.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_revStatOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of review status: the lowest the least significant.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_variants<\/strong><\/p> <ul> <li><strong>id<\/strong>: Variant ID<\/li> <li><strong>type<\/strong>: Variant type<\/li> <li><strong>name<\/strong>: Variant preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaObservedIn<\/strong><\/p> <ul> <li><strong>cvaId<\/strong><\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_cvaSubmitters<\/strong><\/p> <ul> <li><strong>cvaId<\/strong>: ClinVar Assertion ID<\/li> <li><strong>submitter<\/strong>: Submitter name<\/li> <li><strong>primary<\/strong>: The submitter is of pirmary type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traits<\/strong><\/p> <ul> <li><strong>id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varEntrez<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>type<\/strong>: Type of the association between the variant and the gene<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varAttributes<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Type<\/strong><\/li> <li><strong>integerValue<\/strong><\/li> <li><strong>Change<\/strong><\/li> <li><strong>value<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varCytoLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>location<\/strong>: Cytogenic location<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varNames<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>name<\/strong>: Variant name<\/li> <li><strong>type<\/strong>: Variant name type: Preferred or Alternate<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varSeqLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Accession<\/strong>: Sequence accession ID<\/li> <li><strong>alternateAllele<\/strong>: Alternate allele<\/li> <li><strong>Assembly<\/strong>: Assembly<\/li> <li><strong>AssemblyAccessionVersion<\/strong>: Version of assembly accession<\/li> <li><strong>AssemblyStatus<\/strong>: Status of the assembly: current or previous<\/li> <li><strong>Chr<\/strong>: Chromosome name<\/li> <li><strong>display_start<\/strong><\/li> <li><strong>display_stop<\/strong><\/li> <li><strong>innerStart<\/strong><\/li> <li><strong>innerStop<\/strong><\/li> <li><strong>outerStart<\/strong><\/li> <li><strong>outerStop<\/strong><\/li> <li><strong>referenceAllele<\/strong>: Reference allele<\/li> <li><strong>start<\/strong><\/li> <li><strong>stop<\/strong><\/li> <li><strong>Strand<\/strong><\/li> <li><strong>variantLength<\/strong>: Length of the variant<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_varXRef<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitCref<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_traitNames<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait name<\/li> <li><strong>type<\/strong>: \"Alternate\" or \"Preferred\"<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVar_entrezNames<\/strong><\/p> <ul> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>name<\/strong>: Gene name<\/li> <li><strong>symbol<\/strong>: Gene symbol<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left"],"x":[-370,55,-767,-311,-114,-368,-540,-443,430,-766,-771,-371,266,243,486,494,751,721,-507,-264,265],"y":[344,345,338,521,498,168,524,603,346,186,488,31,178,503,489,194,214,461,-87,-91,37],"color.background":["#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0"],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0"],"id":["ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_variants","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_traits","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames","ClinVar_entrezNames"]},"edges":{"id":["ClinVar_ReferenceClinVarAssertion->ClinVar_clinSigOrder: clinicalSignificance->label","ClinVar_ReferenceClinVarAssertion->ClinVar_revStatOrder: reviewStatus->label","ClinVar_rcvaVariant->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaVariant->ClinVar_variants: varId->id","ClinVar_ClinVarAssertions->ClinVar_ReferenceClinVarAssertion: cvs->cvs","ClinVar_rcvaInhMode->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaObservedIn->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_rcvaTraits->ClinVar_traits: t.id->id","ClinVar_rcvaTraits->ClinVar_ReferenceClinVarAssertion: rcvaId->id","ClinVar_cvaObservedIn->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_cvaSubmitters->ClinVar_ClinVarAssertions: cvaId->id","ClinVar_varEntrez->ClinVar_variants: varId->id","ClinVar_varEntrez->ClinVar_entrezNames: entrez->entrez","ClinVar_varAttributes->ClinVar_variants: varId->id","ClinVar_varCytoLoc->ClinVar_variants: varId->id","ClinVar_varNames->ClinVar_variants: varId->id","ClinVar_varSeqLoc->ClinVar_variants: varId->id","ClinVar_varXRef->ClinVar_variants: varId->id","ClinVar_traitCref->ClinVar_traits: t.id->id","ClinVar_traitNames->ClinVar_traits: t.id->id"],"to":["ClinVar_clinSigOrder","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_variants","ClinVar_entrezNames","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traits","ClinVar_traits"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_clinSigOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">clinicalSignificance<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_revStatOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">reviewStatus<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaInhMode<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_cvaSubmitters<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_entrezNames<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varAttributes<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varCytoLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varSeqLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_varXRef<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traitNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVar_traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["clinicalSignificance","reviewStatus","rcvaId","varId","cvs","rcvaId","rcvaId","t.id","rcvaId","cvaId","cvaId","varId","entrez","varId","varId","varId","varId","varId","t.id","t.id"],"tf":["label","label","id","id","cvs","id","id","id","id","id","id","id","entrez","id","id","id","id","id","id","id"],"from":["ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_varEntrez","ClinVar_varEntrez","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traitCref","ClinVar_traitNames"],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[0,-0,-0,-0,-0,-0,-0,-0,-0,0,0,-0,0,-0,-0,0,0,0,-0,-0],"selfReferenceSize":[30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30],"uef":["ClinVar_clinSigOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant","ClinVar_rcvaVariant","ClinVar_ClinVarAssertions","ClinVar_rcvaInhMode","ClinVar_rcvaObservedIn","ClinVar_rcvaTraits","ClinVar_rcvaTraits","ClinVar_ClinVarAssertions","ClinVar_ClinVarAssertions","ClinVar_varEntrez","ClinVar_entrezNames","ClinVar_varAttributes","ClinVar_varCytoLoc","ClinVar_variants","ClinVar_variants","ClinVar_variants","ClinVar_traitCref","ClinVar_traitNames"],"uet":["ClinVar_ReferenceClinVarAssertion","ClinVar_revStatOrder","ClinVar_ReferenceClinVarAssertion","ClinVar_variants","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion","ClinVar_traits","ClinVar_ReferenceClinVarAssertion","ClinVar_cvaObservedIn","ClinVar_cvaSubmitters","ClinVar_variants","ClinVar_varEntrez","ClinVar_variants","ClinVar_variants","ClinVar_varNames","ClinVar_varSeqLoc","ClinVar_varXRef","ClinVar_traits","ClinVar_traits"],"ue":["ClinVar_clinSigOrder<->ClinVar_ReferenceClinVarAssertion","ClinVar_ReferenceClinVarAssertion<->ClinVar_revStatOrder","ClinVar_rcvaVariant<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaVariant<->ClinVar_variants","ClinVar_ClinVarAssertions<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaInhMode<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaObservedIn<->ClinVar_ReferenceClinVarAssertion","ClinVar_rcvaTraits<->ClinVar_traits","ClinVar_rcvaTraits<->ClinVar_ReferenceClinVarAssertion","ClinVar_ClinVarAssertions<->ClinVar_cvaObservedIn","ClinVar_ClinVarAssertions<->ClinVar_cvaSubmitters","ClinVar_varEntrez<->ClinVar_variants","ClinVar_entrezNames<->ClinVar_varEntrez","ClinVar_varAttributes<->ClinVar_variants","ClinVar_varCytoLoc<->ClinVar_variants","ClinVar_variants<->ClinVar_varNames","ClinVar_variants<->ClinVar_varSeqLoc","ClinVar_variants<->ClinVar_varXRef","ClinVar_traitCref<->ClinVar_traits","ClinVar_traitNames<->ClinVar_traits"],"color.color":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><p>Tables names can be listed with the <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> function and
potentially renamed with <code>names()&lt;-</code> or
<code><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename()</a></code> functions (the tables have been renamed here to
improve the readability of the following examples).</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ClinVar_ReferenceClinVarAssertion" "ClinVar_rcvaVariant"              </span></span>
<span><span class="co">##  [3] "ClinVar_ClinVarAssertions"         "ClinVar_rcvaInhMode"              </span></span>
<span><span class="co">##  [5] "ClinVar_rcvaObservedIn"            "ClinVar_rcvaTraits"               </span></span>
<span><span class="co">##  [7] "ClinVar_clinSigOrder"              "ClinVar_revStatOrder"             </span></span>
<span><span class="co">##  [9] "ClinVar_variants"                  "ClinVar_cvaObservedIn"            </span></span>
<span><span class="co">## [11] "ClinVar_cvaSubmitters"             "ClinVar_traits"                   </span></span>
<span><span class="co">## [13] "ClinVar_varEntrez"                 "ClinVar_varAttributes"            </span></span>
<span><span class="co">## [15] "ClinVar_varCytoLoc"                "ClinVar_varNames"                 </span></span>
<span><span class="co">## [17] "ClinVar_varSeqLoc"                 "ClinVar_varXRef"                  </span></span>
<span><span class="co">## [19] "ClinVar_traitCref"                 "ClinVar_traitNames"               </span></span>
<span><span class="co">## [21] "ClinVar_entrezNames"</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_clinvar</span> <span class="op">&lt;-</span> <span class="va">file_clinvar</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu">set_names</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"ClinVar_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ReferenceClinVarAssertion" "rcvaVariant"              </span></span>
<span><span class="co">##  [3] "ClinVarAssertions"         "rcvaInhMode"              </span></span>
<span><span class="co">##  [5] "rcvaObservedIn"            "rcvaTraits"               </span></span>
<span><span class="co">##  [7] "clinSigOrder"              "revStatOrder"             </span></span>
<span><span class="co">##  [9] "variants"                  "cvaObservedIn"            </span></span>
<span><span class="co">## [11] "cvaSubmitters"             "traits"                   </span></span>
<span><span class="co">## [13] "varEntrez"                 "varAttributes"            </span></span>
<span><span class="co">## [15] "varCytoLoc"                "varNames"                 </span></span>
<span><span class="co">## [17] "varSeqLoc"                 "varXRef"                  </span></span>
<span><span class="co">## [19] "traitCref"                 "traitNames"               </span></span>
<span><span class="co">## [21] "entrezNames"</span></span></code></pre>
<p>The different collection members of an MDBs are listed with the
<code><a href="../reference/collection_members.html">collection_members()</a></code> function.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 10 × 9</span></span></span>
<span><span class="co">##    collection cid                  resource   mid table field static value type </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Condition  ClinVar_conditions_… ClinVar      2 trai… cond… TRUE   Dise… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Condition  ClinVar_conditions_… ClinVar      2 trai… iden… FALSE  id    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Condition  ClinVar_conditions_… ClinVar      2 trai… sour… TRUE   Clin… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Condition  ClinVar_conditions_… ClinVar      1 trai… cond… TRUE   Dise… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Condition  ClinVar_conditions_… ClinVar      1 trai… iden… FALSE  id    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Condition  ClinVar_conditions_… ClinVar      1 trai… sour… FALSE  db    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> BE         ClinVar_BE_1.0       ClinVar      1 entr… be    TRUE   Gene  <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> BE         ClinVar_BE_1.0       ClinVar      1 entr… iden… FALSE  entr… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> BE         ClinVar_BE_1.0       ClinVar      1 entr… orga… TRUE   Homo… Scie…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> BE         ClinVar_BE_1.0       ClinVar      1 entr… sour… TRUE   Entr… <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>The following functions are use to get the number of tables, the
number of fields per table and the number of records.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span>        <span class="co"># Number of tables</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 21</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span>       <span class="co"># Number of fields per table</span></span></code></pre></div>
<pre><code><span><span class="co">## ReferenceClinVarAssertion               rcvaVariant         ClinVarAssertions </span></span>
<span><span class="co">##                         8                         2                         4 </span></span>
<span><span class="co">##               rcvaInhMode            rcvaObservedIn                rcvaTraits </span></span>
<span><span class="co">##                         2                         6                         3 </span></span>
<span><span class="co">##              clinSigOrder              revStatOrder                  variants </span></span>
<span><span class="co">##                         2                         2                         3 </span></span>
<span><span class="co">##             cvaObservedIn             cvaSubmitters                    traits </span></span>
<span><span class="co">##                         4                         3                         2 </span></span>
<span><span class="co">##                 varEntrez             varAttributes                varCytoLoc </span></span>
<span><span class="co">##                         3                         5                         2 </span></span>
<span><span class="co">##                  varNames                 varSeqLoc                   varXRef </span></span>
<span><span class="co">##                         3                        18                         4 </span></span>
<span><span class="co">##                 traitCref                traitNames               entrezNames </span></span>
<span><span class="co">##                         4                         3                         3</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/count_records.html">count_records</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span> <span class="co"># Number of records per table</span></span></code></pre></div>
<pre><code><span><span class="co">## ReferenceClinVarAssertion               rcvaVariant         ClinVarAssertions </span></span>
<span><span class="co">##                       166                       166                       409 </span></span>
<span><span class="co">##               rcvaInhMode            rcvaObservedIn                rcvaTraits </span></span>
<span><span class="co">##                        16                       337                       166 </span></span>
<span><span class="co">##              clinSigOrder              revStatOrder                  variants </span></span>
<span><span class="co">##                        11                         2                       138 </span></span>
<span><span class="co">##             cvaObservedIn             cvaSubmitters                    traits </span></span>
<span><span class="co">##                       412                       416                        18 </span></span>
<span><span class="co">##                 varEntrez             varAttributes                varCytoLoc </span></span>
<span><span class="co">##                       145                      2262                       138 </span></span>
<span><span class="co">##                  varNames                 varSeqLoc                   varXRef </span></span>
<span><span class="co">##                       188                       280                       244 </span></span>
<span><span class="co">##                 traitCref                traitNames               entrezNames </span></span>
<span><span class="co">##                        50                        44                        20</span></span></code></pre>
<p>The <code><a href="../reference/count_records.html">count_records()</a></code> function can take a lot of time when
dealing with <em>fileMDB</em> objects if the data files are very large.
In such case it could be more efficient to list data file size
instead.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data_file_size.html">data_file_size</a></span><span class="op">(</span><span class="va">file_clinvar</span>, hr<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 21 × 3</span></span></span>
<span><span class="co">##    table                     size   compressed</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ReferenceClinVarAssertion 4.6 KB TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> rcvaVariant               947 B  TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ClinVarAssertions         4.2 KB TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> rcvaInhMode               152 B  TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> rcvaObservedIn            1.4 KB TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> rcvaTraits                788 B  TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> clinSigOrder              145 B  TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> revStatOrder              101 B  TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> variants                  2.1 KB TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> cvaObservedIn             1.8 KB TRUE      </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 11 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="pulling-subsetting-and-combining">Pulling, subsetting and combining<a class="anchor" aria-label="anchor" href="#pulling-subsetting-and-combining"></a>
</h3>
<p>There are several possible ways to pull data tables from MDBs. The
following lines return the same result displayed below (only once).</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data_tables.html">data_tables</a></span><span class="op">(</span><span class="va">file_clinvar</span>, <span class="st">"traitNames"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">file_clinvar</span><span class="op">[[</span><span class="st">"traitNames"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">file_clinvar</span><span class="op">$</span><span class="st">"traitNames"</span></span>
<span><span class="va">file_clinvar</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">traitNames</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 44 × 3</span></span></span>
<span><span class="co">##     t.id name                                                              type </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>   912 Chudley-McCullough syndrome                                       Pref…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>   912 Deafness, autosomal recessive 82                                  Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>   912 Deafness, bilateral sensorineural, and hydrocephalus due to fora… Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>   912 Deafness, sensorineural, with partial agenesis of the corpus cal… Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>  <span style="text-decoration: underline;">1</span>352 CTSD-Related Neuronal Ceroid-Lipofuscinosis                       Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>  <span style="text-decoration: underline;">1</span>352 Ceroid lipofuscinosis neuronal Cathepsin D-deficient              Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>  <span style="text-decoration: underline;">1</span>352 Neuronal ceroid lipofuscinosis 10                                 Pref…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>  <span style="text-decoration: underline;">1</span>352 Neuronal ceroid lipofuscinosis due to Cathepsin D deficiency      Alte…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>  <span style="text-decoration: underline;">1</span>481 Diabetes mellitus, neonatal, with congenital hypothyroidism       Pref…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>  <span style="text-decoration: underline;">1</span>481 NDH SYNDROME                                                      Alte…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 34 more rows</span></span></span></code></pre>
<p>MDBs can also be subset and combined. The corresponding functions
ensure that the data model is fulfilled by the data tables.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_clinvar</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## fileMDB ClinVar (version 0.9, Patrice Godard &lt;patrice.godard@ucb.com&gt;): Data extracted from the ClinVar database</span></span>
<span><span class="co">##    - 3 tables with 14 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## No collection member</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. This is a very small subset of ClinVar! Visit the reference URL for more information.</span></span>
<span><span class="co">## (https://www.ncbi.nlm.nih.gov/clinvar/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="va">igraph_available</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="va">file_hpo</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">auto_layout</span><span class="op">(</span>force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="va">file_hpo</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-10b3b7155e8045a1b2ad" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-10b3b7155e8045a1b2ad">{"x":{"nodes":{"tableName":["ReferenceClinVarAssertion","rcvaVariant","ClinVarAssertions","HPO_hp","HPO_diseaseHP","HPO_parents"],"label":["<b>ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true],"font.align":["left","left","left","left","left","left"],"x":[92.66618850429707,-76.0835644585421,240,-87.31116826524956,-240,47.49481804896098],"y":[-182.9647964101941,-240,-83.78394967109953,188.8044660731033,92.24285362634416,240],"color.background":["#DFFB86","#D9D9D9","#DFFB86","#9BC8FE","#D9D9D9","#F67FC4"],"color.border":["black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#DFFB86","#D9D9D9","#DFFB86","#9BC8FE","#D9D9D9","#F67FC4"],"id":["ReferenceClinVarAssertion","rcvaVariant","ClinVarAssertions","HPO_hp","HPO_diseaseHP","HPO_parents"]},"edges":{"id":["rcvaVariant->ReferenceClinVarAssertion: rcvaId->id","ClinVarAssertions->ReferenceClinVarAssertion: cvs->cvs","HPO_diseaseHP->HPO_hp: hp->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id"],"to":["ReferenceClinVarAssertion","ReferenceClinVarAssertion","HPO_hp","HPO_hp","HPO_hp"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["rcvaId","cvs","hp","id","parent"],"tf":["id","cvs","id","id","id"],"from":["rcvaVariant","ClinVarAssertions","HPO_diseaseHP","HPO_parents","HPO_parents"],"arrows":["to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0.2,-0.2],"selfReferenceSize":[30,30,30,30,50],"uef":["rcvaVariant","ClinVarAssertions","HPO_diseaseHP","HPO_hp","HPO_hp"],"uet":["ReferenceClinVarAssertion","ReferenceClinVarAssertion","HPO_hp","HPO_parents","HPO_parents"],"ue":["rcvaVariant<->ReferenceClinVarAssertion","ClinVarAssertions<->ReferenceClinVarAssertion","HPO_diseaseHP<->HPO_hp","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents"],"color.color":["black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><p>The function <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> concatenates the provided MDB after
checking that tables names are not duplicated. It does not integrate the
data with any relational table. This can achieved by merging the MDBs as
described in the <a href="#merging-with-collections">Merging with
collections</a> section.</p>
</div>
<div class="section level3">
<h3 id="filtering-and-joining">Filtering and joining<a class="anchor" aria-label="anchor" href="#filtering-and-joining"></a>
</h3>
<p>An MDB can be filtered by filtering one or several tables based on
field values. The filtering is propagated to other tables using the
embedded data model.</p>
<p>In the example below, the <code>file_clinvar</code> object is
filtered in order to focus on a few genes with pathogenic variants. The
table below compares the number of rows before (“ori”) and after
(“filt”) filtering.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_clinvar</span> <span class="op">&lt;-</span> <span class="va">file_clinvar</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>      entrezNames <span class="op">=</span> <span class="va">symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PIK3R2"</span>, <span class="st">"UGT1A8"</span><span class="op">)</span></span>
<span>   <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu">slice</span><span class="op">(</span>ReferenceClinVarAssertion<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span></span>
<span>      <span class="st">"pathogen"</span>,</span>
<span>      <span class="va">.</span><span class="op">$</span><span class="va">ReferenceClinVarAssertion</span><span class="op">$</span><span class="va">clinicalSignificance</span>,</span>
<span>      ignore.case<span class="op">=</span><span class="cn">TRUE</span></span>
<span>   <span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">left_join</span><span class="op">(</span></span>
<span>   <span class="fu"><a href="../reference/dims.html">dims</a></span><span class="op">(</span><span class="va">file_clinvar</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">name</span>, <span class="va">nrow</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="../reference/dims.html">dims</a></span><span class="op">(</span><span class="va">filtered_clinvar</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">name</span>, <span class="va">nrow</span><span class="op">)</span>,</span>
<span>   by<span class="op">=</span><span class="st">"name"</span>,</span>
<span>   suffix<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_ori"</span>, <span class="st">"_filt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 21 × 3</span></span></span>
<span><span class="co">##    name                      nrow_ori nrow_filt</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ReferenceClinVarAssertion      166         4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> rcvaVariant                    166         4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ClinVarAssertions              409        15</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> rcvaInhMode                     16         0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> rcvaObservedIn                 337        10</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> rcvaTraits                     166         4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> clinSigOrder                    11         3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> revStatOrder                     2         1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> variants                       138         3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> cvaObservedIn                  412        15</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 11 more rows</span></span></span></code></pre>
<p>The object returned by <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter()</a></code> or <code>slice</code> is
a <em>memoMDB</em>: all the data are in memory.</p>
<p>Tables can be easily joined to get diseases associated to the genes
of interest in a single table as shown below.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_traits</span> <span class="op">&lt;-</span> <span class="va">filtered_clinvar</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu"><a href="../reference/join_mdb_tables.html">join_mdb_tables</a></span><span class="op">(</span></span>
<span>      <span class="st">"entrezNames"</span>, <span class="st">"varEntrez"</span>, <span class="st">"variants"</span>, <span class="st">"rcvaVariant"</span>,</span>
<span>      <span class="st">"ReferenceClinVarAssertion"</span>, <span class="st">"rcvaTraits"</span>, <span class="st">"traits"</span></span>
<span>   <span class="op">)</span></span>
<span><span class="va">gene_traits</span><span class="op">$</span><span class="va">entrezNames</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu">select</span><span class="op">(</span><span class="va">symbol</span>, <span class="va">name</span>, <span class="va">variants.type</span>, <span class="va">variants.name</span>, <span class="va">traitType</span>, <span class="va">traits.name</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 6</span></span></span>
<span><span class="co">##   symbol name                  variants.type variants.name traitType traits.name</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> PIK3R2 phosphoinositide-3-k… single nucle… NM_005027.4(… Disease   Megalencep…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> PIK3R2 phosphoinositide-3-k… single nucle… NM_005027.4(… Disease   not provid…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> PIK3R2 phosphoinositide-3-k… single nucle… NM_005027.4(… Disease   not provid…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> UGT1A8 UDP glucuronosyltran… Microsatelli… UGT1A1*28     Disease   Gilbert's …</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="merging-with-collections">Merging MDBs with collections<a class="anchor" aria-label="anchor" href="#merging-with-collections"></a>
</h3>
<p>Until now, we have seen how to use individual MDB by exploring
general information about it, extracting tables, filtering and joining
data. This part shows how to use <strong>collections</strong> to
identify relationships between MDBs and to leverage these relationships
to integrate them. Documenting collection members has been <a href="#min-coll-memb">described above</a> and further information about
collections implementation is provided in the <a href="#collections">appendix</a>.</p>
<div class="section level4">
<h4 id="collections-and-collection-members">Collections and collection members<a class="anchor" aria-label="anchor" href="#collections-and-collection-members"></a>
</h4>
<p>As explained <a href="#min-coll-memb">above</a>, some databases refer
to the same concepts and could be integrated accordingly. However they
often use different vocabularies.</p>
<p>For example, both <a href="#chembl">CHEMBL</a> and <a href="#clinvar">ClinVar</a> refer to biological entities (BE) for
documenting drug targets or disease causal genes. CHEMBL refers to drug
target in the <em>CHEMBL_component_sequence</em> table using mainly
Uniprot peptide identifiers from different species.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_chembl</span><span class="op">$</span><span class="va">CHEMBL_component_sequence</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 35 × 5</span></span></span>
<span><span class="co">##    component_id accession organism                  db_source db_version</span></span>
<span><span class="co">##           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>          259 P15260    Homo sapiens              Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>          327 Q99062    Homo sapiens              Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>          752 P35563    Rattus norvegicus         Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>          917 P07339    Homo sapiens              Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>         <span style="text-decoration: underline;">1</span>807 Q54A96    Plasmodium falciparum     Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>         <span style="text-decoration: underline;">2</span>180 P67774    Bos taurus                Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>         <span style="text-decoration: underline;">2</span>398 P25098    Homo sapiens              Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>         <span style="text-decoration: underline;">2</span>541 Q8II92    Plasmodium falciparum 3D7 Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>         <span style="text-decoration: underline;">3</span>803 Q64346    Rattus norvegicus         Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>         <span style="text-decoration: underline;">4</span>395 O60502    Homo sapiens              Uniprot   2019_09   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 25 more rows</span></span></span></code></pre>
<p>Whereas ClinVar refers to causal genes in the <em>entrezNames</em>
table using human Entrez gene identifiers.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_clinvar</span><span class="op">$</span><span class="va">entrezNames</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 20 × 3</span></span></span>
<span><span class="co">##       entrez name                                                         symbol</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>      <span style="text-decoration: underline;">1</span>509 cathepsin D                                                  CTSD  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>      <span style="text-decoration: underline;">1</span>903 sphingosine-1-phosphate receptor 3                           S1PR3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>      <span style="text-decoration: underline;">3</span>300 DnaJ heat shock protein family (Hsp40) member B2             DNAJB2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>      <span style="text-decoration: underline;">3</span>423 iduronate 2-sulfatase                                        IDS   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>      <span style="text-decoration: underline;">3</span>910 laminin subunit alpha 4                                      LAMA4 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>      <span style="text-decoration: underline;">5</span>296 phosphoinositide-3-kinase regulatory subunit 2               PIK3R2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>      <span style="text-decoration: underline;">6</span>748 signal sequence receptor subunit 4                           SSR4  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>      <span style="text-decoration: underline;">7</span>633 zinc finger protein 79                                       ZNF79 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>     <span style="text-decoration: underline;">22</span>906 trafficking kinesin protein 1                                TRAK1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>     <span style="text-decoration: underline;">23</span>155 chloride channel CLIC like 1                                 CLCC1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span>     <span style="text-decoration: underline;">26</span>251 potassium voltage-gated channel modifier subfamily G member… KCNG2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span>     <span style="text-decoration: underline;">29</span>851 inducible T cell costimulator                                ICOS  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span>     <span style="text-decoration: underline;">54</span>576 UDP glucuronosyltransferase family 1 member A8               UGT1A8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">14</span>     <span style="text-decoration: underline;">57</span>684 zinc finger and BTB domain containing 26                     ZBTB26</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">15</span>    <span style="text-decoration: underline;">115</span>948 outer dynein arm docking complex subunit 3                   ODAD3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">16</span>    <span style="text-decoration: underline;">139</span>716 GRB2 associated binding protein 3                            GAB3  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">17</span>    <span style="text-decoration: underline;">169</span>792 GLIS family zinc finger 3                                    GLIS3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">18</span>    <span style="text-decoration: underline;">407</span>054 microRNA 98                                                  MIR98 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">19</span>    <span style="text-decoration: underline;">441</span>531 phosphoglycerate mutase family member 4                      PGAM4 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">20</span> 105<span style="text-decoration: underline;">373</span>557 serous ovarian cancer associated RNA                         SOCAR</span></span></code></pre>
<p>Since peptides are coded by genes, there is a biological relationship
between these two types of BE, and several tools exist to convert such
BE identifiers from one scope to the other (e.g. <a href="https://github.com/patzaw/BED" class="external-link">BED</a> <span class="citation">(<a href="#ref-godard_bed:_2018">Godard and Eyll 2018</a>)</span>, <a href="https://bioconductor.org/packages/release/bioc/html/mygene.html" class="external-link">mygene</a>
<span class="citation">(<a href="#ref-wu2012">Wu, MacLeod, and Su
2012</a>)</span>, <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html" class="external-link">biomaRt</a>
<span class="citation">(<a href="#ref-kinsella2011">Kinsella et al.
2011</a>)</span>).</p>
<p>TKCat provides mechanism to document these scopes in order to allow
automatic conversions from and to any of them. Those concepts are called
<strong>Collections</strong> in TKCat and they should be formally
defined before being able to document any of their members. Two
collection definitions are provided within the TKCat package and other
can be imported with the <code><a href="../reference/import_local_collection.html">import_local_collection()</a></code>
function.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/list_local_collections.html">list_local_collections</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   title     description                                  </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> BE        Collection of biological entity (BE) concepts</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition Collection of condition concepts</span></span></code></pre>
<p>Here are the definition of the BE collection members provided by the
<em>CHEMBL_component_sequence</em> and the <em>entrezNames</em>
tables.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">file_chembl</span>, <span class="st">"BE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 9</span></span></span>
<span><span class="co">##   collection cid           resource   mid table         field static value type </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_compo… be    TRUE   Pept… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_compo… iden… FALSE  acce… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_compo… sour… FALSE  db_s… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> BE         CHEMBL_BE_1.0 CHEMBL       1 CHEMBL_compo… orga… FALSE  orga… Scie…</span></span></code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">file_clinvar</span>, <span class="st">"BE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 9</span></span></span>
<span><span class="co">##   collection cid            resource   mid table       field  static value type </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> BE         ClinVar_BE_1.0 ClinVar      1 entrezNames be     TRUE   Gene  <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> BE         ClinVar_BE_1.0 ClinVar      1 entrezNames ident… FALSE  entr… <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> BE         ClinVar_BE_1.0 ClinVar      1 entrezNames organ… TRUE   Homo… Scie…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> BE         ClinVar_BE_1.0 ClinVar      1 entrezNames source TRUE   Entr… <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>The <em>Collection</em> column indicates the collection to which the
table refers. The <em>cid</em> column indicates the version of the
collection definition which should correspond to the <code>$id</code> of
JSON schema. The <em>resource</em> column indicates the name of the
resource and the <em>mid</em> column an identifier which is unique for
each member of a collection in each resource. The <em>field</em> column
indicates each part of the scope of collection. In the case of BE, 4
fields should be documented:</p>
<ul>
<li>be: the type of BE (e.g. Gene or Peptide)</li>
<li>source: the source of the identifier (e.g. EntrezGene or
Peptide)</li>
<li>organism: the organism to which the identifier refers (e.g Homo
sapiens)</li>
<li>identifier: the identifier itself.</li>
</ul>
<p>Each of these fields can be <em>static</em> or not. <code>TRUE</code>
means that the value of this field is the same for all the records and
is provided in the <em>value</em> column. Whereas <code>FALSE</code>
means that the value can be different for each record and is provided in
the column the name of which is given in the <em>value</em> column. The
<em>type</em> column is only used for the organism field in the case of
the BE collection and can take 2 values: “Scientific name” or “NCBI
taxon identifier”. The definition of the pre-build BE collection members
follows the terminology used in the <a href="https://github.com/patzaw/BED" class="external-link">BED package</a> <span class="citation">(<a href="#ref-godard_bed:_2018">Godard and Eyll
2018</a>)</span>. But it can be adapted according to the solution chosen
for converting BE identifiers from one scope to another.</p>
<p>Setting up the definition of such scope is done using the
<code><a href="../reference/add_collection_member.html">add_collection_member()</a></code> function as shown above in the <a href="#min-coll-memb">minimal example</a> and in the <a href="#hpo">Reading HPO example</a>.</p>
</div>
<div class="section level4">
<h4 id="shared-collections-and-merging">Shared collections and merging<a class="anchor" aria-label="anchor" href="#shared-collections-and-merging"></a>
</h4>
<p>The aim of collections is to identify potential bridges between MDBs.
The <code>get_shared_collection()</code> function is used to list all
the collections shared by two MDBs.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_shared_collections.html">get_shared_collections</a></span><span class="op">(</span><span class="va">filtered_clinvar</span>, <span class="va">file_chembl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 5</span></span></span>
<span><span class="co">##   collection table.x     mid.x table.y                   mid.y</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Condition  traits          2 CHEMBL_drug_indication        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition  traitCref       1 CHEMBL_drug_indication        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> BE         entrezNames     1 CHEMBL_component_sequence     1</span></span></code></pre>
<p>In this example, there are 3 different ways to merge the two MDBs
<em>filtered_clinvar</em> and <em>file_chembl</em>:</p>
<ul>
<li>Based on conditions provided respectively in the <em>traits</em> and
in the <em>CHEMBL_drug_indication</em> tables</li>
<li>Based on conditions provided respectively in the <em>traitsCref</em>
and in the <em>CHEMBL_drug_indication</em> tables</li>
<li>Based on BE provided respectively in the <em>entrezNames</em> and in
the <em>CHEMBL_component_sequence</em> tables</li>
</ul>
<p>The code below shows how to merge these two resources based on BE
information. To achieve this task it relies on a function provided with
TKCat along with BE collection definition (to get the function:
<code>get_collection_mapper("BE")</code>). This function uses the <a href="https://github.com/patzaw/BED" class="external-link">BED package</a> <span class="citation">(<a href="#ref-godard_bed:_2018">Godard and Eyll
2018</a>)</span> and you need this package to be installed with a
connection to BED database in order to run the code below.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/connectToBed.html" class="external-link">connectToBed</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in eval(expr, envir, enclos) : object 'a' not found</span></span></code></pre>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bedCheck</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/checkBedConn.html" class="external-link">checkBedConn</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">inherits</a></span><span class="op">(</span><span class="va">bedCheck</span>, <span class="st">"try-error"</span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="va">bedCheck</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">sel_coll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_shared_collections.html">get_shared_collections</a></span><span class="op">(</span><span class="va">file_clinvar</span>, <span class="va">file_chembl</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">collection</span><span class="op">==</span><span class="st">"BE"</span><span class="op">)</span></span>
<span>   <span class="va">filtered_cv_chembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span></span>
<span>      x<span class="op">=</span><span class="va">file_clinvar</span>,</span>
<span>      y<span class="op">=</span><span class="va">file_chembl</span>,</span>
<span>      by<span class="op">=</span><span class="va">sel_coll</span>,</span>
<span>      dmAutoLayout<span class="op">=</span><span class="va">igraph_available</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The returned object is a <strong>metaMDB</strong> gathering the
original MDBs and a relational table between members of the same
collection as defined by the <code>by</code> parameter.</p>
<p>Additional information about collection can be found below in the <a href="#collections">appendix</a>.</p>
</div>
<div class="section level4">
<h4 id="merging-without-collection">Merging without collection<a class="anchor" aria-label="anchor" href="#merging-without-collection"></a>
</h4>
<p>If the <em>collection</em> column of the <code>by</code> parameter is
<code>NA</code>, then the relational table is built by merging identical
columns in table.x and table.y (No conversion occurs). For example,
<em>file_hpo</em> and <em>file_clinvar</em> MDBs could be merged
according to conditions provided in the <em>HPO_diseases</em> and the
<em>traitCref</em> tables respectively.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_shared_collections.html">get_shared_collections</a></span><span class="op">(</span><span class="va">file_hpo</span>, <span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 5</span></span></span>
<span><span class="co">##   collection table.x      mid.x table.y   mid.y</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Condition  HPO_hp           1 traits        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition  HPO_hp           1 traitCref     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Condition  HPO_diseases     2 traits        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Condition  HPO_diseases     2 traitCref     1</span></span></code></pre>
<p>These conditions could be converted using a function provided with
TKCat (<code>get_collection_mapper("Condition")</code>) and which rely
on the <a href="https://github.com/Elysheba/DODO" class="external-link">DODO package</a> <span class="citation">(<a href="#ref-fran%C3%A7ois2020">François, Eyll, and Godard
2020</a>)</span>. The two tables can also be simply concatenated without
applying any conversion (loosing the advantage of such conversion
obviously).</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_coll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_shared_collections.html">get_shared_collections</a></span><span class="op">(</span><span class="va">file_hpo</span>, <span class="va">file_clinvar</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">table.x</span><span class="op">==</span><span class="st">"HPO_diseases"</span>, <span class="va">table.y</span><span class="op">==</span><span class="st">"traitCref"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>   <span class="fu">mutate</span><span class="op">(</span>collection<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">sel_coll</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 5</span></span></span>
<span><span class="co">##   collection table.x      mid.x table.y   mid.y</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> <span style="color: #BB0000;">NA</span>         HPO_diseases     2 traitCref     1</span></span></code></pre>
<p>The <code><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge()</a></code> function gather the two <em>MDBs</em> in one
<em>metaMDB</em> and create a association table based on the
<code>by</code> argument. This association table
(“HPO_diseases_traitCref”) is displayed in yellow in the data model of
the created <em>metaMDB</em> as shown below.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hpo_clinvar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span></span>
<span>   <span class="va">file_hpo</span>, <span class="va">file_clinvar</span>, by<span class="op">=</span><span class="va">sel_coll</span>, dmAutoLayout<span class="op">=</span><span class="va">igraph_available</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="va">hpo_clinvar</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-4018eef1a407a0df6b52" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-4018eef1a407a0df6b52">{"x":{"nodes":{"tableName":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms","ReferenceClinVarAssertion","rcvaVariant","ClinVarAssertions","rcvaInhMode","rcvaObservedIn","rcvaTraits","clinSigOrder","revStatOrder","variants","cvaObservedIn","cvaSubmitters","traits","varEntrez","varAttributes","varCytoLoc","varNames","varSeqLoc","varXRef","traitCref","traitNames","entrezNames","HPO_diseases_traitCref"],"label":["<b>HPO_hp<\/b>\n    - <b>*id<\/b> {character - uidx.1}\n    - name {character}\n    - (description {character})\n    - level {integer}","<b>HPO_altId<\/b>\n    - id {character - idx.1}\n    - alt {character}","<b>HPO_sourceFiles<\/b>\n    - <b>*url<\/b> {character - uidx.1}\n    - current {Date}","<b>HPO_diseases<\/b>\n    - <b>db<\/b> {character - uidx.1}\n    - <b>id<\/b> {character - uidx.1}\n    - label {character}","<b>HPO_diseaseHP<\/b>\n    - db {character - idx.2}\n    - id {character - idx.2}\n    - hp {character - idx.1}","<b>HPO_diseaseSynonyms<\/b>\n    - db {character - idx.1}\n    - id {character - idx.1}\n    - synonym {character}\n    - preferred {logical}","<b>HPO_parents<\/b>\n    - id {character - idx.1}\n    - parent {character - idx.2}","<b>HPO_descendants<\/b>\n    - id {character - idx.1}\n    - descendant {character - idx.2}","<b>HPO_synonyms<\/b>\n    - id {character}\n    - synonym {character}\n    - type {character}","<b>ReferenceClinVarAssertion<\/b>\n    - *cvs {integer - uidx.3}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - assertion {character}\n    - reviewStatus {character}\n    - clinicalSignificance {character - idx.1}\n    - (explanation {character})\n    - title {character}","<b>rcvaVariant<\/b>\n    - varId {integer}\n    - rcvaId {integer - idx.1}","<b>ClinVarAssertions<\/b>\n    - cvs {integer - idx.1}\n    - <b>*id<\/b> {integer - uidx.2}\n    - accession {character}\n    - (clinicalSignificance {character})","<b>rcvaInhMode<\/b>\n    - rcvaId {integer}\n    - inhMode {character}","<b>rcvaObservedIn<\/b>\n    - rcvaId {integer}\n    - origin {character}\n    - taxonomyId {character}\n    - species {character}\n    - affectedStatus {character}\n    - (numberTested {integer})","<b>rcvaTraits<\/b>\n    - rcvaId {integer}\n    - t.id {integer - idx.1}\n    - traitType {character}","<b>clinSigOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>revStatOrder<\/b>\n    - <b>*label<\/b> {character - uidx.1}\n    - order {integer}","<b>variants<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - type {character}\n    - (name {character})","<b>cvaObservedIn<\/b>\n    - cvaId {integer}\n    - origin {character}\n    - species {character}\n    - affectedStatus {character}","<b>cvaSubmitters<\/b>\n    - cvaId {integer}\n    - submitter {character}\n    - primary {logical}","<b>traits<\/b>\n    - <b>*id<\/b> {integer - uidx.1}\n    - name {character}","<b>varEntrez<\/b>\n    - varId {integer - idx.1}\n    - entrez {integer}\n    - type {character}","<b>varAttributes<\/b>\n    - varId {integer - idx.1}\n    - Type {character}\n    - (integerValue {integer})\n    - (Change {character})\n    - (value {character})","<b>varCytoLoc<\/b>\n    - varId {integer - idx.1}\n    - location {character}","<b>varNames<\/b>\n    - varId {integer - idx.1}\n    - name {character}\n    - type {character}","<b>varSeqLoc<\/b>\n    - varId {integer - idx.1}\n    - Accession {character}\n    - (alternateAllele {character})\n    - Assembly {character}\n    - AssemblyAccessionVersion {character}\n    - AssemblyStatus {character}\n    - Chr {character}\n    - display_start {integer}\n    - (display_stop {integer})\n    - (innerStart {integer})\n    - (innerStop {integer})\n    - (outerStart {integer})\n    - (outerStop {integer})\n    - (referenceAllele {character})\n    - (start {integer})\n    - (stop {integer})\n    - (Strand {character})\n    - (variantLength {integer})","<b>varXRef<\/b>\n    - varId {integer - idx.1}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>traitCref<\/b>\n    - t.id {integer}\n    - id {character}\n    - db {character}\n    - (type {character})","<b>traitNames<\/b>\n    - t.id {integer}\n    - name {character}\n    - type {character}","<b>entrezNames<\/b>\n    - <b>*entrez<\/b> {integer - uidx.1}\n    - name {character}\n    - symbol {character}","<b>HPO_diseases_traitCref<\/b>\n    - db {character}\n    - id {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_hp<\/strong> (Human phenotypes (HP))<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>name<\/strong>: HP name<\/li> <li><strong>description<\/strong>: HP description<\/li> <li><strong>level<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_altId<\/strong> (Alternative HP identifiers)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>alt<\/strong>: Alternative identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_sourceFiles<\/strong> (Files from which data were extracted)<\/p> <ul> <li><strong>url<\/strong>: Source file URL<\/li> <li><strong>current<\/strong>: The date of the current source file<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases<\/strong> (Diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>label<\/strong>: Disease lable (preferred synonym)<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseHP<\/strong> (HP presented by diseases)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>hp<\/strong>: HP identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseaseSynonyms<\/strong> (Disease synonyms)<\/p> <ul> <li><strong>db<\/strong>: Disease database<\/li> <li><strong>id<\/strong>: Disease ID<\/li> <li><strong>synonym<\/strong>: Disease synonym<\/li> <li><strong>preferred<\/strong>: Is synonym preferred<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_parents<\/strong> (Direct parents of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>parent<\/strong>: Parent identifier<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_descendants<\/strong> (Indirect descendants of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>descendant<\/strong>: Descendant (child or child of child...) HP ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_synonyms<\/strong> (Synonyms of human phenotypes)<\/p> <ul> <li><strong>id<\/strong>: HP identifier<\/li> <li><strong>synonym<\/strong>: HP synonym<\/li> <li><strong>type<\/strong>: Synonym type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ReferenceClinVarAssertion<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>assertion<\/strong><\/li> <li><strong>reviewStatus<\/strong><\/li> <li><strong>clinicalSignificance<\/strong><\/li> <li><strong>explanation<\/strong>: Explanation is used only when the description is 'conflicting data from submitters'. The element summarizes the conflict.<\/li> <li><strong>title<\/strong>: The title is contructed from the concatenation of the description of the variation and the phenotype. In the database it is extracted from the title of RCV accession<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">rcvaVariant<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ClinVarAssertions<\/strong><\/p> <ul> <li><strong>cvs<\/strong>: ClinVar Set ID<\/li> <li><strong>id<\/strong>: ClinVar Assertion ID<\/li> <li><strong>accession<\/strong>: The accession assigned by ClinVar.<\/li> <li><strong>clinicalSignificance<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">rcvaInhMode<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>inhMode<\/strong>: Mode of inheritance<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">rcvaObservedIn<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>taxonomyId<\/strong>: Taxonomy ID<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <li><strong>numberTested<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">rcvaTraits<\/strong><\/p> <ul> <li><strong>rcvaId<\/strong>: Reference ClinVar Assertion ID<\/li> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>traitType<\/strong>: Trait type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">clinSigOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of clinical significance: the lowest the best clinical outcome.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">revStatOrder<\/strong><\/p> <ul> <li><strong>label<\/strong><\/li> <li><strong>order<\/strong>: An integer indicating the order of review status: the lowest the least significant.<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">variants<\/strong><\/p> <ul> <li><strong>id<\/strong>: Variant ID<\/li> <li><strong>type<\/strong>: Variant type<\/li> <li><strong>name<\/strong>: Variant preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">cvaObservedIn<\/strong><\/p> <ul> <li><strong>cvaId<\/strong><\/li> <li><strong>origin<\/strong>: Sample origin<\/li> <li><strong>species<\/strong>: Species name<\/li> <li><strong>affectedStatus<\/strong>: Affected status<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">cvaSubmitters<\/strong><\/p> <ul> <li><strong>cvaId<\/strong>: ClinVar Assertion ID<\/li> <li><strong>submitter<\/strong>: Submitter name<\/li> <li><strong>primary<\/strong>: The submitter is of pirmary type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">traits<\/strong><\/p> <ul> <li><strong>id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait preferred or first alternate name<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varEntrez<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>type<\/strong>: Type of the association between the variant and the gene<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varAttributes<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Type<\/strong><\/li> <li><strong>integerValue<\/strong><\/li> <li><strong>Change<\/strong><\/li> <li><strong>value<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varCytoLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>location<\/strong>: Cytogenic location<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varNames<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>name<\/strong>: Variant name<\/li> <li><strong>type<\/strong>: Variant name type: Preferred or Alternate<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varSeqLoc<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>Accession<\/strong>: Sequence accession ID<\/li> <li><strong>alternateAllele<\/strong>: Alternate allele<\/li> <li><strong>Assembly<\/strong>: Assembly<\/li> <li><strong>AssemblyAccessionVersion<\/strong>: Version of assembly accession<\/li> <li><strong>AssemblyStatus<\/strong>: Status of the assembly: current or previous<\/li> <li><strong>Chr<\/strong>: Chromosome name<\/li> <li><strong>display_start<\/strong><\/li> <li><strong>display_stop<\/strong><\/li> <li><strong>innerStart<\/strong><\/li> <li><strong>innerStop<\/strong><\/li> <li><strong>outerStart<\/strong><\/li> <li><strong>outerStop<\/strong><\/li> <li><strong>referenceAllele<\/strong>: Reference allele<\/li> <li><strong>start<\/strong><\/li> <li><strong>stop<\/strong><\/li> <li><strong>Strand<\/strong><\/li> <li><strong>variantLength<\/strong>: Length of the variant<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">varXRef<\/strong><\/p> <ul> <li><strong>varId<\/strong>: Variant ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">traitCref<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>id<\/strong>: External ID<\/li> <li><strong>db<\/strong>: External DB<\/li> <li><strong>type<\/strong>: Reference type<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">traitNames<\/strong><\/p> <ul> <li><strong>t.id<\/strong>: ClinVar trait ID<\/li> <li><strong>name<\/strong>: Trait name<\/li> <li><strong>type<\/strong>: \"Alternate\" or \"Preferred\"<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">entrezNames<\/strong><\/p> <ul> <li><strong>entrez<\/strong>: Entrez Gene ID<\/li> <li><strong>name<\/strong>: Gene name<\/li> <li><strong>symbol<\/strong>: Gene symbol<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">HPO_diseases_traitCref<\/strong><\/p> <ul> <li><strong>db<\/strong><\/li> <li><strong>id<\/strong><\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left","left"],"x":[144.6292046914937,-313.9694981365806,654.8636394524915,947.7374173180356,596.8978395465931,1240,-88.46533927476584,410.8960784700604,14.61661650338591,-145.4533001069884,-468.3758410351223,-703.7503716956736,242.9669033069716,-472.7267861882931,420.2718725318353,-89.91393424381576,41.3419319800351,-808.6705594575176,-929.2551770976681,-1130.849262571149,859.6340269019053,-338.9413004121193,-1240,-1196.429819028398,-648.8261648926989,-969.2139131231592,-995.201088996943,1154.435973946119,1032.183177675014,-24.94466193738693,1145.135158533224],"y":[-942.233684802988,-1053.710071720524,1240,-399.3772912173195,-688.1916261849572,-479.2172577720441,-818.6455173722566,-1042.443995762367,-1240,116.2973603775893,459.8085539837009,-270.4316153986202,1.103469519306701,-11.35993112080448,369.1225241309376,-233.3524389155673,460.6709960332177,747.0557165440247,-603.8120643631456,-338.2382061016107,526.5047947190593,1023.47262618829,513.5447555432276,798.6587313658953,1091.212107306484,1064.033954887046,372.9681946856942,300.8378075417079,805.1036736945532,1238.236074794775,-41.5732800169553],"color.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0","yellow"],"color.border":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#9BC8FE","#9BC8FE","#D9D9D9","#DFFB86","#D9D9D9","#DFFB86","#F67FC4","#F67FC4","#9BC8FE","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#DFFB86","#D9D9D9","#DFFB86","#DFFB86","#F8DEC3","#DFFB86","#DFFB86","#9BC8FE","#8FE6E0","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#F8DEC3","#9BC8FE","#9BC8FE","#8FE6E0","yellow"],"id":["HPO_hp","HPO_altId","HPO_sourceFiles","HPO_diseases","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_descendants","HPO_synonyms","ReferenceClinVarAssertion","rcvaVariant","ClinVarAssertions","rcvaInhMode","rcvaObservedIn","rcvaTraits","clinSigOrder","revStatOrder","variants","cvaObservedIn","cvaSubmitters","traits","varEntrez","varAttributes","varCytoLoc","varNames","varSeqLoc","varXRef","traitCref","traitNames","entrezNames","HPO_diseases_traitCref"]},"edges":{"id":["HPO_altId->HPO_hp: id->id","HPO_diseaseHP->HPO_hp: hp->id","HPO_diseaseHP->HPO_diseases: db->db && id->id","HPO_diseaseSynonyms->HPO_diseases: db->db && id->id","HPO_parents->HPO_hp: id->id","HPO_parents->HPO_hp: parent->id","HPO_descendants->HPO_hp: id->id","HPO_descendants->HPO_hp: descendant->id","HPO_synonyms->HPO_hp: id->id","ReferenceClinVarAssertion->clinSigOrder: clinicalSignificance->label","ReferenceClinVarAssertion->revStatOrder: reviewStatus->label","rcvaVariant->ReferenceClinVarAssertion: rcvaId->id","rcvaVariant->variants: varId->id","ClinVarAssertions->ReferenceClinVarAssertion: cvs->cvs","rcvaInhMode->ReferenceClinVarAssertion: rcvaId->id","rcvaObservedIn->ReferenceClinVarAssertion: rcvaId->id","rcvaTraits->traits: t.id->id","rcvaTraits->ReferenceClinVarAssertion: rcvaId->id","cvaObservedIn->ClinVarAssertions: cvaId->id","cvaSubmitters->ClinVarAssertions: cvaId->id","varEntrez->variants: varId->id","varEntrez->entrezNames: entrez->entrez","varAttributes->variants: varId->id","varCytoLoc->variants: varId->id","varNames->variants: varId->id","varSeqLoc->variants: varId->id","varXRef->variants: varId->id","traitCref->traits: t.id->id","traitNames->traits: t.id->id","HPO_diseases_traitCref->HPO_diseases: db->db && id->id","HPO_diseases_traitCref->traitCref: db->db && id->id"],"to":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseases","HPO_hp","HPO_hp","HPO_hp","HPO_hp","HPO_hp","clinSigOrder","revStatOrder","ReferenceClinVarAssertion","variants","ReferenceClinVarAssertion","ReferenceClinVarAssertion","ReferenceClinVarAssertion","traits","ReferenceClinVarAssertion","ClinVarAssertions","ClinVarAssertions","variants","entrezNames","variants","variants","variants","variants","variants","traits","traits","HPO_diseases","traitCref"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_altId<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">hp<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseHP<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseaseSynonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_parents<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">parent<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_descendants<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">descendant<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_synonyms<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_hp<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">clinSigOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">clinicalSignificance<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">revStatOrder<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">reviewStatus<\/td><td style=\"border: 1px solid black; padding:1px;\">label<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaVariant<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVarAssertions<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><td style=\"border: 1px solid black; padding:1px;\">cvs<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaInhMode<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">rcvaTraits<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ReferenceClinVarAssertion<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">rcvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">cvaObservedIn<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">cvaSubmitters<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">ClinVarAssertions<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">cvaId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varEntrez<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">entrezNames<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><td style=\"border: 1px solid black; padding:1px;\">entrez<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varAttributes<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varCytoLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varSeqLoc<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">varXRef<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">variants<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">varId<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traitNames<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traits<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">t.id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases<br>(0..n)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">HPO_diseases_traitCref<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">traitCref<br>(0..n)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">db<\/td><td style=\"border: 1px solid black; padding:1px;\">db<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">id<\/td><td style=\"border: 1px solid black; padding:1px;\">id<\/td><\/tr><\/table>"],"ff":["id","hp",["db","id"],["db","id"],"id","parent","id","descendant","id","clinicalSignificance","reviewStatus","rcvaId","varId","cvs","rcvaId","rcvaId","t.id","rcvaId","cvaId","cvaId","varId","entrez","varId","varId","varId","varId","varId","t.id","t.id",["db","id"],["db","id"]],"tf":["id","id",["db","id"],["db","id"],"id","id","id","id","id","label","label","id","id","cvs","id","id","id","id","id","id","id","entrez","id","id","id","id","id","id","id",["db","id"],["db","id"]],"from":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_descendants","HPO_descendants","HPO_synonyms","ReferenceClinVarAssertion","ReferenceClinVarAssertion","rcvaVariant","rcvaVariant","ClinVarAssertions","rcvaInhMode","rcvaObservedIn","rcvaTraits","rcvaTraits","cvaObservedIn","cvaSubmitters","varEntrez","varEntrez","varAttributes","varCytoLoc","varNames","varSeqLoc","varXRef","traitCref","traitNames","HPO_diseases_traitCref","HPO_diseases_traitCref"],"arrows":["to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0,0.2,-0.2,-0.2,0.2,0,0,-0,-0,-0,-0,-0,-0,-0,-0,0,0,-0,0,-0,-0,0,0,0,-0,-0,0,-0],"selfReferenceSize":[30,30,30,30,30,50,30,50,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30,30],"uef":["HPO_altId","HPO_diseaseHP","HPO_diseaseHP","HPO_diseases","HPO_hp","HPO_hp","HPO_descendants","HPO_descendants","HPO_hp","clinSigOrder","ReferenceClinVarAssertion","rcvaVariant","rcvaVariant","ClinVarAssertions","rcvaInhMode","rcvaObservedIn","rcvaTraits","rcvaTraits","ClinVarAssertions","ClinVarAssertions","varEntrez","entrezNames","varAttributes","varCytoLoc","variants","variants","variants","traitCref","traitNames","HPO_diseases","HPO_diseases_traitCref"],"uet":["HPO_hp","HPO_hp","HPO_diseases","HPO_diseaseSynonyms","HPO_parents","HPO_parents","HPO_hp","HPO_hp","HPO_synonyms","ReferenceClinVarAssertion","revStatOrder","ReferenceClinVarAssertion","variants","ReferenceClinVarAssertion","ReferenceClinVarAssertion","ReferenceClinVarAssertion","traits","ReferenceClinVarAssertion","cvaObservedIn","cvaSubmitters","variants","varEntrez","variants","variants","varNames","varSeqLoc","varXRef","traits","traits","HPO_diseases_traitCref","traitCref"],"ue":["HPO_altId<->HPO_hp","HPO_diseaseHP<->HPO_hp","HPO_diseaseHP<->HPO_diseases","HPO_diseases<->HPO_diseaseSynonyms","HPO_hp<->HPO_parents","HPO_hp<->HPO_parents","HPO_descendants<->HPO_hp","HPO_descendants<->HPO_hp","HPO_hp<->HPO_synonyms","clinSigOrder<->ReferenceClinVarAssertion","ReferenceClinVarAssertion<->revStatOrder","rcvaVariant<->ReferenceClinVarAssertion","rcvaVariant<->variants","ClinVarAssertions<->ReferenceClinVarAssertion","rcvaInhMode<->ReferenceClinVarAssertion","rcvaObservedIn<->ReferenceClinVarAssertion","rcvaTraits<->traits","rcvaTraits<->ReferenceClinVarAssertion","ClinVarAssertions<->cvaObservedIn","ClinVarAssertions<->cvaSubmitters","varEntrez<->variants","entrezNames<->varEntrez","varAttributes<->variants","varCytoLoc<->variants","variants<->varNames","variants<->varSeqLoc","variants<->varXRef","traitCref<->traits","traitNames<->traits","HPO_diseases<->HPO_diseases_traitCref","HPO_diseases_traitCref<->traitCref"],"color.color":["black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hpo_clinvar</span><span class="op">$</span><span class="va">HPO_diseases_traitCref</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,950 × 2</span></span></span>
<span><span class="co">##    db       id    </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> DECIPHER 15    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> DECIPHER 45    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> DECIPHER 65    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> OMIM     100050</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> OMIM     100650</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> OMIM     101800</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> OMIM     102500</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> OMIM     102510</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> OMIM     102700</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> OMIM     102800</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,940 more rows</span></span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="chTKCat">A centralized catalog of MDB in ClickHouse (chTKCat)<a class="anchor" aria-label="anchor" href="#chTKCat"></a>
</h2>
<div class="section level3">
<h3 id="local-tkcat">Local TKCat<a class="anchor" aria-label="anchor" href="#local-tkcat"></a>
</h3>
<p>MDB can be gathered in a <em>TKCat</em> (Tailored Knowledge Catalog)
object.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TKCat.html">TKCat</a></span><span class="op">(</span><span class="va">file_hpo</span>, <span class="va">file_clinvar</span><span class="op">)</span></span></code></pre></div>
<p>Gathering MDBs in such a catalog facilitate their exploration and
their preparation for potential integration. Several functions are
available to achieve this goal.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/list_MDBs.html">list_MDBs</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span>                     <span class="co"># list all the MDBs in a TKCat object</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">##   name    title         description url   version maintainer timestamp</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HPO     Data extract… This is a … http… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>         <span style="color: #BB0000;">NA</span>       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ClinVar Data extract… ClinVar is… http… 0.9     Patrice G… <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_MDB.html">get_MDB</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"HPO"</span><span class="op">)</span>                <span class="co"># get a specific MDBs from the catalog</span></span></code></pre></div>
<pre><code><span><span class="co">## fileMDB HPO: Data extracted from the HPO database</span></span>
<span><span class="co">##    - 9 tables with 25 fields</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Collection members: </span></span>
<span><span class="co">##    - 2 Condition members</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This is a very small subset of the HPO! Visit the reference URL for more information.</span></span>
<span><span class="co">## (http://human-phenotype-ontology.github.io/)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/search_MDB_tables.html">search_MDB_tables</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"disease"</span><span class="op">)</span>  <span class="co"># Search table about "disease"</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">##   resource name                comment                 </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HPO      HPO_diseases        Diseases                </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> HPO      HPO_diseaseHP       HP presented by diseases</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> HPO      HPO_diseaseSynonyms Disease synonyms</span></span></code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/search_MDB_fields.html">search_MDB_fields</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"disease"</span><span class="op">)</span>  <span class="co"># Search a field about "disease"</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 7</span></span></span>
<span><span class="co">##   resource table               name    type      nullable unique comment        </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HPO      HPO_diseases        db      character FALSE    FALSE  Disease databa…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> HPO      HPO_diseases        id      character FALSE    FALSE  Disease ID     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> HPO      HPO_diseases        label   character FALSE    FALSE  Disease lable …</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> HPO      HPO_diseaseHP       db      character FALSE    FALSE  Disease databa…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> HPO      HPO_diseaseHP       id      character FALSE    FALSE  Disease ID     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> HPO      HPO_diseaseSynonyms db      character FALSE    FALSE  Disease databa…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> HPO      HPO_diseaseSynonyms id      character FALSE    FALSE  Disease ID     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> HPO      HPO_diseaseSynonyms synonym character FALSE    FALSE  Disease synonym</span></span></code></pre>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span>            <span class="co"># Get collection members of the different MDBs</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 5 × 3</span></span></span>
<span><span class="co">##   resource collection table       </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HPO      Condition  HPO_hp      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> HPO      Condition  HPO_diseases</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ClinVar  Condition  traits      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ClinVar  Condition  traitCref   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ClinVar  BE         entrezNames</span></span></code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">k</span>, <span class="fu"><a href="../reference/TKCat.html">TKCat</a></span><span class="op">(</span><span class="va">file_chembl</span><span class="op">)</span><span class="op">)</span>         <span class="co"># Merge 2 TKCat objects</span></span></code></pre></div>
<pre><code><span><span class="co">## TKCat gathering 3 MDB objects</span></span></code></pre>
<p>The function <code><a href="../reference/explore_MDBs.html">explore_MDBs()</a></code> launches a shiny interface
to explore MDBs in a <em>TKCat</em> object. This exploration interface
can be easily deployed using an <em>app.R</em> file with content similar
to the one below.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patzaw.github.io/TKCat/">TKCat</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/explore_MDBs.html">explore_MDBs</a></span><span class="op">(</span><span class="va">k</span>, download<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore_MDBs-local.png"></p>
<p>In this interface the users can explore the resources available in
the catalog. They can browse the data model of each of them with some
sample data. They can also search for information provided in resources,
tables or fields. Finally, if the parameter <code>download</code> is set
to <code>TRUE</code>, the users will also be able to download the data:
either each table individually or an archive of the whole MDB.</p>
</div>
<div class="section level3">
<h3 id="chtkcat">chTKCat<a class="anchor" aria-label="anchor" href="#chtkcat"></a>
</h3>
<p>A <em>chTKCat</em> object is a catalog of MDB as a <em>TKCat</em>
object described above but relying on a <a href="https://clickhouse.com/" class="external-link">ClickHouse</a> database. This part
focuses on using and querying a <em>chTKCat</em> object. The
installation and the initialization of a <a href="https://clickhouse.com/" class="external-link">ClickHouse</a> database ready for TKCat
are described below in the <a href="#ch-op">appendix</a>.</p>
<p>The connection to the ClickHouse TKCat database is achieved using the
<code><a href="../reference/chTKCat.html">chTKCat()</a></code> function.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chTKCat.html">chTKCat</a></span><span class="op">(</span></span>
<span>   host<span class="op">=</span><span class="st">"localhost"</span>,                     <span class="co"># default parameter</span></span>
<span>   port<span class="op">=</span><span class="fl">9111L</span>,                           <span class="co"># default parameter</span></span>
<span>   drv<span class="op">=</span><span class="fu">ClickHouseHTTP</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ClickHouseHTTP/man/ClickHouseHTTP.html" class="external-link">ClickHouseHTTP</a></span><span class="op">(</span><span class="op">)</span>, <span class="co"># default parameter</span></span>
<span>   user<span class="op">=</span><span class="st">"default"</span>,                       <span class="co"># default parameter</span></span>
<span>   password<span class="op">=</span><span class="st">""</span>                           <span class="co"># if not provided the</span></span>
<span>                                         <span class="co"># password is requested interactively </span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By default, this function connects anonymously (“default” user
without password) to the database, using the <a href="https://clickhouse.com/docs/en/interfaces/http/" class="external-link">HTTP
interface</a> of ClickHouse thanks to the <a href="https://github.com/patzaw/ClickHouseHTTP" class="external-link">ClickHouseHTTP</a>
driver. If the database is configured appropriately (see <a href="#ch-op">appendix</a>), connection can be achieved through HTTPS
with or without SSL peer verification (see the manual of
<code>ClickHouseHTTP::\</code>ClickHouseHTTPDriver-class`<code>for further information). Also, the</code>RClickhouse::clickhouse()<code>driver from the [RClickhouse][rclickhouse] package can be used (</code>drv<code>parameter of the</code>chTKCat()`
function) to leverage the native <a href="https://clickhouse.com/docs/en/interfaces/tcp/" class="external-link">TCP interface</a>
of ClickHouse which has the strong advantage of having less overhead.
But TLS wrapping is not supported yet by the RClickhouse package.</p>
<p>Once connected, this <em>chTKCat</em> object can be used as a
<em>TKCat</em> object.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/list_MDBs.html">list_MDBs</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span>             <span class="co"># get a specific MDBs from the catalog</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 12</span></span></span>
<span><span class="co">##   name    title description url   version maintainer public populated timestamps</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> CHEMBL  Data… CHEMBL is … http… 0.2     Liesbeth … TRUE   TRUE      TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ClinVar Data… ClinVar is… http… 0.9     Patrice G… TRUE   TRUE      TRUE      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> HPO     Data… This is a … http… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>         TRUE   TRUE      TRUE      </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: timestamp &lt;dttm&gt;, access &lt;fct&gt;, total_size &lt;dbl&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/search_MDB_tables.html">search_MDB_tables</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"disease"</span><span class="op">)</span>  <span class="co"># Search table about "disease"</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">##   resource name                comment                 </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> HPO      HPO_diseaseHP       HP presented by diseases</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> HPO      HPO_diseaseSynonyms Disease synonyms        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> HPO      HPO_diseases        Diseases</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/search_MDB_fields.html">search_MDB_fields</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"disease"</span><span class="op">)</span>  <span class="co"># Search a field about "disease"</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 12 × 7</span></span></span>
<span><span class="co">##    resource table                  name            comment type  nullable unique</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> CHEMBL   CHEMBL_drug_indication DB              <span style="color: #949494;">"</span>Datab… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CHEMBL   CHEMBL_drug_mechanism  disease_effica… <span style="color: #949494;">""</span>      char… TRUE     FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CHEMBL   CHEMBL_drug_indication id              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> CHEMBL   CHEMBL_drug_indication name            <span style="color: #949494;">"</span>Term … char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> HPO      HPO_diseaseHP          db              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> HPO      HPO_diseaseSynonyms    db              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> HPO      HPO_diseases           db              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> HPO      HPO_diseaseHP          id              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> HPO      HPO_diseaseSynonyms    id              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> HPO      HPO_diseases           id              <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> HPO      HPO_diseases           label           <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> HPO      HPO_diseaseSynonyms    synonym         <span style="color: #949494;">"</span>Disea… char… FALSE    FALSE</span></span></code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collection_members.html">collection_members</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span>  </span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 7 × 3</span></span></span>
<span><span class="co">##   resource collection table                    </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> CHEMBL   BE         CHEMBL_component_sequence</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> CHEMBL   Condition  CHEMBL_drug_indication   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ClinVar  BE         ClinVar_entrezNames      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ClinVar  Condition  ClinVar_traitCref        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ClinVar  Condition  ClinVar_traits           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> HPO      Condition  HPO_hp                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> HPO      Condition  HPO_diseases</span></span></code></pre>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/explore_MDBs.html">explore_MDBs</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pushing-an-mdb-in-a-chtkcat-instance">Pushing an MDB in a chTKCat instance<a class="anchor" aria-label="anchor" href="#pushing-an-mdb-in-a-chtkcat-instance"></a>
</h3>
<p>Any <code>MDB</code> object can be imported in a TKCat ClickHouse
instance as following:</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chTKCat.html">chTKCat</a></span><span class="op">(</span>host<span class="op">=</span><span class="st">"localhost"</span>, port<span class="op">=</span><span class="fl">9111L</span>, user<span class="op">=</span><span class="st">"pgodard"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/create_chMDB.html">create_chMDB</a></span><span class="op">(</span><span class="va">kw</span>, <span class="st">"HPO"</span>, public<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ch_hpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_chMDB.html">as_chMDB</a></span><span class="op">(</span><span class="va">file_hpo</span>, <span class="va">kw</span><span class="op">)</span></span></code></pre></div>
<p>It is then accessible to anyone with relevant permissions on the
Clickhouse database. Pushing data in a ClickHouse database works only if
the user is allowed to write in the database.</p>
</div>
<div class="section level3">
<h3 id="specific-operations-on-chmdb-objects">Specific operations on chMDB objects<a class="anchor" aria-label="anchor" href="#specific-operations-on-chmdb-objects"></a>
</h3>
<p>The function <code><a href="../reference/get_MDB.html">get_MDB()</a></code> returns a <em>chMDB</em> object
that can be used as any <em>MDB</em> object. The data are located in the
ClickHouse database and pulled on request.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ch_hpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_MDB.html">get_MDB</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"HPO"</span><span class="op">)</span></span></code></pre></div>
<p>To avoid pulling a whole table from ClickHouse (which can take time
if the table is big), SQL queries can be made on the <em>chMDB</em>
object as shown below.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_query.html">get_query</a></span><span class="op">(</span></span>
<span>   <span class="va">ch_hpo</span>,</span>
<span>   query<span class="op">=</span><span class="st">"SELECT * from HPO_diseases WHERE lower(label) LIKE '%epilep%'"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 16 × 3</span></span></span>
<span><span class="co">##    db    id     label                                                          </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> OMIM  266100 EPILEPSY, PYRIDOXINE-DEPENDENT                                 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> OMIM  601068 EPILEPSY, FAMILIAL ADULT MYOCLONIC, 1; FAME1                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> OMIM  607876 EPILEPSY, FAMILIAL ADULT MYOCLONIC, 2; FAME2                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> OMIM  610003 CEROID LIPOFUSCINOSIS, NEURONAL, 8, NORTHERN EPILEPSY VARIANT  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> OMIM  612164 EPILEPTIC ENCEPHALOPATHY, EARLY INFANTILE, 4                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> OMIM  613608 EPILEPSY, FAMILIAL ADULT MYOCLONIC, 3; FAME3                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> OMIM  615127 EPILEPSY, FAMILIAL ADULT MYOCLONIC, 4; FAME4                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> OMIM  616457 EPILEPTIC ENCEPHALOPATHY, EARLY INFANTILE, 50; EIEE50          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ORPHA 1192   Atherosclerosis-deafness-diabetes-epilepsy-nephropathy syndrome</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ORPHA 1934   Early infantile epileptic encephalopathy                       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> ORPHA 1945   Rolandic epilepsy                                              </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> ORPHA 263516 Progressive myoclonic epilepsy type 3                          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span> ORPHA 352582 Familial infantile myoclonic epilepsy                          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">14</span> ORPHA 352596 Progressive myoclonic epilepsy with dystonia                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">15</span> ORPHA 439218 KCNQ2-related epileptic encephalopathy                         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">16</span> ORPHA 86909  Myoclonic epilepsy of infancy</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="defining-and-using-requirements-for-knowledge-management-kmr">Defining and using Requirements for Knowledge Management (KMR)<a class="anchor" aria-label="anchor" href="#defining-and-using-requirements-for-knowledge-management-kmr"></a>
</h2>
<p>Beside the relational model, no additional constraints are applied to
an MDB. This allows for high flexibility in the data that can be
managed. However, in some cases, it could be useful to add further
constraints to ensure that the data is compatible with specific analysis
or integration workflows. In TKCat, this feature is supported by KMR
(Knowledge Management Requirements). A KMR object is meant to be shared
and centrally managed. MDBs intended to meet these requirements must
contain technical tables referring to the corresponding KMR. When
grouped in the same TKCat catalog, KMRs and MDBs form a coherent corpus
of knowledge that can be leveraged consistently by KMR-tailored
functions.</p>
<p>This set of features is described in the vignette <a href="TKCat-KMR-POK.html">Defining and using Requirements for Knowledge
Management (KMR) in TKCat</a>.</p>
</div>
<div class="section level2">
<h2 id="appendices">Appendices<a class="anchor" aria-label="anchor" href="#appendices"></a>
</h2>
<div class="section level3">
<h3 id="ch-op">chTKCat operations<a class="anchor" aria-label="anchor" href="#ch-op"></a>
</h3>
<div class="section level4">
<h4 id="instantiating-the-clickhouse-database">Instantiating the ClickHouse database<a class="anchor" aria-label="anchor" href="#instantiating-the-clickhouse-database"></a>
</h4>
<div class="section level5">
<h5 id="install-clickhouse-initialize-and-configure-the-tkcat-instance">Install ClickHouse, initialize and configure the TKCat instance<a class="anchor" aria-label="anchor" href="#install-clickhouse-initialize-and-configure-the-tkcat-instance"></a>
</h5>
<p>The ClickHouse docker container supporting TKCat, its initialization
and its configuration procedures are implemented here: <a href="https://github.com/patzaw/TKCat/blob/master/ClickHouse/S01-install-and-init.R" class="external-link">S01-install-and-init.R</a>.
This script should be adapted according to requirements and needs.</p>
<p>Specific attention should be paid on available ports: TCP native port
(but not TLS wrapping yet) is supported by the <a href="https://github.com/IMSMWU/RClickhouse" class="external-link">RClickhouse</a> R package
whereas HTTP and HTTP ports are supported by the <a href="https://github.com/patzaw/ClickHouseHTTP" class="external-link">ClickHouseHTTP</a> R
package.</p>
<p>The data are stored in the <code>TKCAT_HOME</code> folder.</p>
</div>
<div class="section level5">
<h5 id="cleaning-and-removing-a-tkcat-instance">Cleaning and removing a TKCat instance<a class="anchor" aria-label="anchor" href="#cleaning-and-removing-a-tkcat-instance"></a>
</h5>
<p>When no longer needed, stoping and removing the docker container can
be achieved as exemplified below</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb135-1"><a href="#cb135-1" tabindex="-1"></a><span class="co"># In shell</span></span>
<span id="cb135-2"><a href="#cb135-2" tabindex="-1"></a><span class="ex">docker</span> stop tkcat_test</span>
<span id="cb135-3"><a href="#cb135-3" tabindex="-1"></a><span class="ex">docker</span> rm tkcat_test</span>
<span id="cb135-4"><a href="#cb135-4" tabindex="-1"></a><span class="ex">docker</span> volume prune <span class="at">-f</span></span>
<span id="cb135-5"><a href="#cb135-5" tabindex="-1"></a><span class="co"># Remove the folder with all the data: `$TKCAT_HOME`.`</span></span>
<span id="cb135-6"><a href="#cb135-6" tabindex="-1"></a><span class="fu">sudo</span> rm <span class="at">-rf</span> ~/Documents/Projects/TKCat_Test</span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="user-management">User management<a class="anchor" aria-label="anchor" href="#user-management"></a>
</h4>
<p>User management requires admin rights on the database.</p>
<div class="section level5">
<h5 id="ch-u-create">Creation<a class="anchor" aria-label="anchor" href="#ch-u-create"></a>
</h5>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chTKCat.html">chTKCat</a></span><span class="op">(</span>user<span class="op">=</span><span class="st">"pgodard"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/create_chTKCat_user.html">create_chTKCat_user</a></span><span class="op">(</span></span>
<span>   <span class="va">k</span>, login<span class="op">=</span><span class="st">"lfrancois"</span>, contact<span class="op">=</span><span class="cn">NA</span>, admin<span class="op">=</span><span class="cn">FALSE</span>, provider<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function will require to setup a password for the new user. The
admin parameter indicates if the new user have admin right on the whole
chTKCat instance (default: FALSE). The provider parameter indicates if
the new user can create and populate new databases whithin the chTKCat
instance (default: FALSE).</p>
</div>
<div class="section level5">
<h5 id="update">Update<a class="anchor" aria-label="anchor" href="#update"></a>
</h5>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chTKCat.html">chTKCat</a></span><span class="op">(</span>user<span class="op">=</span><span class="st">"pgodard"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/change_chTKCat_password.html">change_chTKCat_password</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"lfrancois"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/update_chTKCat_user.html">update_chTKCat_user</a></span><span class="op">(</span><span class="va">k</span>, contact<span class="op">=</span><span class="st">"email"</span>, admin<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>A shiny application can be launched for updating user settings:</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/manage_chTKCat_users.html">manage_chTKCat_users</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span></span></code></pre></div>
<p>If this application is deployed, it can be made directly accessible
from the <code><a href="../reference/explore_MDBs.html">explore_MDBs()</a></code> Shiny application by providing the
URL as the <code>userManager</code> parameter.</p>
</div>
<div class="section level5">
<h5 id="drop">Drop<a class="anchor" aria-label="anchor" href="#drop"></a>
</h5>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/drop_chTKCat_user.html">drop_chTKCat_user</a></span><span class="op">(</span><span class="va">k</span>, login<span class="op">=</span><span class="st">"lfrancois"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="chmdb-management">chMDB management<a class="anchor" aria-label="anchor" href="#chmdb-management"></a>
</h4>
<div class="section level5">
<h5 id="chmdb-creation">chMDB Creation<a class="anchor" aria-label="anchor" href="#chmdb-creation"></a>
</h5>
<p>Before MDB data can be uploaded, the database should be created. This
operation can only be achieved by data providers (see <a href="#ch-u-create">above</a>).</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/create_chMDB.html">create_chMDB</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span>, public<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>By default chMDB are not public. It can be changed through the
<code>public</code> parameter when creating the chMDB or by using the
<code><a href="../reference/set_chMDB_access.html">set_chMDB_access()</a></code> function afterward.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/set_chMDB_access.html">set_chMDB_access</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span>, public<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Then, users having access to the chMDB can be identified with or
without admin rights on the chMDB. Admin rights allow the user to update
the chMDB data.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/add_chMDB_user.html">add_chMDB_user</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span>, <span class="st">"lfrancois"</span>, admin<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># remove_chMDB_user(k, "CHEMBL", "lfrancois")</span></span>
<span><span class="fu"><a href="../reference/list_chMDB_users.html">list_chMDB_users</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="populating-chmdb">Populating chMDB<a class="anchor" aria-label="anchor" href="#populating-chmdb"></a>
</h5>
<p>Each chMDB can be populated individualy using the
<code><a href="../reference/as_chMDB.html">as_chMDB()</a></code> function. The code chunk below shows how to scan
a directory for all <em>fileMDB</em> it contains. The
<code><a href="../reference/as_memoMDB.html">as_memoMDB()</a></code> function load all the data in memory and
checks that all the model constraints are fulfilled (this step is
optional). When <code>overwrite</code> parameter of the
<code><a href="../reference/as_chMDB.html">as_chMDB()</a></code> function is set to FALSE (default), the
potential existing version is archived before being updated. When
<code>overwrite</code> is set to TRUE, the potential existing version is
overwritten without being archived.</p>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scan_fileMDBs.html">scan_fileMDBs</a></span><span class="op">(</span><span class="st">"fileMDB_directory"</span><span class="op">)</span></span>
<span><span class="co">## The commented line below allows the exploration of the data models in lc.</span></span>
<span><span class="co"># explore_MDBs(lc)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="va">toFeed</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span>   <span class="va">lr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_memoMDB.html">as_memoMDB</a></span><span class="op">(</span><span class="va">lc</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>   <span class="va">cr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_chMDB.html">as_chMDB</a></span><span class="op">(</span><span class="va">lr</span>, <span class="va">k</span>, overwrite<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="deleting-a-chmdb">Deleting a chMDB<a class="anchor" aria-label="anchor" href="#deleting-a-chmdb"></a>
</h5>
<p>Any admin user of a chMDB can delete the corresponding data.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/empty_chMDB.html">empty_chMDB</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span><span class="op">)</span></span></code></pre></div>
<p>But only a system admin can drop the chMDB from the ClickHouse
database.</p>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/drop_chMDB.html">drop_chMDB</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"CHEMBL"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="collection-management">Collection management<a class="anchor" aria-label="anchor" href="#collection-management"></a>
</h4>
<p>Details about collections are provided in the <a href="#collections">following appendix</a>.</p>
<p>Collections needs to be added to a chTKCat instance in order to
support collection members of the different chMDB. They can be taken
from the TKCat package environment, from a JSON file or directly from a
JSON text variable. Additional functions are available to list and
remove chTKCat collections.</p>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/add_chTKCat_collection.html">add_chTKCat_collection</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"BE"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/list_chTKCat_collections.html">list_chTKCat_collections</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/remove_chTKCat_collection.html">remove_chTKCat_collection</a></span><span class="op">(</span><span class="va">k</span>, <span class="st">"BE"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h4>
<div class="section level5">
<h5 id="data-models">Data models<a class="anchor" aria-label="anchor" href="#data-models"></a>
</h5>
<div class="section level6">
<h6 id="default-database">Default database<a class="anchor" aria-label="anchor" href="#default-database"></a>
</h6>
<p>The default database stores information about chTKCat instance, users
and user access.</p>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-5b1b2f4ad92281566982" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-5b1b2f4ad92281566982">{"x":{"nodes":{"tableName":["System","Collections","Users"],"label":["<b>System<\/b>\n    - *name {character}\n    - *instance {character}\n    - *version {character}\n    - *contact {character}\n    - *path {character}","<b>Collections<\/b>\n    - <b>*title<\/b> {character - uidx.1}\n    - description {character}\n    - json {character}","<b>Users<\/b>\n    - <b>*login<\/b> {character - uidx.1}\n    - (contact {character})\n    - admin {logical}\n    - provider {logical}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">System<\/strong> (Technical table for documenting the instance of chTKCat. This table has only one row.)<\/p> <ul> <li><strong>name<\/strong>: Always \"chTKCat\"<\/li> <li><strong>instance<\/strong>: The name of the chTKCat instance<\/li> <li><strong>version<\/strong>: Version of the chTKCat instance<\/li> <li><strong>contact<\/strong>: Maintainer of the chTKCat instance<\/li> <li><strong>path<\/strong>: Local path to the ClickHouse folder<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">Collections<\/strong> (Collections of concepts: define information that must be provided by a table to identify a specific concept such as a disease or a biological entity.)<\/p> <ul> <li><strong>title<\/strong>: The name of the collection<\/li> <li><strong>description<\/strong>: Description of the collection<\/li> <li><strong>json<\/strong>: A formal definition of the collection in json format fitting the \"Collection-Schema.json\"<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">Users<\/strong><\/p> <ul> <li><strong>login<\/strong><\/li> <li><strong>contact<\/strong>: Contact information<\/li> <li><strong>admin<\/strong>: If the user is an admin of the chTKCat instance<\/li> <li><strong>provider<\/strong>: If the user is a data provider (TRUE), or a data consumer only (FALSE)<\/li> <\/ul><\/div>"],"shape":["box","box","box"],"font.multi":[true,true,true],"font.align":["left","left","left"],"x":[-627,-514,-385],"y":[-335,-198,-335],"color.background":["#F67FC4","#DFFB86","#9BC8FE"],"color.border":["black","black","black"],"color.highlight.border":["orange","orange","orange"],"color.highlight.background":["#F67FC4","#DFFB86","#9BC8FE"],"id":["System","Collections","Users"]},"edges":{"id":[],"from":[],"to":[]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level6">
<h6 id="modeled-databases">Modeled databases<a class="anchor" aria-label="anchor" href="#modeled-databases"></a>
</h6>
<p>Modeled databases (MDB) are stored in dedicated database in chTKCat.
Their data model is provided in dedicated tables described below.</p>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-25c3e940e6859592f801" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-25c3e940e6859592f801">{"x":{"nodes":{"tableName":["___MDB___","___CollectionMembers___","___Tables___","___Fields___","___PrimaryKeys___","___Indexes___","___ForeignKeys___","___MDBUsers___","___Public___","___Timestamps___"],"label":["<b>___MDB___<\/b>\n    - <b>*name<\/b> {character - uidx.1}\n    - (title {character})\n    - (description {character})\n    - (url {character})\n    - (version {character})\n    - (maintainer {character})","<b>___CollectionMembers___<\/b>\n    - <b>collection<\/b> {character - uidx.1}\n    - cid {character}\n    - <b>table<\/b> {character - uidx.1}\n    - <b>mid<\/b> {integer - uidx.1}\n    - <b>field<\/b> {character - uidx.1}\n    - static {logical}\n    - value {character}\n    - (type {character})","<b>___Tables___<\/b>\n    - <b>*name<\/b> {character - uidx.1}\n    - (x {numeric})\n    - (y {numeric})\n    - (color {character})\n    - (comment {character})","<b>___Fields___<\/b>\n    - <b>table<\/b> {character - uidx.1}\n    - <b>name<\/b> {character - uidx.1}\n    - type {character}\n    - nullable {logical}\n    - unique {logical}\n    - comment {character}\n    - fieldOrder {integer}","<b>___PrimaryKeys___<\/b>\n    - <b>table<\/b> {character - uidx.1}\n    - <b>field<\/b> {character - uidx.1}","<b>___Indexes___<\/b>\n    - <b>table<\/b> {character - uidx.1}\n    - <b>idx<\/b> {integer - uidx.1}\n    - <b>field<\/b> {character - uidx.1}\n    - unique {logical}","<b>___ForeignKeys___<\/b>\n    - <b>table<\/b> {character - uidx.1}\n    - <b>fki<\/b> {integer - uidx.1}\n    - <b>field<\/b> {character - uidx.1}\n    - <b>refTable<\/b> {character - uidx.1}\n    - <b>refField<\/b> {character - uidx.1}\n    - fmin {integer}\n    - fmax {integer}\n    - tmin {integer}\n    - tmax {integer}","<b>___MDBUsers___<\/b>\n    - <b>*login<\/b> {character - uidx.1}\n    - admin {logical}","<b>___Public___<\/b>\n    - public {logical}","<b>___Timestamps___<\/b>\n    - timestamp {POSIXct}\n    - table {character}\n    - instance {character}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___MDB___<\/strong> (The modeled DataBase)<\/p> <ul> <li><strong>name<\/strong>: The name of the modeled resource<\/li> <li><strong>title<\/strong>: A longer title for the resource<\/li> <li><strong>description<\/strong>: Description of the resource<\/li> <li><strong>url<\/strong>: URL for more information about the original resource<\/li> <li><strong>version<\/strong>: Version of resource data model. The version of the data should be included in the database itself.<\/li> <li><strong>maintainer<\/strong>: Maintainer of the resource<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___CollectionMembers___<\/strong> (Tables in MDBs which provide a collection of concepts.)<\/p> <ul> <li><strong>collection<\/strong>: The title of the collection<\/li> <li><strong>cid<\/strong>: Collection identifier<\/li> <li><strong>table<\/strong>: The name of the table providing the collection<\/li> <li><strong>mid<\/strong>: The identifier of the member<\/li> <li><strong>field<\/strong>: Part of the collection definition<\/li> <li><strong>static<\/strong>: If the field value is the same for all elements in the collection<\/li> <li><strong>value<\/strong>: If static, the value is the value of the field, else the value is the name of the column providing the field value.<\/li> <li><strong>type<\/strong>: Information about the type of the value helping its interpretation<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___Tables___<\/strong> (Tables belonging to each MDB)<\/p> <ul> <li><strong>name<\/strong>: Table name<\/li> <li><strong>x<\/strong>: (ReDaMoR) Position of the table node in the data model<\/li> <li><strong>y<\/strong>: (ReDaMoR) Position of the table node in the data model<\/li> <li><strong>color<\/strong>: (ReDaMoR) Color of the table node in the data model<\/li> <li><strong>comment<\/strong>: (ReDaMoR) table description<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___Fields___<\/strong><\/p> <ul> <li><strong>table<\/strong><\/li> <li><strong>name<\/strong>: Field name<\/li> <li><strong>type<\/strong>: Field type (R type)<\/li> <li><strong>nullable<\/strong>: If the field is nullable<\/li> <li><strong>unique<\/strong>: If the field is unique<\/li> <li><strong>comment<\/strong>: Field description<\/li> <li><strong>fieldOrder<\/strong>: (ReDaMoR) Display order of the field<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___PrimaryKeys___<\/strong> (Table primary keys)<\/p> <ul> <li><strong>table<\/strong><\/li> <li><strong>field<\/strong>: Table field belonging to the primary key<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___Indexes___<\/strong> (Table indexes)<\/p> <ul> <li><strong>table<\/strong><\/li> <li><strong>idx<\/strong>: Index identifier for the table<\/li> <li><strong>field<\/strong>: field belonging to the index<\/li> <li><strong>unique<\/strong><\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___ForeignKeys___<\/strong> (Table foreign keys)<\/p> <ul> <li><strong>table<\/strong><\/li> <li><strong>fki<\/strong>: Foreign key identifier for the table<\/li> <li><strong>field<\/strong>: field belonging to the foreign key<\/li> <li><strong>refTable<\/strong>: Targeted table<\/li> <li><strong>refField<\/strong>: Targeted field in the targeted table<\/li> <li><strong>fmin<\/strong>: Minimum cardinality from the origin table<\/li> <li><strong>fmax<\/strong>: Maximum cardinality from the origin table<\/li> <li><strong>tmin<\/strong>: Minimum cardinality to the targeted table<\/li> <li><strong>tmax<\/strong>: Minimum cardinality to the targeted table<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___MDBUsers___<\/strong> (Users having access to the MDB)<\/p> <ul> <li><strong>login<\/strong>: User login<\/li> <li><strong>admin<\/strong>: If the user can edit the MDB: the data and the model<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___Public___<\/strong> (If the resource data are available for anybody having access to the chTKCat instance)<\/p> <ul> <li><strong>public<\/strong>: If the resource data are available for anybody having access to the chTKCat instance<\/li> <\/ul><\/div>","<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">___Timestamps___<\/strong> (Instance of each table at each timestamp. For each chMDB instance all the DB tables (including the model but not the \"___Public___\", the \"___Timestamps___\" nor the \"___MDBUsers___\" tables) are associated with an actual instance table in Clickouse. For the current instance, instance has the same value as table.)<\/p> <ul> <li><strong>timestamp<\/strong>: chMDB instance timestamp<\/li> <li><strong>table<\/strong>: Table name<\/li> <li><strong>instance<\/strong>: Table instance in Clickhouse<\/li> <\/ul><\/div>"],"shape":["box","box","box","box","box","box","box","box","box","box"],"font.multi":[true,true,true,true,true,true,true,true,true,true],"font.align":["left","left","left","left","left","left","left","left","left","left"],"x":[-31,-350,-355,-354,-248,-25,-658,-662,-695,-677],"y":[-249,-304,-112,197,36,39,45,-267,-338,-179],"color.background":["#A9ECC9","#DFFB86","#8FE6E0","#8FE6E0","#C6BDF1","#C6BDF1","#C6BDF1","#FAC6DC","#FAC6DC","#FAC6DC"],"color.border":["black","black","black","black","black","black","black","black","black","black"],"color.highlight.border":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"],"color.highlight.background":["#A9ECC9","#DFFB86","#8FE6E0","#8FE6E0","#C6BDF1","#C6BDF1","#C6BDF1","#FAC6DC","#FAC6DC","#FAC6DC"],"id":["___MDB___","___CollectionMembers___","___Tables___","___Fields___","___PrimaryKeys___","___Indexes___","___ForeignKeys___","___MDBUsers___","___Public___","___Timestamps___"]},"edges":{"id":["___CollectionMembers___->___Tables___: table->name","___Fields___->___Tables___: table->name","___PrimaryKeys___->___Tables___: table->name","___PrimaryKeys___->___Fields___: field->name && table->table","___Indexes___->___Tables___: table->name","___Indexes___->___Fields___: field->name && table->table","___ForeignKeys___->___Tables___: table->name","___ForeignKeys___->___Fields___: field->name && table->table","___ForeignKeys___->___Fields___: refField->name && refTable->table","___ForeignKeys___->___Tables___: refTable->name"],"to":["___Tables___","___Tables___","___Tables___","___Fields___","___Tables___","___Fields___","___Tables___","___Fields___","___Fields___","___Tables___"],"title":["<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___CollectionMembers___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Fields___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___PrimaryKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___PrimaryKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Fields___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">field<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">table<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Indexes___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Indexes___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Fields___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">field<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">table<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___ForeignKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___ForeignKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Fields___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">field<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">table<\/td><td style=\"border: 1px solid black; padding:1px;\">table<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___ForeignKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Fields___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">refField<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">refTable<\/td><td style=\"border: 1px solid black; padding:1px;\">table<\/td><\/tr><\/table>","<table style=\"border: 1px solid black; padding:1px;\"><tr style=\"border: 1px solid black; padding:1px;\"><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___ForeignKeys___<br>(0..n)<\/th><th style=\"border: 1px solid black; padding:1px;text-align:center;\">___Tables___<br>(1..1)<\/th><\/tr><tr style=\"border: 1px solid black; padding:1px;\"><td style=\"border: 1px solid black; padding:1px;\">refTable<\/td><td style=\"border: 1px solid black; padding:1px;\">name<\/td><\/tr><\/table>"],"ff":["table","table","table",["field","table"],"table",["field","table"],"table",["field","table"],["refField","refTable"],"refTable"],"tf":["name","name","name",["name","table"],"name",["name","table"],"name",["name","table"],["name","table"],"name"],"from":["___CollectionMembers___","___Fields___","___PrimaryKeys___","___PrimaryKeys___","___Indexes___","___Indexes___","___ForeignKeys___","___ForeignKeys___","___ForeignKeys___","___ForeignKeys___"],"arrows":["to","to","to","to","to","to","to","to","to","to"],"font.align":["bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom","bottom"],"smooth.type":["curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW","curvedCCW"],"smooth.roundness":[-0,-0,-0,0,-0,0,-0.2,0.2,-0.2,0.2],"selfReferenceSize":[30,30,30,30,30,30,30,30,50,50],"uef":["___CollectionMembers___","___Fields___","___PrimaryKeys___","___Fields___","___Indexes___","___Fields___","___ForeignKeys___","___Fields___","___Fields___","___ForeignKeys___"],"uet":["___Tables___","___Tables___","___Tables___","___PrimaryKeys___","___Tables___","___Indexes___","___Tables___","___ForeignKeys___","___ForeignKeys___","___Tables___"],"ue":["___CollectionMembers___<->___Tables___","___Fields___<->___Tables___","___PrimaryKeys___<->___Tables___","___Fields___<->___PrimaryKeys___","___Indexes___<->___Tables___","___Fields___<->___Indexes___","___ForeignKeys___<->___Tables___","___Fields___<->___ForeignKeys___","___Fields___<->___ForeignKeys___","___ForeignKeys___<->___Tables___"],"color.color":["black","black","black","black","black","black","black","black","black","black"],"color.highlight":["orange","orange","orange","orange","orange","orange","orange","orange","orange","orange"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="collections">TKCat collections<a class="anchor" aria-label="anchor" href="#collections"></a>
</h3>
<p>Some MDBs refer to the same concepts and can be integrated
accordingly. However they often use different vocabularies or scopes.
Collections are used to identify such concepts and to define a way to
document formally the scope used by the different members of these
collections. Thanks to this formal description, tools can be used to
automatically combine MDBs referring to the same collection but using
different scopes, as shown <a href="#merging-with-collections">above</a>.</p>
<p>This appendix describes how to create TKCat Collections, document
collection members and create functions to support the merging of
MDBs.</p>
<div class="section level4">
<h4 id="creating-a-collection">Creating a collection<a class="anchor" aria-label="anchor" href="#creating-a-collection"></a>
</h4>
<p>A collection is defined by a JSON document. This document should
fulfill the requirements defined by the <a href="https://github.com/patzaw/TKCat/blob/master/inst/Collections/Collection-Schema.json" class="external-link">Collection-Schema.json</a>.
Two collections are available by default in the TKCat package.</p>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/list_local_collections.html">list_local_collections</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   title     description                                  </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> BE        Collection of biological entity (BE) concepts</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Condition Collection of condition concepts</span></span></code></pre>
<p>Here is how the <em>BE</em> collection is defined.</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_local_collection.html">get_local_collection</a></span><span class="op">(</span><span class="st">"BE"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb150"><pre class="sourceCode json"><code class="sourceCode json"><span id="cb150-1"><a href="#cb150-1" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb150-2"><a href="#cb150-2" tabindex="-1"></a>   <span class="dt">"$schema"</span><span class="fu">:</span> <span class="st">"https://json-schema.org/draft/2019-09/schema"</span><span class="fu">,</span></span>
<span id="cb150-3"><a href="#cb150-3" tabindex="-1"></a>   <span class="dt">"$id"</span><span class="fu">:</span><span class="st">"TKCat_BE_collection_1.0"</span><span class="fu">,</span></span>
<span id="cb150-4"><a href="#cb150-4" tabindex="-1"></a>    <span class="dt">"title"</span><span class="fu">:</span> <span class="st">"BE collection"</span><span class="fu">,</span></span>
<span id="cb150-5"><a href="#cb150-5" tabindex="-1"></a>    <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-6"><a href="#cb150-6" tabindex="-1"></a>    <span class="dt">"description"</span><span class="fu">:</span> <span class="st">"Collection of biological entity (BE) concepts"</span><span class="fu">,</span></span>
<span id="cb150-7"><a href="#cb150-7" tabindex="-1"></a>    <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-8"><a href="#cb150-8" tabindex="-1"></a>      <span class="dt">"$schema"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"enum"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"TKCat_BE_collection_1.0"</span><span class="ot">]</span><span class="fu">},</span></span>
<span id="cb150-9"><a href="#cb150-9" tabindex="-1"></a>      <span class="dt">"$id"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">},</span></span>
<span id="cb150-10"><a href="#cb150-10" tabindex="-1"></a>        <span class="dt">"collection"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"enum"</span><span class="fu">:</span><span class="ot">[</span><span class="st">"BE"</span><span class="ot">]</span><span class="fu">},</span></span>
<span id="cb150-11"><a href="#cb150-11" tabindex="-1"></a>        <span class="dt">"resource"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">},</span></span>
<span id="cb150-12"><a href="#cb150-12" tabindex="-1"></a>        <span class="dt">"tables"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-13"><a href="#cb150-13" tabindex="-1"></a>            <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"array"</span><span class="fu">,</span></span>
<span id="cb150-14"><a href="#cb150-14" tabindex="-1"></a>            <span class="dt">"minItems"</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></span>
<span id="cb150-15"><a href="#cb150-15" tabindex="-1"></a>            <span class="dt">"items"</span><span class="fu">:{</span></span>
<span id="cb150-16"><a href="#cb150-16" tabindex="-1"></a>                <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-17"><a href="#cb150-17" tabindex="-1"></a>                <span class="dt">"properties"</span><span class="fu">:{</span></span>
<span id="cb150-18"><a href="#cb150-18" tabindex="-1"></a>                    <span class="dt">"name"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">},</span></span>
<span id="cb150-19"><a href="#cb150-19" tabindex="-1"></a>                    <span class="dt">"fields"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-20"><a href="#cb150-20" tabindex="-1"></a>                        <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-21"><a href="#cb150-21" tabindex="-1"></a>                        <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-22"><a href="#cb150-22" tabindex="-1"></a>                            <span class="dt">"be"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-23"><a href="#cb150-23" tabindex="-1"></a>                                <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-24"><a href="#cb150-24" tabindex="-1"></a>                                <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-25"><a href="#cb150-25" tabindex="-1"></a>                                    <span class="dt">"static"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"boolean"</span><span class="fu">},</span></span>
<span id="cb150-26"><a href="#cb150-26" tabindex="-1"></a>                                    <span class="dt">"value"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">}</span></span>
<span id="cb150-27"><a href="#cb150-27" tabindex="-1"></a>                                <span class="fu">},</span></span>
<span id="cb150-28"><a href="#cb150-28" tabindex="-1"></a>                                <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"static"</span><span class="ot">,</span> <span class="st">"value"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-29"><a href="#cb150-29" tabindex="-1"></a>                                <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-30"><a href="#cb150-30" tabindex="-1"></a>                            <span class="fu">},</span></span>
<span id="cb150-31"><a href="#cb150-31" tabindex="-1"></a>                            <span class="dt">"source"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-32"><a href="#cb150-32" tabindex="-1"></a>                                <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-33"><a href="#cb150-33" tabindex="-1"></a>                                <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-34"><a href="#cb150-34" tabindex="-1"></a>                                    <span class="dt">"static"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"boolean"</span><span class="fu">},</span></span>
<span id="cb150-35"><a href="#cb150-35" tabindex="-1"></a>                                    <span class="dt">"value"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">}</span></span>
<span id="cb150-36"><a href="#cb150-36" tabindex="-1"></a>                                <span class="fu">},</span></span>
<span id="cb150-37"><a href="#cb150-37" tabindex="-1"></a>                                <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"static"</span><span class="ot">,</span> <span class="st">"value"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-38"><a href="#cb150-38" tabindex="-1"></a>                                <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-39"><a href="#cb150-39" tabindex="-1"></a>                            <span class="fu">},</span></span>
<span id="cb150-40"><a href="#cb150-40" tabindex="-1"></a>                            <span class="dt">"organism"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-41"><a href="#cb150-41" tabindex="-1"></a>                                <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-42"><a href="#cb150-42" tabindex="-1"></a>                                <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-43"><a href="#cb150-43" tabindex="-1"></a>                                    <span class="dt">"static"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"boolean"</span><span class="fu">},</span></span>
<span id="cb150-44"><a href="#cb150-44" tabindex="-1"></a>                                    <span class="dt">"value"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">},</span></span>
<span id="cb150-45"><a href="#cb150-45" tabindex="-1"></a>                                    <span class="dt">"type"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"enum"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"Scientific name"</span><span class="ot">,</span> <span class="st">"NCBI taxon identifier"</span><span class="ot">]</span><span class="fu">}</span></span>
<span id="cb150-46"><a href="#cb150-46" tabindex="-1"></a>                                <span class="fu">},</span></span>
<span id="cb150-47"><a href="#cb150-47" tabindex="-1"></a>                                <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"static"</span><span class="ot">,</span> <span class="st">"value"</span><span class="ot">,</span> <span class="st">"type"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-48"><a href="#cb150-48" tabindex="-1"></a>                                <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-49"><a href="#cb150-49" tabindex="-1"></a>                            <span class="fu">},</span></span>
<span id="cb150-50"><a href="#cb150-50" tabindex="-1"></a>                            <span class="dt">"identifier"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-51"><a href="#cb150-51" tabindex="-1"></a>                                <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"object"</span><span class="fu">,</span></span>
<span id="cb150-52"><a href="#cb150-52" tabindex="-1"></a>                                <span class="dt">"properties"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb150-53"><a href="#cb150-53" tabindex="-1"></a>                                    <span class="dt">"static"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"boolean"</span><span class="fu">},</span></span>
<span id="cb150-54"><a href="#cb150-54" tabindex="-1"></a>                                    <span class="dt">"value"</span><span class="fu">:</span> <span class="fu">{</span><span class="dt">"type"</span><span class="fu">:</span> <span class="st">"string"</span><span class="fu">}</span></span>
<span id="cb150-55"><a href="#cb150-55" tabindex="-1"></a>                                <span class="fu">},</span></span>
<span id="cb150-56"><a href="#cb150-56" tabindex="-1"></a>                                <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"static"</span><span class="ot">,</span> <span class="st">"value"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-57"><a href="#cb150-57" tabindex="-1"></a>                                <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-58"><a href="#cb150-58" tabindex="-1"></a>                            <span class="fu">}</span></span>
<span id="cb150-59"><a href="#cb150-59" tabindex="-1"></a>                        <span class="fu">},</span></span>
<span id="cb150-60"><a href="#cb150-60" tabindex="-1"></a>                        <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"be"</span><span class="ot">,</span> <span class="st">"source"</span><span class="ot">,</span> <span class="st">"identifier"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-61"><a href="#cb150-61" tabindex="-1"></a>                        <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-62"><a href="#cb150-62" tabindex="-1"></a>                    <span class="fu">}</span></span>
<span id="cb150-63"><a href="#cb150-63" tabindex="-1"></a>                <span class="fu">},</span></span>
<span id="cb150-64"><a href="#cb150-64" tabindex="-1"></a>                <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"name"</span><span class="ot">,</span> <span class="st">"fields"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-65"><a href="#cb150-65" tabindex="-1"></a>                <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-66"><a href="#cb150-66" tabindex="-1"></a>            <span class="fu">}</span></span>
<span id="cb150-67"><a href="#cb150-67" tabindex="-1"></a>        <span class="fu">}</span></span>
<span id="cb150-68"><a href="#cb150-68" tabindex="-1"></a>    <span class="fu">},</span></span>
<span id="cb150-69"><a href="#cb150-69" tabindex="-1"></a>    <span class="dt">"required"</span><span class="fu">:</span> <span class="ot">[</span><span class="st">"$schema"</span><span class="ot">,</span> <span class="st">"$id"</span><span class="ot">,</span> <span class="st">"collection"</span><span class="ot">,</span> <span class="st">"resource"</span><span class="ot">,</span> <span class="st">"tables"</span><span class="ot">]</span><span class="fu">,</span></span>
<span id="cb150-70"><a href="#cb150-70" tabindex="-1"></a>    <span class="dt">"additionalProperties"</span><span class="fu">:</span> <span class="kw">false</span></span>
<span id="cb150-71"><a href="#cb150-71" tabindex="-1"></a><span class="fu">}</span></span></code></pre></div>
<p>A collection should refer to the <code>"TKCat_collections_1.0"</code>
<strong>$schema</strong>. It should then have the following
properties:</p>
<ul>
<li><p><strong>$id</strong>: the identifier of the collection</p></li>
<li><p><strong>title</strong>: the title of the collection</p></li>
<li><p><strong>type</strong>: always <code>object</code></p></li>
<li><p><strong>description</strong>: a short description of the
collection</p></li>
<li>
<p><strong>properties</strong>: the properties that should be
provided by collection members. In this case:</p>
<ul>
<li><p><strong>$schema</strong>: should be the <em>$id</em> of the
collection</p></li>
<li><p><strong>$id</strong>: the identifier of the collection member: a
string</p></li>
<li><p><strong>collection</strong>: should be “BE”</p></li>
<li><p><strong>resource</strong>: the name of the resource having
collection members: a string</p></li>
<li>
<p><strong>tables</strong>: an array of tables corresponding to
collection members. Each item being a table with the following
features:</p>
<ul>
<li><p><strong>name</strong>: the name of the table</p></li>
<li>
<p><strong>fields</strong>: the required fields</p>
<ul>
<li>
<strong>be</strong>: if <strong>static</strong> is true then
<strong>value</strong> correspond to the be value valid for all the
records. If not <strong>value</strong> correspond to the table column
with the be value for each record.</li>
<li>
<strong>source</strong>: if <strong>static</strong> is true then
<strong>value</strong> correspond to the source value valid for all the
records. If not <strong>value</strong> correspond to the table column
with the source value for each record.</li>
<li>
<strong>organism</strong>: if <strong>static</strong> is true then
<strong>value</strong> correspond to the organism value valid for all
the records. If not <strong>value</strong> correspond to the table
column with the organism value for each record. <strong>type</strong>
indicate how organisms are identified: <code>"Scientific name"</code> or
<code>"NCBI taxon identifier"</code>.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>The main specifications defined in a JSON document can be simply
displayed in R session by calling the <code><a href="../reference/show_collection_def.html">show_collection_def()</a></code>
function.</p>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_local_collection.html">get_local_collection</a></span><span class="op">(</span><span class="st">"BE"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>   <span class="fu"><a href="../reference/show_collection_def.html">show_collection_def</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## BE collection: Collection of biological entity (BE) concepts</span></span>
<span><span class="co">## Arguments (non-mandatory arguments are between parentheses):</span></span>
<span><span class="co">##    - be:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span>
<span><span class="co">##    - source:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span>
<span><span class="co">##    - (organism):</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span>
<span><span class="co">##       + type: character in 'Scientific name', 'NCBI taxon identifier'</span></span>
<span><span class="co">##    - identifier:</span></span>
<span><span class="co">##       + static: logical</span></span>
<span><span class="co">##       + value: character</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="documenting-collection-members">Documenting collection members<a class="anchor" aria-label="anchor" href="#documenting-collection-members"></a>
</h4>
<p>Documenting collection members of an <em>MDB</em> can be done by
using the <code><a href="../reference/add_collection_member.html">add_collection_member()</a></code> function (as <a href="#min-coll-memb">formerly described</a>), or by writing a JSON file
like the following one which correspond to BE members of the CHEMBL
MDB.</p>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>   <span class="st">"examples/CHEMBL/model/Collections/BE-CHEMBL_BE_1.0.json"</span>,</span>
<span>   package<span class="op">=</span><span class="st">"TKCat"</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>collapse<span class="op">=</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb154"><pre class="sourceCode json"><code class="sourceCode json"><span id="cb154-1"><a href="#cb154-1" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb154-2"><a href="#cb154-2" tabindex="-1"></a>  <span class="dt">"$schema"</span><span class="fu">:</span> <span class="st">"TKCat_BE_collection_1.0"</span><span class="fu">,</span></span>
<span id="cb154-3"><a href="#cb154-3" tabindex="-1"></a>  <span class="dt">"$id"</span><span class="fu">:</span> <span class="st">"CHEMBL_BE_1.0"</span><span class="fu">,</span></span>
<span id="cb154-4"><a href="#cb154-4" tabindex="-1"></a>  <span class="dt">"collection"</span><span class="fu">:</span> <span class="st">"BE"</span><span class="fu">,</span></span>
<span id="cb154-5"><a href="#cb154-5" tabindex="-1"></a>  <span class="dt">"resource"</span><span class="fu">:</span> <span class="st">"CHEMBL"</span><span class="fu">,</span></span>
<span id="cb154-6"><a href="#cb154-6" tabindex="-1"></a>  <span class="dt">"tables"</span><span class="fu">:</span> <span class="ot">[</span></span>
<span id="cb154-7"><a href="#cb154-7" tabindex="-1"></a>    <span class="fu">{</span></span>
<span id="cb154-8"><a href="#cb154-8" tabindex="-1"></a>      <span class="dt">"name"</span><span class="fu">:</span> <span class="st">"CHEMBL_component_sequence"</span><span class="fu">,</span></span>
<span id="cb154-9"><a href="#cb154-9" tabindex="-1"></a>      <span class="dt">"fields"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb154-10"><a href="#cb154-10" tabindex="-1"></a>        <span class="dt">"be"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb154-11"><a href="#cb154-11" tabindex="-1"></a>          <span class="dt">"static"</span><span class="fu">:</span> <span class="kw">true</span><span class="fu">,</span></span>
<span id="cb154-12"><a href="#cb154-12" tabindex="-1"></a>          <span class="dt">"value"</span><span class="fu">:</span> <span class="st">"Peptide"</span></span>
<span id="cb154-13"><a href="#cb154-13" tabindex="-1"></a>        <span class="fu">},</span></span>
<span id="cb154-14"><a href="#cb154-14" tabindex="-1"></a>        <span class="dt">"identifier"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb154-15"><a href="#cb154-15" tabindex="-1"></a>          <span class="dt">"static"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb154-16"><a href="#cb154-16" tabindex="-1"></a>          <span class="dt">"value"</span><span class="fu">:</span> <span class="st">"accession"</span></span>
<span id="cb154-17"><a href="#cb154-17" tabindex="-1"></a>        <span class="fu">},</span></span>
<span id="cb154-18"><a href="#cb154-18" tabindex="-1"></a>        <span class="dt">"source"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb154-19"><a href="#cb154-19" tabindex="-1"></a>          <span class="dt">"static"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb154-20"><a href="#cb154-20" tabindex="-1"></a>          <span class="dt">"value"</span><span class="fu">:</span> <span class="st">"db_source"</span></span>
<span id="cb154-21"><a href="#cb154-21" tabindex="-1"></a>        <span class="fu">},</span></span>
<span id="cb154-22"><a href="#cb154-22" tabindex="-1"></a>        <span class="dt">"organism"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb154-23"><a href="#cb154-23" tabindex="-1"></a>          <span class="dt">"static"</span><span class="fu">:</span> <span class="kw">false</span><span class="fu">,</span></span>
<span id="cb154-24"><a href="#cb154-24" tabindex="-1"></a>          <span class="dt">"value"</span><span class="fu">:</span> <span class="st">"organism"</span><span class="fu">,</span></span>
<span id="cb154-25"><a href="#cb154-25" tabindex="-1"></a>          <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"Scientific name"</span></span>
<span id="cb154-26"><a href="#cb154-26" tabindex="-1"></a>        <span class="fu">}</span></span>
<span id="cb154-27"><a href="#cb154-27" tabindex="-1"></a>      <span class="fu">}</span></span>
<span id="cb154-28"><a href="#cb154-28" tabindex="-1"></a>    <span class="fu">}</span></span>
<span id="cb154-29"><a href="#cb154-29" tabindex="-1"></a>  <span class="ot">]</span></span>
<span id="cb154-30"><a href="#cb154-30" tabindex="-1"></a><span class="fu">}</span></span></code></pre></div>
<p>The identification of collection members should fulfill the
requirements defined by the collection JSON document, and therefore pass
the following validation.</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">jsonvalidate</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/jsonvalidate/reference/json_validate.html" class="external-link">json_validate</a></span><span class="op">(</span></span>
<span>   json<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>      <span class="st">"examples/CHEMBL/model/Collections/BE-CHEMBL_BE_1.0.json"</span>,</span>
<span>      package<span class="op">=</span><span class="st">"TKCat"</span></span>
<span>   <span class="op">)</span>,</span>
<span>   schema<span class="op">=</span><span class="fu"><a href="../reference/get_local_collection.html">get_local_collection</a></span><span class="op">(</span><span class="st">"BE"</span><span class="op">)</span>,</span>
<span>   engine<span class="op">=</span><span class="st">"ajv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>This validation is done automatically when reading a <em>fileMDB</em>
object or when setting collection members with the
<code><a href="../reference/add_collection_member.html">add_collection_member()</a></code> function.</p>
</div>
<div class="section level4">
<h4 id="collection-mapper-functions">Collection mapper functions<a class="anchor" aria-label="anchor" href="#collection-mapper-functions"></a>
</h4>
<p>The <code><a href="../reference/MDB.html">merge.MDB()</a></code> and the
<code><a href="../reference/map_collection_members.html">map_collection_members()</a></code> functions rely on functions to map
members of the same collection. When recorded (using the
<code><a href="../reference/import_collection_mapper.html">import_collection_mapper()</a></code> function), these functions can
be automatically identified by TKCat, otherwise or according to user
needs, these functions could be provided using the <code>funs</code>
(for <code><a href="../reference/MDB.html">merge.MDB()</a></code>) or the <code>fun</code> (for
<code><a href="../reference/map_collection_members.html">map_collection_members()</a></code>) parameters. Two mappers are
pre-recorded in TKCat, one for the <em>BE</em> collection and one for
the <em>Condition</em> collection. They can be retrieved with the
<code><a href="../reference/get_collection_mapper.html">get_collection_mapper()</a></code> function.</p>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_collection_mapper.html">get_collection_mapper</a></span><span class="op">(</span><span class="st">"BE"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">orthologs</span> <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">restricted</span> <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">...</span><span class="op">)</span> </span>
<span><span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BED"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The BED package is required"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/checkBedConn.html" class="external-link">checkBedConn</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"You need to connect to a BED database using"</span>, <span class="st">" the BED::connectToBed() function"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"organism"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">d</span> <span class="op">&lt;-</span> <span class="va">x</span></span>
<span>        <span class="va">scopes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">d</span>, <span class="va">be</span>, <span class="va">source</span><span class="op">)</span></span>
<span>        <span class="va">nd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">scopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">be</span> <span class="op">&lt;-</span> <span class="va">scopes</span><span class="op">$</span><span class="va">be</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>            <span class="va">source</span> <span class="op">&lt;-</span> <span class="va">scopes</span><span class="op">$</span><span class="va">source</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>            <span class="va">toadd</span> <span class="op">&lt;-</span> <span class="va">d</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">be</span> <span class="op">==</span> <span class="va">be</span>, <span class="va">source</span> <span class="op">==</span> </span>
<span>                <span class="va">source</span><span class="op">)</span></span>
<span>            <span class="va">organism</span> <span class="op">&lt;-</span> <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/guessIdScope.html" class="external-link">guessIdScope</a></span><span class="op">(</span><span class="va">toadd</span><span class="op">$</span><span class="va">identifier</span>, be <span class="op">=</span> <span class="va">be</span>, </span>
<span>                source <span class="op">=</span> <span class="va">source</span>, tcLim <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"details"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">be</span> <span class="op">==</span> <span class="op">!</span><span class="op">!</span><span class="va">be</span> <span class="op">&amp;</span> <span class="va">source</span> <span class="op">==</span> <span class="op">!</span><span class="op">!</span><span class="va">source</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">organism</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span>            <span class="va">toadd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">toadd</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="va">organism</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="va">nd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">nd</span>, <span class="va">toadd</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">nd</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>organism_type <span class="op">=</span> <span class="st">"Scientific name"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"organism"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">d</span> <span class="op">&lt;-</span> <span class="va">y</span></span>
<span>        <span class="va">scopes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">d</span>, <span class="va">be</span>, <span class="va">source</span><span class="op">)</span></span>
<span>        <span class="va">nd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">scopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">be</span> <span class="op">&lt;-</span> <span class="va">scopes</span><span class="op">$</span><span class="va">be</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>            <span class="va">source</span> <span class="op">&lt;-</span> <span class="va">scopes</span><span class="op">$</span><span class="va">source</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>            <span class="va">toadd</span> <span class="op">&lt;-</span> <span class="va">d</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">be</span> <span class="op">==</span> <span class="va">be</span>, <span class="va">source</span> <span class="op">==</span> </span>
<span>                <span class="va">source</span><span class="op">)</span></span>
<span>            <span class="va">organism</span> <span class="op">&lt;-</span> <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/guessIdScope.html" class="external-link">guessIdScope</a></span><span class="op">(</span><span class="va">toadd</span><span class="op">$</span><span class="va">identifier</span>, be <span class="op">=</span> <span class="va">be</span>, </span>
<span>                source <span class="op">=</span> <span class="va">source</span>, tcLim <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"details"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">be</span> <span class="op">==</span> <span class="op">!</span><span class="op">!</span><span class="va">be</span> <span class="op">&amp;</span> <span class="va">source</span> <span class="op">==</span> <span class="op">!</span><span class="op">!</span><span class="va">source</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">organism</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span>            <span class="va">toadd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">toadd</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="va">organism</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="va">nd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">nd</span>, <span class="va">toadd</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">nd</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>organism_type <span class="op">=</span> <span class="st">"Scientific name"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">xscopes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">be</span>, <span class="va">source</span>, <span class="va">organism</span>, <span class="va">organism_type</span><span class="op">)</span></span>
<span>    <span class="va">yscopes</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">be</span>, <span class="va">source</span>, <span class="va">organism</span>, <span class="va">organism_type</span><span class="op">)</span></span>
<span>    <span class="va">toRet</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">xscopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">xscope</span> <span class="op">&lt;-</span> <span class="va">xscopes</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">xscope</span>, <span class="fl">2</span>, <span class="va">is.na</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="op">(</span><span class="kw">next</span><span class="op">)</span><span class="op">(</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">xi</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">xscope</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"be"</span>, <span class="st">"source"</span>, </span>
<span>            <span class="st">"organism"</span>, <span class="st">"organism_type"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">xorg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">xscope</span><span class="op">$</span><span class="va">organism_type</span> <span class="op">==</span> <span class="st">"NCBI taxon identifier"</span>, </span>
<span>            <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/getOrgNames.html" class="external-link">getOrgNames</a></span><span class="op">(</span><span class="va">xscope</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nameClass</span> <span class="op">==</span> </span>
<span>                <span class="st">"scientific name"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span>, <span class="va">xscope</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">yscopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">yscope</span> <span class="op">&lt;-</span> <span class="va">yscopes</span><span class="op">[</span><span class="va">j</span>, <span class="op">]</span></span>
<span>            <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">yscope</span>, <span class="fl">2</span>, <span class="va">is.na</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>                <span class="op">(</span><span class="kw">next</span><span class="op">)</span><span class="op">(</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>            <span class="va">yi</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">yscope</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"be"</span>, <span class="st">"source"</span>, </span>
<span>                <span class="st">"organism"</span>, <span class="st">"organism_type"</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="va">yorg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">yscope</span><span class="op">$</span><span class="va">organism_type</span> <span class="op">==</span> <span class="st">"NCBI taxon identifier"</span>, </span>
<span>                <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/getOrgNames.html" class="external-link">getOrgNames</a></span><span class="op">(</span><span class="va">yscope</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nameClass</span> <span class="op">==</span> </span>
<span>                  <span class="st">"scientific name"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span>, <span class="va">yscope</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span></span>
<span>            <span class="kw">if</span> <span class="op">(</span><span class="va">xorg</span> <span class="op">==</span> <span class="va">yorg</span> <span class="op">||</span> <span class="va">orthologs</span><span class="op">)</span> <span class="op">{</span></span>
<span>                <span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/convBeIds.html" class="external-link">convBeIds</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">xi</span><span class="op">$</span><span class="va">identifier</span>, from <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">be</span>, </span>
<span>                  from.source <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">source</span>, from.org <span class="op">=</span> <span class="va">xorg</span>, </span>
<span>                  to <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">be</span>, to.source <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">source</span>, </span>
<span>                  to.org <span class="op">=</span> <span class="va">yorg</span>, restricted <span class="op">=</span> <span class="va">restricted</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">from</span>, </span>
<span>                  <span class="va">to</span><span class="op">)</span></span>
<span>                <span class="kw">if</span> <span class="op">(</span><span class="va">restricted</span><span class="op">)</span> <span class="op">{</span></span>
<span>                  <span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">xy</span>, <span class="fu">BED</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/BED/reference/convBeIds.html" class="external-link">convBeIds</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">yi</span><span class="op">$</span><span class="va">identifier</span>, </span>
<span>                    from <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">be</span>, from.source <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">source</span>, </span>
<span>                    from.org <span class="op">=</span> <span class="va">yorg</span>, to <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">be</span>, to.source <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">source</span>, </span>
<span>                    to.org <span class="op">=</span> <span class="va">xorg</span>, restricted <span class="op">=</span> <span class="va">restricted</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>to <span class="op">=</span> <span class="va">from</span>, </span>
<span>                    from <span class="op">=</span> <span class="va">to</span><span class="op">)</span><span class="op">)</span></span>
<span>                <span class="op">}</span></span>
<span>                <span class="va">xy</span> <span class="op">&lt;-</span> <span class="va">xy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>identifier_x <span class="op">=</span> <span class="st">"from"</span>, </span>
<span>                  identifier_y <span class="op">=</span> <span class="st">"to"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>be_x <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">be</span>, </span>
<span>                  source_x <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">source</span>, organism_x <span class="op">=</span> <span class="va">xscope</span><span class="op">$</span><span class="va">organism</span>, </span>
<span>                  be_y <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">be</span>, source_y <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">source</span>, </span>
<span>                  organism_y <span class="op">=</span> <span class="va">yscope</span><span class="op">$</span><span class="va">organism</span><span class="op">)</span></span>
<span>                <span class="va">toRet</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">toRet</span>, <span class="va">xy</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">toRet</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">toRet</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">toRet</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>A mapper function must have at least an x and a y parameters. Each of
them should be a data.frame with all the field values corresponding to
the fields defined in the collection. Additional parameters can be
defined and will be forwarded using <code>...</code>. This function
should return a data frame with all the fields values followed by “_x”
and “_y” suffix accordingly.</p>
</div>
</div>
<div class="section level3">
<h3 id="remarks-about-supported-data-format-and-data-types">Remarks about supported data format and data types<a class="anchor" aria-label="anchor" href="#remarks-about-supported-data-format-and-data-types"></a>
</h3>
<p>Most of the data format and data types supported by the ReDaMoR and
the TKCat packages are taken into account in the examples described in
the main sections of this vignette. Nevertheless, one specific data
format (matrix) and one specific data type (base64) are not exemplified.
This appendix provides a short description of these format and type.</p>
<div class="section level4">
<h4 id="matrices-of-values">Matrices of values<a class="anchor" aria-label="anchor" href="#matrices-of-values"></a>
</h4>
<p>ReDaMoR and TKCat support data frame and matrix objectq. Data frame
is the most used data format from far. However, matrices of values can
be useful in some use cases. The example below shows how such data
format are modeled in ReDaMoR as a 3 columns table: one of type “row”
corresponding to the row names of the matrix, one of type “column”
corresponding to the column names of the matrix, and one of any type
(excepted “row”, “column”, or “base64”).</p>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">40</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">10</span>,</span>
<span>   dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"g"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"s"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/df_to_model.html" class="external-link">df_to_model</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/rename_field.html" class="external-link">rename_field</a></span><span class="op">(</span><span class="st">"d"</span>, <span class="st">"row"</span>, <span class="st">"gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>   <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/update_field.html" class="external-link">update_field</a></span><span class="op">(</span><span class="st">"d"</span>, <span class="st">"gene"</span>, comment<span class="op">=</span><span class="st">"Gene identifier"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/rename_field.html" class="external-link">rename_field</a></span><span class="op">(</span><span class="st">"d"</span>, <span class="st">"column"</span>, <span class="st">"sample"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/update_field.html" class="external-link">update_field</a></span><span class="op">(</span><span class="st">"d"</span>, <span class="st">"sample"</span>, comment<span class="op">=</span><span class="st">"Sample identifier"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu">ReDaMoR</span><span class="fu">::</span><span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/rename_field.html" class="external-link">rename_field</a></span><span class="op">(</span><span class="st">"d"</span>, <span class="st">"value"</span>, <span class="st">"expression"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/update_field.html" class="external-link">update_field</a></span><span class="op">(</span></span>
<span>      <span class="st">"d"</span>, <span class="st">"expression"</span>, nullable<span class="op">=</span><span class="cn">FALSE</span>, comment<span class="op">=</span><span class="st">"Gene expression value"</span></span>
<span>   <span class="op">)</span></span>
<span><span class="va">md</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/memoMDB.html">memoMDB</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>d<span class="op">=</span><span class="va">d</span><span class="op">)</span>, <span class="va">m</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"Matrix example"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="va">md</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-3f27c09be0c60bb52829" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-3f27c09be0c60bb52829">{"x":{"nodes":{"tableName":["d"],"label":["<b>d []<\/b>\n    - <b>gene<\/b> {row - uidx.1}\n    - <b>sample<\/b> {column - uidx.1}\n    - expression {numeric}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">d<\/strong><\/p> <ul> <li><strong>gene<\/strong>: Gene identifier<\/li> <li><strong>sample<\/strong>: Sample identifier<\/li> <li><strong>expression<\/strong>: Gene expression value<\/li> <\/ul><\/div>"],"shape":["box"],"font.multi":[true],"font.align":["left"],"x":[null],"y":[null],"color.background":["lightgrey"],"color.border":["black"],"color.highlight.border":["orange"],"color.highlight.background":["lightgrey"],"id":["d"]},"edges":{"id":[],"from":[],"to":[]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="documents-stored-as-base64-values">Documents stored as base64 values<a class="anchor" aria-label="anchor" href="#documents-stored-as-base64-values"></a>
</h4>
<p>Whole documents can be stored in MDB as “base64” character values.
The example below shows how a document can be put in a table and the
corresponding data model.</p>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ch_config_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>   name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"config.xml"</span>, <span class="st">"users.xml"</span><span class="op">)</span>,</span>
<span>   file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="fu">base64enc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/base64enc/man/base64.html" class="external-link">base64encode</a></span><span class="op">(</span></span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"ClickHouse/config.xml"</span>, package<span class="op">=</span><span class="st">"TKCat"</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>      <span class="fu">base64enc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/base64enc/man/base64.html" class="external-link">base64encode</a></span><span class="op">(</span></span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"ClickHouse/users.xml"</span>, package<span class="op">=</span><span class="st">"TKCat"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>   <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/df_to_model.html" class="external-link">df_to_model</a></span><span class="op">(</span><span class="va">ch_config_files</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/update_field.html" class="external-link">update_field</a></span><span class="op">(</span></span>
<span>      <span class="st">"ch_config_files"</span>, <span class="st">"name"</span>,</span>
<span>      type<span class="op">=</span><span class="st">"base64"</span>, comment<span class="op">=</span><span class="st">"Name of the config file"</span>,</span>
<span>      nullable<span class="op">=</span><span class="cn">FALSE</span>, unique<span class="op">=</span><span class="cn">TRUE</span></span>
<span>   <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://patzaw.github.io/ReDaMoR/reference/update_field.html" class="external-link">update_field</a></span><span class="op">(</span></span>
<span>      <span class="st">"ch_config_files"</span>, <span class="st">"file"</span>,</span>
<span>      type<span class="op">=</span><span class="st">"base64"</span>, comment<span class="op">=</span><span class="st">"Config file in base64 format"</span>,</span>
<span>      nullable<span class="op">=</span><span class="cn">FALSE</span></span>
<span>   <span class="op">)</span></span>
<span><span class="va">md</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/memoMDB.html">memoMDB</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ch_config_files<span class="op">=</span><span class="va">ch_config_files</span><span class="op">)</span>, <span class="va">m</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"base64 example"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/data_model.html">data_model</a></span><span class="op">(</span><span class="va">md</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="visNetwork html-widget html-fill-item" id="htmlwidget-416566eb193bf50d04e6" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-416566eb193bf50d04e6">{"x":{"nodes":{"tableName":["ch_config_files"],"label":["<b>ch_config_files<\/b>\n    - *name {base64}\n    - file {base64}"],"title":["<div style=\"max-width:400px; max-height:300px;overflow:scroll;\"><p><strong style=\"text-decoration:underline;\">ch_config_files<\/strong><\/p> <ul> <li><strong>name<\/strong>: Name of the config file<\/li> <li><strong>file<\/strong>: Config file in base64 format<\/li> <\/ul><\/div>"],"shape":["box"],"font.multi":[true],"font.align":["left"],"x":[null],"y":[null],"color.background":["lightgrey"],"color.border":["black"],"color.highlight.border":["orange"],"color.highlight.background":["lightgrey"],"id":["ch_config_files"]},"edges":{"id":[],"from":[],"to":[]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2,"labelHighlightBold":false},"manipulation":{"enabled":false},"edges":{"width":2,"selectionWidth":2},"interaction":{"multiselect":true,"zoomSpeed":1},"layout":{"randomSeed":2},"physics":{"enabled":false}},"groups":null,"width":null,"height":null,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<!-- Bibliography -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-françois2020" class="csl-entry">
François, Liesbeth, Jonathan van Eyll, and Patrice Godard. 2020.
<span>“Dictionary of Disease Ontologies (DODO): A Graph Database to
Facilitate Access and Interaction with Disease and Phenotype
Ontologies.”</span> <em>F1000Research</em> 9 (August): 942. <a href="https://doi.org/10.12688/f1000research.25144.1" class="external-link">https://doi.org/10.12688/f1000research.25144.1</a>.
</div>
<div id="ref-godard_bed:_2018" class="csl-entry">
Godard, Patrice, and Jonathan van Eyll. 2018. <span>“<span>BED</span>: A
<span>Biological</span> <span>Entity</span> <span>Dictionary</span>
Based on a Graph Data Model.”</span> <em>F1000Research</em> 7: 195. <a href="https://doi.org/10.12688/f1000research.13925.3" class="external-link">https://doi.org/10.12688/f1000research.13925.3</a>.
</div>
<div id="ref-kinsella2011" class="csl-entry">
Kinsella, R. J., A. Kahari, S. Haider, J. Zamora, G. Proctor, G.
Spudich, J. Almeida-King, et al. 2011. <span>“Ensembl BioMarts: A Hub
for Data Retrieval Across Taxonomic Space.”</span> <em>Database</em>
2011 (0): bar030–30. <a href="https://doi.org/10.1093/database/bar030" class="external-link">https://doi.org/10.1093/database/bar030</a>.
</div>
<div id="ref-kohler_expansion_2019" class="csl-entry">
Köhler, Sebastian, Leigh Carmody, Nicole Vasilevsky, Julius O B
Jacobsen, Daniel Danis, Jean-Philippe Gourdine, Michael Gargano, et al.
2019. <span>“Expansion of the <span>Human</span> <span>Phenotype</span>
<span>Ontology</span> (<span>HPO</span>) Knowledge Base and
Resources.”</span> <em>Nucleic Acids Research</em> 47 (D1): D1018–27. <a href="https://doi.org/10.1093/nar/gky1105" class="external-link">https://doi.org/10.1093/nar/gky1105</a>.
</div>
<div id="ref-landrum_clinvar_2018" class="csl-entry">
Landrum, Melissa J., Jennifer M. Lee, Mark Benson, Garth R. Brown, Chen
Chao, Shanmuga Chitipiralla, Baoshan Gu, et al. 2018.
<span>“<span>ClinVar</span>: Improving Access to Variant Interpretations
and Supporting Evidence.”</span> <em>Nucleic Acids Research</em> 46
(D1): D1062–67. <a href="https://doi.org/10.1093/nar/gkx1153" class="external-link">https://doi.org/10.1093/nar/gkx1153</a>.
</div>
<div id="ref-mendez_chembl_2019" class="csl-entry">
Mendez, David, Anna Gaulton, A Patrícia Bento, Jon Chambers, Marleen De
Veij, Eloy Félix, María Paula Magariños, et al. 2019.
<span>“<span>ChEMBL</span>: Towards Direct Deposition of Bioassay
Data.”</span> <em>Nucleic Acids Research</em> 47 (D1): D930–40. <a href="https://doi.org/10.1093/nar/gky1075" class="external-link">https://doi.org/10.1093/nar/gky1075</a>.
</div>
<div id="ref-wu2012" class="csl-entry">
Wu, Chunlei, Ian MacLeod, and Andrew I. Su. 2012. <span>“BioGPS and
MyGene.info: Organizing Online, Gene-Centric Information.”</span>
<em>Nucleic Acids Research</em> 41 (D1): D561–65. <a href="https://doi.org/10.1093/nar/gks1114" class="external-link">https://doi.org/10.1093/nar/gks1114</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://patzaw.github.io" class="external-link">Patrice Godard</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
